{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "032a1885-9f47-45f2-ada8-06bfba747110",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 1. Set up working environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "4c719de2-4ed9-4620-878f-180cc68b562d",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
      "Replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "Bioconductor version 3.18 (BiocManager 1.30.23), R 4.3.2 (2023-10-31 ucrt)\n",
      "\n",
      "Warning message:\n",
      "\"package(s) not installed when version(s) same as or greater than current; use\n",
      "  `force = TRUE` to re-install: 'ANCOMBC'\"\n",
      "Old packages: 'colorspace', 'yaml', 'cli', 'digest', 'fastmap', 'htmltools',\n",
      "  'parallelly', 'rlang'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "    install.packages(\"BiocManager\")\n",
    "BiocManager::install(\"ANCOMBC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "69f07a87-ec56-4477-ab37-5da219489b50",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "library(\"ANCOMBC\")\n",
    "library(phyloseq)\n",
    "library(mia)\n",
    "library(tidyverse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e9db2985-5036-495d-bb3b-0e6668923375",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'C:/Users/trang/OneDrive/Máy tính/New folder/Meta_Genomics/ANCOMBC'"
      ],
      "text/latex": [
       "'C:/Users/trang/OneDrive/Máy tính/New folder/Meta\\_Genomics/ANCOMBC'"
      ],
      "text/markdown": [
       "'C:/Users/trang/OneDrive/Máy tính/New folder/Meta_Genomics/ANCOMBC'"
      ],
      "text/plain": [
       "[1] \"C:/Users/trang/OneDrive/Máy tính/New folder/Meta_Genomics/ANCOMBC\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "setwd(\"C:/Users/trang/OneDrive/Máy tính/New folder/Meta_Genomics/ANCOMBC\")\n",
    "getwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53fb150b-f09d-4abf-bd63-4c82f63277a4",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## 2. Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "e68541c5-a11f-465a-87a0-49922264acc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "load('smd') # metadata\n",
    "load('otu_data') # count data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b2a4a777-0b50-4e48-95eb-e466d52e33b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 3 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>stand_bmi</th><th scope=col>stand_choles</th><th scope=col>health_status</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>S1</th><td>-7.046117</td><td>-0.05133271</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>S2</th><td>-8.183509</td><td>-1.30178093</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>S3</th><td>-7.369319</td><td>-1.20561056</td><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 3 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & stand\\_bmi & stand\\_choles & health\\_status\\\\\n",
       "  & <dbl> & <dbl> & <fct>\\\\\n",
       "\\hline\n",
       "\tS1 & -7.046117 & -0.05133271 & 1\\\\\n",
       "\tS2 & -8.183509 & -1.30178093 & 1\\\\\n",
       "\tS3 & -7.369319 & -1.20561056 & 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 3 × 3\n",
       "\n",
       "| <!--/--> | stand_bmi &lt;dbl&gt; | stand_choles &lt;dbl&gt; | health_status &lt;fct&gt; |\n",
       "|---|---|---|---|\n",
       "| S1 | -7.046117 | -0.05133271 | 1 |\n",
       "| S2 | -8.183509 | -1.30178093 | 1 |\n",
       "| S3 | -7.369319 | -1.20561056 | 1 |\n",
       "\n"
      ],
      "text/plain": [
       "   stand_bmi stand_choles health_status\n",
       "S1 -7.046117 -0.05133271  1            \n",
       "S2 -8.183509 -1.30178093  1            \n",
       "S3 -7.369319 -1.20561056  1            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 3 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>stand_bmi</th><th scope=col>stand_choles</th><th scope=col>health_status</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>S148</th><td>-2.439321</td><td> 0.5519218</td><td>3</td></tr>\n",
       "\t<tr><th scope=row>S149</th><td>-3.094921</td><td> 0.7709166</td><td>3</td></tr>\n",
       "\t<tr><th scope=row>S150</th><td>-2.664466</td><td>-0.5753851</td><td>3</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 3 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & stand\\_bmi & stand\\_choles & health\\_status\\\\\n",
       "  & <dbl> & <dbl> & <fct>\\\\\n",
       "\\hline\n",
       "\tS148 & -2.439321 &  0.5519218 & 3\\\\\n",
       "\tS149 & -3.094921 &  0.7709166 & 3\\\\\n",
       "\tS150 & -2.664466 & -0.5753851 & 3\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 3 × 3\n",
       "\n",
       "| <!--/--> | stand_bmi &lt;dbl&gt; | stand_choles &lt;dbl&gt; | health_status &lt;fct&gt; |\n",
       "|---|---|---|---|\n",
       "| S148 | -2.439321 |  0.5519218 | 3 |\n",
       "| S149 | -3.094921 |  0.7709166 | 3 |\n",
       "| S150 | -2.664466 | -0.5753851 | 3 |\n",
       "\n"
      ],
      "text/plain": [
       "     stand_bmi stand_choles health_status\n",
       "S148 -2.439321  0.5519218   3            \n",
       "S149 -3.094921  0.7709166   3            \n",
       "S150 -2.664466 -0.5753851   3            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(smd,3)\n",
    "tail(smd,3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e7f5dcf8-44e9-4e85-aa60-dfbdbee412ea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       " 1  2  3 \n",
       "50 50 50 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# 3 levels of heatlh status\n",
    "table(smd$health_status)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "074707ad-cdd3-4b8f-b694-aee383f66750",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 3 × 4 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>stand_choles</th><th scope=col>health_status1</th><th scope=col>health_status2</th><th scope=col>health_status3</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>S1</th><td>-0.05133271</td><td>1</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>S2</th><td>-1.30178093</td><td>1</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>S3</th><td>-1.20561056</td><td>1</td><td>0</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 3 × 4 of type dbl\n",
       "\\begin{tabular}{r|llll}\n",
       "  & stand\\_choles & health\\_status1 & health\\_status2 & health\\_status3\\\\\n",
       "\\hline\n",
       "\tS1 & -0.05133271 & 1 & 0 & 0\\\\\n",
       "\tS2 & -1.30178093 & 1 & 0 & 0\\\\\n",
       "\tS3 & -1.20561056 & 1 & 0 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 3 × 4 of type dbl\n",
       "\n",
       "| <!--/--> | stand_choles | health_status1 | health_status2 | health_status3 |\n",
       "|---|---|---|---|---|\n",
       "| S1 | -0.05133271 | 1 | 0 | 0 |\n",
       "| S2 | -1.30178093 | 1 | 0 | 0 |\n",
       "| S3 | -1.20561056 | 1 | 0 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "   stand_choles health_status1 health_status2 health_status3\n",
       "S1 -0.05133271  1              0              0             \n",
       "S2 -1.30178093  1              0              0             \n",
       "S3 -1.20561056  1              0              0             "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Meta data of each sample in extended form\n",
    "smd_dmy = model.matrix(~ 0 + stand_choles  + health_status, data = smd)\n",
    "head(smd_dmy,3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "21788efe-4e52-483e-b60d-299a0b5b3cb9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 3 × 4 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>stand_choles</th><th scope=col>health_status1</th><th scope=col>health_status2</th><th scope=col>health_status3</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>S148</th><td> 0.5519218</td><td>0</td><td>0</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>S149</th><td> 0.7709166</td><td>0</td><td>0</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>S150</th><td>-0.5753851</td><td>0</td><td>0</td><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 3 × 4 of type dbl\n",
       "\\begin{tabular}{r|llll}\n",
       "  & stand\\_choles & health\\_status1 & health\\_status2 & health\\_status3\\\\\n",
       "\\hline\n",
       "\tS148 &  0.5519218 & 0 & 0 & 1\\\\\n",
       "\tS149 &  0.7709166 & 0 & 0 & 1\\\\\n",
       "\tS150 & -0.5753851 & 0 & 0 & 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 3 × 4 of type dbl\n",
       "\n",
       "| <!--/--> | stand_choles | health_status1 | health_status2 | health_status3 |\n",
       "|---|---|---|---|---|\n",
       "| S148 |  0.5519218 | 0 | 0 | 1 |\n",
       "| S149 |  0.7709166 | 0 | 0 | 1 |\n",
       "| S150 | -0.5753851 | 0 | 0 | 1 |\n",
       "\n"
      ],
      "text/plain": [
       "     stand_choles health_status1 health_status2 health_status3\n",
       "S148  0.5519218   0              0              1             \n",
       "S149  0.7709166   0              0              1             \n",
       "S150 -0.5753851   0              0              1             "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "tail(smd_dmy,3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4194efe7-199c-474a-b291-36eebd640f57",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 7 × 150 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>S1</th><th scope=col>S2</th><th scope=col>S3</th><th scope=col>S4</th><th scope=col>S5</th><th scope=col>S6</th><th scope=col>S7</th><th scope=col>S8</th><th scope=col>S9</th><th scope=col>S10</th><th scope=col>⋯</th><th scope=col>S141</th><th scope=col>S142</th><th scope=col>S143</th><th scope=col>S144</th><th scope=col>S145</th><th scope=col>S146</th><th scope=col>S147</th><th scope=col>S148</th><th scope=col>S149</th><th scope=col>S150</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>T1</th><td>  1</td><td>0</td><td>  1</td><td>  3</td><td>  2</td><td>0</td><td> 1</td><td> 0</td><td> 0</td><td>  3</td><td>⋯</td><td>  142</td><td>  2</td><td>  208</td><td>   232</td><td>  105</td><td>   64</td><td>   317</td><td>   41</td><td>  320</td><td>  14</td></tr>\n",
       "\t<tr><th scope=row>T2</th><td>102</td><td>1</td><td>156</td><td>535</td><td>178</td><td>0</td><td>99</td><td>21</td><td>28</td><td>400</td><td>⋯</td><td>41555</td><td>940</td><td>69158</td><td>150981</td><td>68451</td><td>40298</td><td>258659</td><td>23841</td><td>79188</td><td>7491</td></tr>\n",
       "\t<tr><th scope=row>T3</th><td>  0</td><td>0</td><td>  0</td><td>  0</td><td>  0</td><td>0</td><td> 0</td><td> 0</td><td> 0</td><td>  0</td><td>⋯</td><td>   17</td><td>  0</td><td>   22</td><td>    23</td><td>   10</td><td>    7</td><td>    32</td><td>    4</td><td>   35</td><td>   2</td></tr>\n",
       "\t<tr><th scope=row>T4</th><td>  4</td><td>0</td><td>  7</td><td> 23</td><td> 11</td><td>0</td><td> 6</td><td> 1</td><td> 1</td><td> 19</td><td>⋯</td><td>  913</td><td> 12</td><td> 1364</td><td>  1673</td><td>  767</td><td>  465</td><td>  2340</td><td>  293</td><td> 2036</td><td>  96</td></tr>\n",
       "\t<tr><th scope=row>T5</th><td> 38</td><td>0</td><td> 65</td><td>214</td><td> 90</td><td>0</td><td>52</td><td> 7</td><td> 9</td><td>173</td><td>⋯</td><td> 7678</td><td>118</td><td>11913</td><td> 17441</td><td> 7893</td><td> 4766</td><td> 26071</td><td> 2888</td><td>16343</td><td> 975</td></tr>\n",
       "\t<tr><th scope=row>T6</th><td> 15</td><td>0</td><td> 26</td><td> 85</td><td> 38</td><td>0</td><td>23</td><td> 3</td><td> 3</td><td> 70</td><td>⋯</td><td> 3302</td><td> 45</td><td> 4985</td><td>  6334</td><td> 2864</td><td> 1756</td><td>  8992</td><td> 1100</td><td> 7295</td><td> 364</td></tr>\n",
       "\t<tr><th scope=row>T7</th><td> 11</td><td>0</td><td> 18</td><td> 60</td><td> 27</td><td>0</td><td>15</td><td> 2</td><td> 2</td><td> 50</td><td>⋯</td><td> 6187</td><td> 87</td><td> 9441</td><td> 12614</td><td> 5712</td><td> 3455</td><td> 18203</td><td> 2122</td><td>13587</td><td> 724</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 7 × 150 of type dbl\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & S1 & S2 & S3 & S4 & S5 & S6 & S7 & S8 & S9 & S10 & ⋯ & S141 & S142 & S143 & S144 & S145 & S146 & S147 & S148 & S149 & S150\\\\\n",
       "\\hline\n",
       "\tT1 &   1 & 0 &   1 &   3 &   2 & 0 &  1 &  0 &  0 &   3 & ⋯ &   142 &   2 &   208 &    232 &   105 &    64 &    317 &    41 &   320 &   14\\\\\n",
       "\tT2 & 102 & 1 & 156 & 535 & 178 & 0 & 99 & 21 & 28 & 400 & ⋯ & 41555 & 940 & 69158 & 150981 & 68451 & 40298 & 258659 & 23841 & 79188 & 7491\\\\\n",
       "\tT3 &   0 & 0 &   0 &   0 &   0 & 0 &  0 &  0 &  0 &   0 & ⋯ &    17 &   0 &    22 &     23 &    10 &     7 &     32 &     4 &    35 &    2\\\\\n",
       "\tT4 &   4 & 0 &   7 &  23 &  11 & 0 &  6 &  1 &  1 &  19 & ⋯ &   913 &  12 &  1364 &   1673 &   767 &   465 &   2340 &   293 &  2036 &   96\\\\\n",
       "\tT5 &  38 & 0 &  65 & 214 &  90 & 0 & 52 &  7 &  9 & 173 & ⋯ &  7678 & 118 & 11913 &  17441 &  7893 &  4766 &  26071 &  2888 & 16343 &  975\\\\\n",
       "\tT6 &  15 & 0 &  26 &  85 &  38 & 0 & 23 &  3 &  3 &  70 & ⋯ &  3302 &  45 &  4985 &   6334 &  2864 &  1756 &   8992 &  1100 &  7295 &  364\\\\\n",
       "\tT7 &  11 & 0 &  18 &  60 &  27 & 0 & 15 &  2 &  2 &  50 & ⋯ &  6187 &  87 &  9441 &  12614 &  5712 &  3455 &  18203 &  2122 & 13587 &  724\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 7 × 150 of type dbl\n",
       "\n",
       "| <!--/--> | S1 | S2 | S3 | S4 | S5 | S6 | S7 | S8 | S9 | S10 | ⋯ | S141 | S142 | S143 | S144 | S145 | S146 | S147 | S148 | S149 | S150 |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| T1 |   1 | 0 |   1 |   3 |   2 | 0 |  1 |  0 |  0 |   3 | ⋯ |   142 |   2 |   208 |    232 |   105 |    64 |    317 |    41 |   320 |   14 |\n",
       "| T2 | 102 | 1 | 156 | 535 | 178 | 0 | 99 | 21 | 28 | 400 | ⋯ | 41555 | 940 | 69158 | 150981 | 68451 | 40298 | 258659 | 23841 | 79188 | 7491 |\n",
       "| T3 |   0 | 0 |   0 |   0 |   0 | 0 |  0 |  0 |  0 |   0 | ⋯ |    17 |   0 |    22 |     23 |    10 |     7 |     32 |     4 |    35 |    2 |\n",
       "| T4 |   4 | 0 |   7 |  23 |  11 | 0 |  6 |  1 |  1 |  19 | ⋯ |   913 |  12 |  1364 |   1673 |   767 |   465 |   2340 |   293 |  2036 |   96 |\n",
       "| T5 |  38 | 0 |  65 | 214 |  90 | 0 | 52 |  7 |  9 | 173 | ⋯ |  7678 | 118 | 11913 |  17441 |  7893 |  4766 |  26071 |  2888 | 16343 |  975 |\n",
       "| T6 |  15 | 0 |  26 |  85 |  38 | 0 | 23 |  3 |  3 |  70 | ⋯ |  3302 |  45 |  4985 |   6334 |  2864 |  1756 |   8992 |  1100 |  7295 |  364 |\n",
       "| T7 |  11 | 0 |  18 |  60 |  27 | 0 | 15 |  2 |  2 |  50 | ⋯ |  6187 |  87 |  9441 |  12614 |  5712 |  3455 |  18203 |  2122 | 13587 |  724 |\n",
       "\n"
      ],
      "text/plain": [
       "   S1  S2 S3  S4  S5  S6 S7 S8 S9 S10 ⋯ S141  S142 S143  S144   S145  S146 \n",
       "T1   1 0    1   3   2 0   1  0  0   3 ⋯   142   2    208    232   105    64\n",
       "T2 102 1  156 535 178 0  99 21 28 400 ⋯ 41555 940  69158 150981 68451 40298\n",
       "T3   0 0    0   0   0 0   0  0  0   0 ⋯    17   0     22     23    10     7\n",
       "T4   4 0    7  23  11 0   6  1  1  19 ⋯   913  12   1364   1673   767   465\n",
       "T5  38 0   65 214  90 0  52  7  9 173 ⋯  7678 118  11913  17441  7893  4766\n",
       "T6  15 0   26  85  38 0  23  3  3  70 ⋯  3302  45   4985   6334  2864  1756\n",
       "T7  11 0   18  60  27 0  15  2  2  50 ⋯  6187  87   9441  12614  5712  3455\n",
       "   S147   S148  S149  S150\n",
       "T1    317    41   320   14\n",
       "T2 258659 23841 79188 7491\n",
       "T3     32     4    35    2\n",
       "T4   2340   293  2036   96\n",
       "T5  26071  2888 16343  975\n",
       "T6   8992  1100  7295  364\n",
       "T7  18203  2122 13587  724"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(otu_data,7) # raw count data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "627d8477-9c1a-4172-a8a3-5bdb6b02dc92",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## 3.1 Differential Abundance Analysis: Health status 1 and 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "e28980d1-0789-4a44-af91-a010cdfbf971",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 3 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>stand_bmi</th><th scope=col>stand_choles</th><th scope=col>health_status</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>S1</th><td>-7.046117</td><td>-0.05133271</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>S2</th><td>-8.183509</td><td>-1.30178093</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>S3</th><td>-7.369319</td><td>-1.20561056</td><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 3 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & stand\\_bmi & stand\\_choles & health\\_status\\\\\n",
       "  & <dbl> & <dbl> & <fct>\\\\\n",
       "\\hline\n",
       "\tS1 & -7.046117 & -0.05133271 & 1\\\\\n",
       "\tS2 & -8.183509 & -1.30178093 & 1\\\\\n",
       "\tS3 & -7.369319 & -1.20561056 & 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 3 × 3\n",
       "\n",
       "| <!--/--> | stand_bmi &lt;dbl&gt; | stand_choles &lt;dbl&gt; | health_status &lt;fct&gt; |\n",
       "|---|---|---|---|\n",
       "| S1 | -7.046117 | -0.05133271 | 1 |\n",
       "| S2 | -8.183509 | -1.30178093 | 1 |\n",
       "| S3 | -7.369319 | -1.20561056 | 1 |\n",
       "\n"
      ],
      "text/plain": [
       "   stand_bmi stand_choles health_status\n",
       "S1 -7.046117 -0.05133271  1            \n",
       "S2 -8.183509 -1.30178093  1            \n",
       "S3 -7.369319 -1.20561056  1            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Meta data\n",
    "head(smd,3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "a31fb468-2e65-4604-9adf-6bdfe14885b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       " 1  2  3 \n",
       "50 50 50 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "table(smd$health_status)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "e918c7d9-917e-452c-83bf-6844e8bb00d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       " 1  2 \n",
       "50 50 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Meta data : Choose only sample with health status 1 or 2\n",
    "smd_H1_H2 = smd[smd$health_status != 3,]\n",
    "smd_H1_H2$health_status = factor(smd_H1_H2$health_status, levels = c(1,2))\n",
    "table(smd_H1_H2$health_status)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "614309b9-ca6b-467f-809e-aa12a8cc740f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 2 × 150 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>S1</th><th scope=col>S2</th><th scope=col>S3</th><th scope=col>S4</th><th scope=col>S5</th><th scope=col>S6</th><th scope=col>S7</th><th scope=col>S8</th><th scope=col>S9</th><th scope=col>S10</th><th scope=col>⋯</th><th scope=col>S141</th><th scope=col>S142</th><th scope=col>S143</th><th scope=col>S144</th><th scope=col>S145</th><th scope=col>S146</th><th scope=col>S147</th><th scope=col>S148</th><th scope=col>S149</th><th scope=col>S150</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>T1</th><td>  1</td><td>0</td><td>  1</td><td>  3</td><td>  2</td><td>0</td><td> 1</td><td> 0</td><td> 0</td><td>  3</td><td>⋯</td><td>  142</td><td>  2</td><td>  208</td><td>   232</td><td>  105</td><td>   64</td><td>   317</td><td>   41</td><td>  320</td><td>  14</td></tr>\n",
       "\t<tr><th scope=row>T2</th><td>102</td><td>1</td><td>156</td><td>535</td><td>178</td><td>0</td><td>99</td><td>21</td><td>28</td><td>400</td><td>⋯</td><td>41555</td><td>940</td><td>69158</td><td>150981</td><td>68451</td><td>40298</td><td>258659</td><td>23841</td><td>79188</td><td>7491</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 2 × 150 of type dbl\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & S1 & S2 & S3 & S4 & S5 & S6 & S7 & S8 & S9 & S10 & ⋯ & S141 & S142 & S143 & S144 & S145 & S146 & S147 & S148 & S149 & S150\\\\\n",
       "\\hline\n",
       "\tT1 &   1 & 0 &   1 &   3 &   2 & 0 &  1 &  0 &  0 &   3 & ⋯ &   142 &   2 &   208 &    232 &   105 &    64 &    317 &    41 &   320 &   14\\\\\n",
       "\tT2 & 102 & 1 & 156 & 535 & 178 & 0 & 99 & 21 & 28 & 400 & ⋯ & 41555 & 940 & 69158 & 150981 & 68451 & 40298 & 258659 & 23841 & 79188 & 7491\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 2 × 150 of type dbl\n",
       "\n",
       "| <!--/--> | S1 | S2 | S3 | S4 | S5 | S6 | S7 | S8 | S9 | S10 | ⋯ | S141 | S142 | S143 | S144 | S145 | S146 | S147 | S148 | S149 | S150 |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| T1 |   1 | 0 |   1 |   3 |   2 | 0 |  1 |  0 |  0 |   3 | ⋯ |   142 |   2 |   208 |    232 |   105 |    64 |    317 |    41 |   320 |   14 |\n",
       "| T2 | 102 | 1 | 156 | 535 | 178 | 0 | 99 | 21 | 28 | 400 | ⋯ | 41555 | 940 | 69158 | 150981 | 68451 | 40298 | 258659 | 23841 | 79188 | 7491 |\n",
       "\n"
      ],
      "text/plain": [
       "   S1  S2 S3  S4  S5  S6 S7 S8 S9 S10 ⋯ S141  S142 S143  S144   S145  S146 \n",
       "T1   1 0    1   3   2 0   1  0  0   3 ⋯   142   2    208    232   105    64\n",
       "T2 102 1  156 535 178 0  99 21 28 400 ⋯ 41555 940  69158 150981 68451 40298\n",
       "   S147   S148  S149  S150\n",
       "T1    317    41   320   14\n",
       "T2 258659 23841 79188 7491"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Count data\n",
    "head(otu_data,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "1282d4b7-4535-484c-890d-46c0afc39df8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>91</li><li>100</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 91\n",
       "\\item 100\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 91\n",
       "2. 100\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1]  91 100"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 3 × 100 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>S1</th><th scope=col>S2</th><th scope=col>S3</th><th scope=col>S4</th><th scope=col>S5</th><th scope=col>S6</th><th scope=col>S7</th><th scope=col>S8</th><th scope=col>S9</th><th scope=col>S10</th><th scope=col>⋯</th><th scope=col>S91</th><th scope=col>S92</th><th scope=col>S93</th><th scope=col>S94</th><th scope=col>S95</th><th scope=col>S96</th><th scope=col>S97</th><th scope=col>S98</th><th scope=col>S99</th><th scope=col>S100</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>T1</th><td>  1</td><td>0</td><td>  1</td><td>  3</td><td>  2</td><td>0</td><td> 1</td><td> 0</td><td> 0</td><td>  3</td><td>⋯</td><td>  51</td><td>  8</td><td>0</td><td>0</td><td>0</td><td> 29</td><td> 4</td><td> 22</td><td>0</td><td> 1</td></tr>\n",
       "\t<tr><th scope=row>T2</th><td>102</td><td>1</td><td>156</td><td>535</td><td>178</td><td>0</td><td>99</td><td>21</td><td>28</td><td>400</td><td>⋯</td><td>1213</td><td>256</td><td>9</td><td>1</td><td>1</td><td>787</td><td>76</td><td>783</td><td>0</td><td>11</td></tr>\n",
       "\t<tr><th scope=row>T3</th><td>  0</td><td>0</td><td>  0</td><td>  0</td><td>  0</td><td>0</td><td> 0</td><td> 0</td><td> 0</td><td>  0</td><td>⋯</td><td>   5</td><td>  1</td><td>0</td><td>0</td><td>0</td><td>  3</td><td> 0</td><td>  2</td><td>0</td><td> 0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 3 × 100 of type dbl\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & S1 & S2 & S3 & S4 & S5 & S6 & S7 & S8 & S9 & S10 & ⋯ & S91 & S92 & S93 & S94 & S95 & S96 & S97 & S98 & S99 & S100\\\\\n",
       "\\hline\n",
       "\tT1 &   1 & 0 &   1 &   3 &   2 & 0 &  1 &  0 &  0 &   3 & ⋯ &   51 &   8 & 0 & 0 & 0 &  29 &  4 &  22 & 0 &  1\\\\\n",
       "\tT2 & 102 & 1 & 156 & 535 & 178 & 0 & 99 & 21 & 28 & 400 & ⋯ & 1213 & 256 & 9 & 1 & 1 & 787 & 76 & 783 & 0 & 11\\\\\n",
       "\tT3 &   0 & 0 &   0 &   0 &   0 & 0 &  0 &  0 &  0 &   0 & ⋯ &    5 &   1 & 0 & 0 & 0 &   3 &  0 &   2 & 0 &  0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 3 × 100 of type dbl\n",
       "\n",
       "| <!--/--> | S1 | S2 | S3 | S4 | S5 | S6 | S7 | S8 | S9 | S10 | ⋯ | S91 | S92 | S93 | S94 | S95 | S96 | S97 | S98 | S99 | S100 |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| T1 |   1 | 0 |   1 |   3 |   2 | 0 |  1 |  0 |  0 |   3 | ⋯ |   51 |   8 | 0 | 0 | 0 |  29 |  4 |  22 | 0 |  1 |\n",
       "| T2 | 102 | 1 | 156 | 535 | 178 | 0 | 99 | 21 | 28 | 400 | ⋯ | 1213 | 256 | 9 | 1 | 1 | 787 | 76 | 783 | 0 | 11 |\n",
       "| T3 |   0 | 0 |   0 |   0 |   0 | 0 |  0 |  0 |  0 |   0 | ⋯ |    5 |   1 | 0 | 0 | 0 |   3 |  0 |   2 | 0 |  0 |\n",
       "\n"
      ],
      "text/plain": [
       "   S1  S2 S3  S4  S5  S6 S7 S8 S9 S10 ⋯ S91  S92 S93 S94 S95 S96 S97 S98 S99\n",
       "T1   1 0    1   3   2 0   1  0  0   3 ⋯   51   8 0   0   0    29  4   22 0  \n",
       "T2 102 1  156 535 178 0  99 21 28 400 ⋯ 1213 256 9   1   1   787 76  783 0  \n",
       "T3   0 0    0   0   0 0   0  0  0   0 ⋯    5   1 0   0   0     3  0    2 0  \n",
       "   S100\n",
       "T1  1  \n",
       "T2 11  \n",
       "T3  0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Count data : Choose only sample with health status 1 or 2\n",
    "otu_data_H1_H2 <- otu_data[, colnames(otu_data) %in% rownames(smd_H1_H2)]\n",
    "dim(otu_data_H1_H2)\n",
    "head(otu_data_H1_H2,3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "11a44ac8-c05c-4b1c-860e-ed14e5a1c5af",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the tse object\n",
    "assays = S4Vectors::SimpleList(counts = otu_data_H1_H2) # count data\n",
    "smd_H1_H2 = S4Vectors::DataFrame(smd_H1_H2) # meta data\n",
    "tse = TreeSummarizedExperiment::TreeSummarizedExperiment(assays = assays, colData = smd_H1_H2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "6f7a2c42-2fcf-4a7a-8d5b-c1c3981f9f30",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: TreeSummarizedExperiment \n",
       "dim: 91 100 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(91): T1 T2 ... T90 T91\n",
       "rowData names(0):\n",
       "colnames(100): S1 S2 ... S99 S100\n",
       "colData names(3): stand_bmi stand_choles health_status\n",
       "reducedDimNames(0):\n",
       "mainExpName: NULL\n",
       "altExpNames(0):\n",
       "rowLinks: NULL\n",
       "rowTree: NULL\n",
       "colLinks: NULL\n",
       "colTree: NULL"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "tse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "a6731132-b974-42e7-ba38-69eda07d91bf",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"The group variable has < 3 categories \n",
      "The multi-group comparisons (global/pairwise/dunnet/trend) will be deactivated\"\n",
      "Obtaining initial estimates ...\n",
      "\n",
      "Estimating sample-specific biases ...\n",
      "\n",
      "Warning message in sqrt(sweep(var_hat, 2, var_delta, \"*\")):\n",
      "\"NaNs produced\"\n",
      "Sensitivity analysis for pseudo-count addition to 0s: ...\n",
      "\n",
      "ANCOM-BC2 primary results ...\n",
      "\n"
     ]
    }
   ],
   "source": [
    "set.seed(123)\n",
    "output = ancombc2(data = tse, \n",
    "                  assay_name = \"counts\",\n",
    "                  fix_formula = \"health_status\", # response ~ covariate\n",
    "                  p_adj_method = \"holm\",\n",
    "                  pseudo_sens = TRUE,                 # sensitivity analysis to the pseudo-count addition\n",
    "                  prv_cut = 0.10,                     # (0,1) Taxa with prevalences less than prv_cut will be excluded\n",
    "                  lib_cut = 1000,                     # filter sampling with sampling depth\n",
    "                  s0_perc = 0.05,                     # advoid significant feature due to small standard error, (rare taxa)\n",
    "                  group = \"health_status\",            # Group to be differencially analyzed (healthy status - 1,2)             \n",
    "                  alpha = 0.05,                       # Level of significant\n",
    "                  n_cl = 2,                           # CPU cores to compute\n",
    "                  verbose = TRUE,\n",
    "                  pairwise = FALSE,                   # CUP cores to compute\n",
    "                  iter_control = list(tol = 1e-5, max_iter = 20, # Liner regression model parameter\n",
    "                                      verbose = FALSE),\n",
    "                  em_control = list(tol = 1e-5, max_iter = 100),\n",
    "                  lme_control = NULL, \n",
    "                  mdfdr_control = list(fwer_ctrl_method = \"holm\", B = 100))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "10a71799-0697-4e29-8fd0-351f1d3a27d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 2 × 93 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>S1</th><th scope=col>S3</th><th scope=col>S4</th><th scope=col>S5</th><th scope=col>S7</th><th scope=col>S8</th><th scope=col>S9</th><th scope=col>S10</th><th scope=col>S11</th><th scope=col>S12</th><th scope=col>⋯</th><th scope=col>S89</th><th scope=col>S90</th><th scope=col>S91</th><th scope=col>S92</th><th scope=col>S93</th><th scope=col>S94</th><th scope=col>S96</th><th scope=col>S97</th><th scope=col>S98</th><th scope=col>S100</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>T1</th><td>  1</td><td>  1</td><td>  3</td><td>  2</td><td> 1</td><td> 0</td><td> 0</td><td>  3</td><td> 0</td><td>  3</td><td>⋯</td><td>0</td><td> 1</td><td>  51</td><td>  8</td><td>0</td><td>0</td><td> 29</td><td> 4</td><td> 22</td><td> 1</td></tr>\n",
       "\t<tr><th scope=row>T2</th><td>102</td><td>156</td><td>535</td><td>178</td><td>99</td><td>21</td><td>28</td><td>400</td><td>60</td><td>486</td><td>⋯</td><td>5</td><td>26</td><td>1213</td><td>256</td><td>9</td><td>1</td><td>787</td><td>76</td><td>783</td><td>11</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 2 × 93 of type dbl\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & S1 & S3 & S4 & S5 & S7 & S8 & S9 & S10 & S11 & S12 & ⋯ & S89 & S90 & S91 & S92 & S93 & S94 & S96 & S97 & S98 & S100\\\\\n",
       "\\hline\n",
       "\tT1 &   1 &   1 &   3 &   2 &  1 &  0 &  0 &   3 &  0 &   3 & ⋯ & 0 &  1 &   51 &   8 & 0 & 0 &  29 &  4 &  22 &  1\\\\\n",
       "\tT2 & 102 & 156 & 535 & 178 & 99 & 21 & 28 & 400 & 60 & 486 & ⋯ & 5 & 26 & 1213 & 256 & 9 & 1 & 787 & 76 & 783 & 11\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 2 × 93 of type dbl\n",
       "\n",
       "| <!--/--> | S1 | S3 | S4 | S5 | S7 | S8 | S9 | S10 | S11 | S12 | ⋯ | S89 | S90 | S91 | S92 | S93 | S94 | S96 | S97 | S98 | S100 |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| T1 |   1 |   1 |   3 |   2 |  1 |  0 |  0 |   3 |  0 |   3 | ⋯ | 0 |  1 |   51 |   8 | 0 | 0 |  29 |  4 |  22 |  1 |\n",
       "| T2 | 102 | 156 | 535 | 178 | 99 | 21 | 28 | 400 | 60 | 486 | ⋯ | 5 | 26 | 1213 | 256 | 9 | 1 | 787 | 76 | 783 | 11 |\n",
       "\n"
      ],
      "text/plain": [
       "   S1  S3  S4  S5  S7 S8 S9 S10 S11 S12 ⋯ S89 S90 S91  S92 S93 S94 S96 S97 S98\n",
       "T1   1   1   3   2  1  0  0   3  0    3 ⋯ 0    1    51   8 0   0    29  4   22\n",
       "T2 102 156 535 178 99 21 28 400 60  486 ⋯ 5   26  1213 256 9   1   787 76  783\n",
       "   S100\n",
       "T1  1  \n",
       "T2 11  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(output$feature_table,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "8cecaaa0-6335-42b7-afb4-4f5536fa99ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 93</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>S1</th><th scope=col>S3</th><th scope=col>S4</th><th scope=col>S5</th><th scope=col>S7</th><th scope=col>S8</th><th scope=col>S9</th><th scope=col>S10</th><th scope=col>S11</th><th scope=col>S12</th><th scope=col>⋯</th><th scope=col>S89</th><th scope=col>S90</th><th scope=col>S91</th><th scope=col>S92</th><th scope=col>S93</th><th scope=col>S94</th><th scope=col>S96</th><th scope=col>S97</th><th scope=col>S98</th><th scope=col>S100</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>T1</th><td>-0.138931</td><td>-0.4848042</td><td>-0.7692875</td><td>-0.4039632</td><td>-0.5681792</td><td>      NA</td><td>       NA</td><td>-0.4074126</td><td>      NA</td><td>-0.8452492</td><td>⋯</td><td>      NA</td><td>-1.027670</td><td>-0.5299654</td><td>-0.9886069</td><td>        NA</td><td>       NA</td><td>-0.8305132</td><td>-0.6235703</td><td>-0.9671347</td><td>-0.1206625</td></tr>\n",
       "\t<tr><th scope=row>T2</th><td> 1.055422</td><td> 1.1344323</td><td> 0.9837475</td><td> 0.6540537</td><td> 0.5963212</td><td>1.154336</td><td>0.8960626</td><td> 1.0548202</td><td>1.206111</td><td> 0.8117277</td><td>⋯</td><td>-1.50715</td><td>-1.200193</td><td>-0.7915586</td><td>-0.9534905</td><td>-0.7103483</td><td>-1.481377</td><td>-0.9602002</td><td>-1.1097508</td><td>-0.8256639</td><td>-1.1533867</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 93\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & S1 & S3 & S4 & S5 & S7 & S8 & S9 & S10 & S11 & S12 & ⋯ & S89 & S90 & S91 & S92 & S93 & S94 & S96 & S97 & S98 & S100\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\tT1 & -0.138931 & -0.4848042 & -0.7692875 & -0.4039632 & -0.5681792 &       NA &        NA & -0.4074126 &       NA & -0.8452492 & ⋯ &       NA & -1.027670 & -0.5299654 & -0.9886069 &         NA &        NA & -0.8305132 & -0.6235703 & -0.9671347 & -0.1206625\\\\\n",
       "\tT2 &  1.055422 &  1.1344323 &  0.9837475 &  0.6540537 &  0.5963212 & 1.154336 & 0.8960626 &  1.0548202 & 1.206111 &  0.8117277 & ⋯ & -1.50715 & -1.200193 & -0.7915586 & -0.9534905 & -0.7103483 & -1.481377 & -0.9602002 & -1.1097508 & -0.8256639 & -1.1533867\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 93\n",
       "\n",
       "| <!--/--> | S1 &lt;dbl&gt; | S3 &lt;dbl&gt; | S4 &lt;dbl&gt; | S5 &lt;dbl&gt; | S7 &lt;dbl&gt; | S8 &lt;dbl&gt; | S9 &lt;dbl&gt; | S10 &lt;dbl&gt; | S11 &lt;dbl&gt; | S12 &lt;dbl&gt; | ⋯ ⋯ | S89 &lt;dbl&gt; | S90 &lt;dbl&gt; | S91 &lt;dbl&gt; | S92 &lt;dbl&gt; | S93 &lt;dbl&gt; | S94 &lt;dbl&gt; | S96 &lt;dbl&gt; | S97 &lt;dbl&gt; | S98 &lt;dbl&gt; | S100 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| T1 | -0.138931 | -0.4848042 | -0.7692875 | -0.4039632 | -0.5681792 |       NA |        NA | -0.4074126 |       NA | -0.8452492 | ⋯ |       NA | -1.027670 | -0.5299654 | -0.9886069 |         NA |        NA | -0.8305132 | -0.6235703 | -0.9671347 | -0.1206625 |\n",
       "| T2 |  1.055422 |  1.1344323 |  0.9837475 |  0.6540537 |  0.5963212 | 1.154336 | 0.8960626 |  1.0548202 | 1.206111 |  0.8117277 | ⋯ | -1.50715 | -1.200193 | -0.7915586 | -0.9534905 | -0.7103483 | -1.481377 | -0.9602002 | -1.1097508 | -0.8256639 | -1.1533867 |\n",
       "\n"
      ],
      "text/plain": [
       "   S1        S3         S4         S5         S7         S8       S9       \n",
       "T1 -0.138931 -0.4848042 -0.7692875 -0.4039632 -0.5681792       NA        NA\n",
       "T2  1.055422  1.1344323  0.9837475  0.6540537  0.5963212 1.154336 0.8960626\n",
       "   S10        S11      S12        ⋯ S89      S90       S91        S92       \n",
       "T1 -0.4074126       NA -0.8452492 ⋯       NA -1.027670 -0.5299654 -0.9886069\n",
       "T2  1.0548202 1.206111  0.8117277 ⋯ -1.50715 -1.200193 -0.7915586 -0.9534905\n",
       "   S93        S94       S96        S97        S98        S100      \n",
       "T1         NA        NA -0.8305132 -0.6235703 -0.9671347 -0.1206625\n",
       "T2 -0.7103483 -1.481377 -0.9602002 -1.1097508 -0.8256639 -1.1533867"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(output$bias_correct_log_table,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "687a5212-ff88-475c-9eae-93d7487eb64e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'taxon'</li><li>'lfc_(Intercept)'</li><li>'lfc_health_status2'</li><li>'se_(Intercept)'</li><li>'se_health_status2'</li><li>'W_(Intercept)'</li><li>'W_health_status2'</li><li>'p_(Intercept)'</li><li>'p_health_status2'</li><li>'q_(Intercept)'</li><li>'q_health_status2'</li><li>'diff_(Intercept)'</li><li>'diff_health_status2'</li><li>'passed_ss_(Intercept)'</li><li>'passed_ss_health_status2'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'taxon'\n",
       "\\item 'lfc\\_(Intercept)'\n",
       "\\item 'lfc\\_health\\_status2'\n",
       "\\item 'se\\_(Intercept)'\n",
       "\\item 'se\\_health\\_status2'\n",
       "\\item 'W\\_(Intercept)'\n",
       "\\item 'W\\_health\\_status2'\n",
       "\\item 'p\\_(Intercept)'\n",
       "\\item 'p\\_health\\_status2'\n",
       "\\item 'q\\_(Intercept)'\n",
       "\\item 'q\\_health\\_status2'\n",
       "\\item 'diff\\_(Intercept)'\n",
       "\\item 'diff\\_health\\_status2'\n",
       "\\item 'passed\\_ss\\_(Intercept)'\n",
       "\\item 'passed\\_ss\\_health\\_status2'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'taxon'\n",
       "2. 'lfc_(Intercept)'\n",
       "3. 'lfc_health_status2'\n",
       "4. 'se_(Intercept)'\n",
       "5. 'se_health_status2'\n",
       "6. 'W_(Intercept)'\n",
       "7. 'W_health_status2'\n",
       "8. 'p_(Intercept)'\n",
       "9. 'p_health_status2'\n",
       "10. 'q_(Intercept)'\n",
       "11. 'q_health_status2'\n",
       "12. 'diff_(Intercept)'\n",
       "13. 'diff_health_status2'\n",
       "14. 'passed_ss_(Intercept)'\n",
       "15. 'passed_ss_health_status2'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"taxon\"                    \"lfc_(Intercept)\"         \n",
       " [3] \"lfc_health_status2\"       \"se_(Intercept)\"          \n",
       " [5] \"se_health_status2\"        \"W_(Intercept)\"           \n",
       " [7] \"W_health_status2\"         \"p_(Intercept)\"           \n",
       " [9] \"p_health_status2\"         \"q_(Intercept)\"           \n",
       "[11] \"q_health_status2\"         \"diff_(Intercept)\"        \n",
       "[13] \"diff_health_status2\"      \"passed_ss_(Intercept)\"   \n",
       "[15] \"passed_ss_health_status2\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "colnames(output$res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "8c84eb47-fbf9-41db-863e-4c803679f83c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 15</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>taxon</th><th scope=col>lfc_(Intercept)</th><th scope=col>lfc_health_status2</th><th scope=col>se_(Intercept)</th><th scope=col>se_health_status2</th><th scope=col>W_(Intercept)</th><th scope=col>W_health_status2</th><th scope=col>p_(Intercept)</th><th scope=col>p_health_status2</th><th scope=col>q_(Intercept)</th><th scope=col>q_health_status2</th><th scope=col>diff_(Intercept)</th><th scope=col>diff_health_status2</th><th scope=col>passed_ss_(Intercept)</th><th scope=col>passed_ss_health_status2</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>T1</td><td>-0.6815617</td><td>-0.06342587</td><td>0.04311135</td><td>0.07016352</td><td>-15.80933</td><td> -0.9039722</td><td>1.821665e-24</td><td>3.692016e-01</td><td>1.439115e-22</td><td>1.000000e+00</td><td>TRUE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>T2</td><td> 1.0222721</td><td>-1.99577204</td><td>0.05060129</td><td>0.06833953</td><td> 20.20249</td><td>-29.2037721</td><td>2.143589e-35</td><td>5.258069e-48</td><td>1.907794e-33</td><td>4.679681e-46</td><td>TRUE</td><td> TRUE</td><td> TRUE</td><td> TRUE</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 15\n",
       "\\begin{tabular}{r|lllllllllllllll}\n",
       "  & taxon & lfc\\_(Intercept) & lfc\\_health\\_status2 & se\\_(Intercept) & se\\_health\\_status2 & W\\_(Intercept) & W\\_health\\_status2 & p\\_(Intercept) & p\\_health\\_status2 & q\\_(Intercept) & q\\_health\\_status2 & diff\\_(Intercept) & diff\\_health\\_status2 & passed\\_ss\\_(Intercept) & passed\\_ss\\_health\\_status2\\\\\n",
       "  & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <lgl> & <lgl> & <lgl> & <lgl>\\\\\n",
       "\\hline\n",
       "\t1 & T1 & -0.6815617 & -0.06342587 & 0.04311135 & 0.07016352 & -15.80933 &  -0.9039722 & 1.821665e-24 & 3.692016e-01 & 1.439115e-22 & 1.000000e+00 & TRUE & FALSE & FALSE & FALSE\\\\\n",
       "\t2 & T2 &  1.0222721 & -1.99577204 & 0.05060129 & 0.06833953 &  20.20249 & -29.2037721 & 2.143589e-35 & 5.258069e-48 & 1.907794e-33 & 4.679681e-46 & TRUE &  TRUE &  TRUE &  TRUE\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 15\n",
       "\n",
       "| <!--/--> | taxon &lt;chr&gt; | lfc_(Intercept) &lt;dbl&gt; | lfc_health_status2 &lt;dbl&gt; | se_(Intercept) &lt;dbl&gt; | se_health_status2 &lt;dbl&gt; | W_(Intercept) &lt;dbl&gt; | W_health_status2 &lt;dbl&gt; | p_(Intercept) &lt;dbl&gt; | p_health_status2 &lt;dbl&gt; | q_(Intercept) &lt;dbl&gt; | q_health_status2 &lt;dbl&gt; | diff_(Intercept) &lt;lgl&gt; | diff_health_status2 &lt;lgl&gt; | passed_ss_(Intercept) &lt;lgl&gt; | passed_ss_health_status2 &lt;lgl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | T1 | -0.6815617 | -0.06342587 | 0.04311135 | 0.07016352 | -15.80933 |  -0.9039722 | 1.821665e-24 | 3.692016e-01 | 1.439115e-22 | 1.000000e+00 | TRUE | FALSE | FALSE | FALSE |\n",
       "| 2 | T2 |  1.0222721 | -1.99577204 | 0.05060129 | 0.06833953 |  20.20249 | -29.2037721 | 2.143589e-35 | 5.258069e-48 | 1.907794e-33 | 4.679681e-46 | TRUE |  TRUE |  TRUE |  TRUE |\n",
       "\n"
      ],
      "text/plain": [
       "  taxon lfc_(Intercept) lfc_health_status2 se_(Intercept) se_health_status2\n",
       "1 T1    -0.6815617      -0.06342587        0.04311135     0.07016352       \n",
       "2 T2     1.0222721      -1.99577204        0.05060129     0.06833953       \n",
       "  W_(Intercept) W_health_status2 p_(Intercept) p_health_status2 q_(Intercept)\n",
       "1 -15.80933      -0.9039722      1.821665e-24  3.692016e-01     1.439115e-22 \n",
       "2  20.20249     -29.2037721      2.143589e-35  5.258069e-48     1.907794e-33 \n",
       "  q_health_status2 diff_(Intercept) diff_health_status2 passed_ss_(Intercept)\n",
       "1 1.000000e+00     TRUE             FALSE               FALSE                \n",
       "2 4.679681e-46     TRUE              TRUE                TRUE                \n",
       "  passed_ss_health_status2\n",
       "1 FALSE                   \n",
       "2  TRUE                   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(output$res,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "74048420-4a4c-4d5d-a7d3-2b39bde322bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# lfc_cont_cov: Log-fold change for the continuous covariate\n",
    "# lfc_cat_cov2: Log-fold change for the categorical 2 to the reference category 1\n",
    "# se: standard error of log fold change\n",
    "# W: Test statistic for the log-fold change (Wald statistic test ~ t test with no variance assumption)\n",
    "# diff: log fold change is different from zero?\n",
    "# pass: pass the sensitivity analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc6775ba-66ef-4c75-b1f5-90532a5dd89c",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## 3.2 Differential Abundance Analysis: Health status 1, 2 and 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "aa0216c6-ce63-496f-8db0-9f7e4568af66",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 3 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>stand_bmi</th><th scope=col>stand_choles</th><th scope=col>health_status</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>S1</th><td>-7.046117</td><td>-0.05133271</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>S2</th><td>-8.183509</td><td>-1.30178093</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>S3</th><td>-7.369319</td><td>-1.20561056</td><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 3 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & stand\\_bmi & stand\\_choles & health\\_status\\\\\n",
       "  & <dbl> & <dbl> & <fct>\\\\\n",
       "\\hline\n",
       "\tS1 & -7.046117 & -0.05133271 & 1\\\\\n",
       "\tS2 & -8.183509 & -1.30178093 & 1\\\\\n",
       "\tS3 & -7.369319 & -1.20561056 & 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 3 × 3\n",
       "\n",
       "| <!--/--> | stand_bmi &lt;dbl&gt; | stand_choles &lt;dbl&gt; | health_status &lt;fct&gt; |\n",
       "|---|---|---|---|\n",
       "| S1 | -7.046117 | -0.05133271 | 1 |\n",
       "| S2 | -8.183509 | -1.30178093 | 1 |\n",
       "| S3 | -7.369319 | -1.20561056 | 1 |\n",
       "\n"
      ],
      "text/plain": [
       "   stand_bmi stand_choles health_status\n",
       "S1 -7.046117 -0.05133271  1            \n",
       "S2 -8.183509 -1.30178093  1            \n",
       "S3 -7.369319 -1.20561056  1            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>150</li><li>3</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 150\n",
       "\\item 3\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 150\n",
       "2. 3\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 150   3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Meta data\n",
    "head(smd,3)\n",
    "dim(smd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "b31bea3d-ec07-41dc-8524-14c2249e49ff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 2 × 150 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>S1</th><th scope=col>S2</th><th scope=col>S3</th><th scope=col>S4</th><th scope=col>S5</th><th scope=col>S6</th><th scope=col>S7</th><th scope=col>S8</th><th scope=col>S9</th><th scope=col>S10</th><th scope=col>⋯</th><th scope=col>S141</th><th scope=col>S142</th><th scope=col>S143</th><th scope=col>S144</th><th scope=col>S145</th><th scope=col>S146</th><th scope=col>S147</th><th scope=col>S148</th><th scope=col>S149</th><th scope=col>S150</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>T1</th><td>  1</td><td>0</td><td>  1</td><td>  3</td><td>  2</td><td>0</td><td> 1</td><td> 0</td><td> 0</td><td>  3</td><td>⋯</td><td>  142</td><td>  2</td><td>  208</td><td>   232</td><td>  105</td><td>   64</td><td>   317</td><td>   41</td><td>  320</td><td>  14</td></tr>\n",
       "\t<tr><th scope=row>T2</th><td>102</td><td>1</td><td>156</td><td>535</td><td>178</td><td>0</td><td>99</td><td>21</td><td>28</td><td>400</td><td>⋯</td><td>41555</td><td>940</td><td>69158</td><td>150981</td><td>68451</td><td>40298</td><td>258659</td><td>23841</td><td>79188</td><td>7491</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 2 × 150 of type dbl\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & S1 & S2 & S3 & S4 & S5 & S6 & S7 & S8 & S9 & S10 & ⋯ & S141 & S142 & S143 & S144 & S145 & S146 & S147 & S148 & S149 & S150\\\\\n",
       "\\hline\n",
       "\tT1 &   1 & 0 &   1 &   3 &   2 & 0 &  1 &  0 &  0 &   3 & ⋯ &   142 &   2 &   208 &    232 &   105 &    64 &    317 &    41 &   320 &   14\\\\\n",
       "\tT2 & 102 & 1 & 156 & 535 & 178 & 0 & 99 & 21 & 28 & 400 & ⋯ & 41555 & 940 & 69158 & 150981 & 68451 & 40298 & 258659 & 23841 & 79188 & 7491\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 2 × 150 of type dbl\n",
       "\n",
       "| <!--/--> | S1 | S2 | S3 | S4 | S5 | S6 | S7 | S8 | S9 | S10 | ⋯ | S141 | S142 | S143 | S144 | S145 | S146 | S147 | S148 | S149 | S150 |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| T1 |   1 | 0 |   1 |   3 |   2 | 0 |  1 |  0 |  0 |   3 | ⋯ |   142 |   2 |   208 |    232 |   105 |    64 |    317 |    41 |   320 |   14 |\n",
       "| T2 | 102 | 1 | 156 | 535 | 178 | 0 | 99 | 21 | 28 | 400 | ⋯ | 41555 | 940 | 69158 | 150981 | 68451 | 40298 | 258659 | 23841 | 79188 | 7491 |\n",
       "\n"
      ],
      "text/plain": [
       "   S1  S2 S3  S4  S5  S6 S7 S8 S9 S10 ⋯ S141  S142 S143  S144   S145  S146 \n",
       "T1   1 0    1   3   2 0   1  0  0   3 ⋯   142   2    208    232   105    64\n",
       "T2 102 1  156 535 178 0  99 21 28 400 ⋯ 41555 940  69158 150981 68451 40298\n",
       "   S147   S148  S149  S150\n",
       "T1    317    41   320   14\n",
       "T2 258659 23841 79188 7491"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Count data\n",
    "head(otu_data,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "22f7cbc8-7199-43ce-82d6-e9716057c99f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the tse object\n",
    "assays = S4Vectors::SimpleList(counts = otu_data) # count data\n",
    "smd_3H = S4Vectors::DataFrame(smd) # meta data\n",
    "tse = TreeSummarizedExperiment::TreeSummarizedExperiment(assays = assays, colData = smd_3H)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "960588af-56bc-48e0-8d3f-d5b946818816",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Obtaining initial estimates ...\n",
      "\n",
      "Estimating sample-specific biases ...\n",
      "\n",
      "Sensitivity analysis for pseudo-count addition to 0s: ...\n",
      "\n",
      "ANCOM-BC2 primary results ...\n",
      "\n",
      "ANCOM-BC2 multiple pairwise comparisons ...\n",
      "\n"
     ]
    }
   ],
   "source": [
    "set.seed(123)\n",
    "output = ancombc2(data = tse, \n",
    "                  assay_name = \"counts\",\n",
    "                  fix_formula = \"health_status\", # formular for linear regression model: response ~ covariate1 + covariate2\n",
    "                  p_adj_method = \"holm\",\n",
    "                  pseudo_sens = TRUE,                 # sensitivity analysis to the pseudo-count addition\n",
    "                  prv_cut = 0.10,                     # (0,1) Taxa with prevalences less than prv_cut will be excluded\n",
    "                  lib_cut = 1000,                     # filter sampling with sampling depth\n",
    "                  s0_perc = 0.05,                     # advoid significant feature due to small standard error, (rare taxa)\n",
    "                  group = \"health_status\",            # Group to be differencially analyzed (healthy status - normal,A,B             \n",
    "                  alpha = 0.05,                       # Level of significant\n",
    "                  n_cl = 2,                           # CUP cores to compute\n",
    "                  verbose = TRUE,\n",
    "                  pairwise = TRUE,                   \n",
    "                  iter_control = list(tol = 1e-5, max_iter = 20, # Liner regression model parameter\n",
    "                                      verbose = FALSE),\n",
    "                  em_control = list(tol = 1e-5, max_iter = 100),\n",
    "                  lme_control = NULL, \n",
    "                  mdfdr_control = list(fwer_ctrl_method = \"holm\", B = 100))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "cbd1df9c-6c11-483c-9976-b63726724aca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 4 × 22</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>taxon</th><th scope=col>lfc_health_status2</th><th scope=col>lfc_health_status3</th><th scope=col>lfc_health_status3_health_status2</th><th scope=col>se_health_status2</th><th scope=col>se_health_status3</th><th scope=col>se_health_status3_health_status2</th><th scope=col>W_health_status2</th><th scope=col>W_health_status3</th><th scope=col>W_health_status3_health_status2</th><th scope=col>⋯</th><th scope=col>p_health_status3_health_status2</th><th scope=col>q_health_status2</th><th scope=col>q_health_status3</th><th scope=col>q_health_status3_health_status2</th><th scope=col>diff_health_status2</th><th scope=col>diff_health_status3</th><th scope=col>diff_health_status3_health_status2</th><th scope=col>passed_ss_health_status2</th><th scope=col>passed_ss_health_status3</th><th scope=col>passed_ss_health_status3_health_status2</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>T1</td><td>-0.06827952</td><td>-0.06796177</td><td> 0.0003177525</td><td>0.09406369</td><td>0.09558814</td><td>0.08748904</td><td> -0.7258861</td><td>-0.7109854</td><td> 0.003631912</td><td>⋯</td><td>1.000000e+00</td><td>1.000000e+00</td><td>1.000000e+00</td><td>1.000000e+00</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td> TRUE</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>T2</td><td>-2.00062569</td><td> 1.00584236</td><td> 3.0064680500</td><td>0.06866012</td><td>0.06751198</td><td>0.08178809</td><td>-29.1381035</td><td>14.8987239</td><td>36.759241928</td><td>⋯</td><td>8.507227e-74</td><td>2.760942e-60</td><td>1.061862e-29</td><td>9.676971e-73</td><td> TRUE</td><td> TRUE</td><td> TRUE</td><td> TRUE</td><td> TRUE</td><td> TRUE</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>T3</td><td>-0.75218203</td><td>-0.71827224</td><td> 0.0339097941</td><td>0.13448308</td><td>0.13622569</td><td>0.09292088</td><td> -5.5931349</td><td>-5.2726637</td><td> 0.364931927</td><td>⋯</td><td>7.164041e-01</td><td>6.093504e-06</td><td>1.874949e-05</td><td>1.000000e+00</td><td> TRUE</td><td> TRUE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>T4</td><td> 0.02750635</td><td> 0.02336361</td><td>-0.0041427448</td><td>0.06301810</td><td>0.06939446</td><td>0.07585942</td><td>  0.4364834</td><td> 0.3366783</td><td>-0.054610814</td><td>⋯</td><td>1.000000e+00</td><td>1.000000e+00</td><td>1.000000e+00</td><td>1.000000e+00</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td> TRUE</td><td> TRUE</td><td> TRUE</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 4 × 22\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & taxon & lfc\\_health\\_status2 & lfc\\_health\\_status3 & lfc\\_health\\_status3\\_health\\_status2 & se\\_health\\_status2 & se\\_health\\_status3 & se\\_health\\_status3\\_health\\_status2 & W\\_health\\_status2 & W\\_health\\_status3 & W\\_health\\_status3\\_health\\_status2 & ⋯ & p\\_health\\_status3\\_health\\_status2 & q\\_health\\_status2 & q\\_health\\_status3 & q\\_health\\_status3\\_health\\_status2 & diff\\_health\\_status2 & diff\\_health\\_status3 & diff\\_health\\_status3\\_health\\_status2 & passed\\_ss\\_health\\_status2 & passed\\_ss\\_health\\_status3 & passed\\_ss\\_health\\_status3\\_health\\_status2\\\\\n",
       "  & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl>\\\\\n",
       "\\hline\n",
       "\t1 & T1 & -0.06827952 & -0.06796177 &  0.0003177525 & 0.09406369 & 0.09558814 & 0.08748904 &  -0.7258861 & -0.7109854 &  0.003631912 & ⋯ & 1.000000e+00 & 1.000000e+00 & 1.000000e+00 & 1.000000e+00 & FALSE & FALSE & FALSE & FALSE & FALSE &  TRUE\\\\\n",
       "\t2 & T2 & -2.00062569 &  1.00584236 &  3.0064680500 & 0.06866012 & 0.06751198 & 0.08178809 & -29.1381035 & 14.8987239 & 36.759241928 & ⋯ & 8.507227e-74 & 2.760942e-60 & 1.061862e-29 & 9.676971e-73 &  TRUE &  TRUE &  TRUE &  TRUE &  TRUE &  TRUE\\\\\n",
       "\t3 & T3 & -0.75218203 & -0.71827224 &  0.0339097941 & 0.13448308 & 0.13622569 & 0.09292088 &  -5.5931349 & -5.2726637 &  0.364931927 & ⋯ & 7.164041e-01 & 6.093504e-06 & 1.874949e-05 & 1.000000e+00 &  TRUE &  TRUE & FALSE & FALSE & FALSE & FALSE\\\\\n",
       "\t4 & T4 &  0.02750635 &  0.02336361 & -0.0041427448 & 0.06301810 & 0.06939446 & 0.07585942 &   0.4364834 &  0.3366783 & -0.054610814 & ⋯ & 1.000000e+00 & 1.000000e+00 & 1.000000e+00 & 1.000000e+00 & FALSE & FALSE & FALSE &  TRUE &  TRUE &  TRUE\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 4 × 22\n",
       "\n",
       "| <!--/--> | taxon &lt;chr&gt; | lfc_health_status2 &lt;dbl&gt; | lfc_health_status3 &lt;dbl&gt; | lfc_health_status3_health_status2 &lt;dbl&gt; | se_health_status2 &lt;dbl&gt; | se_health_status3 &lt;dbl&gt; | se_health_status3_health_status2 &lt;dbl&gt; | W_health_status2 &lt;dbl&gt; | W_health_status3 &lt;dbl&gt; | W_health_status3_health_status2 &lt;dbl&gt; | ⋯ ⋯ | p_health_status3_health_status2 &lt;dbl&gt; | q_health_status2 &lt;dbl&gt; | q_health_status3 &lt;dbl&gt; | q_health_status3_health_status2 &lt;dbl&gt; | diff_health_status2 &lt;lgl&gt; | diff_health_status3 &lt;lgl&gt; | diff_health_status3_health_status2 &lt;lgl&gt; | passed_ss_health_status2 &lt;lgl&gt; | passed_ss_health_status3 &lt;lgl&gt; | passed_ss_health_status3_health_status2 &lt;lgl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | T1 | -0.06827952 | -0.06796177 |  0.0003177525 | 0.09406369 | 0.09558814 | 0.08748904 |  -0.7258861 | -0.7109854 |  0.003631912 | ⋯ | 1.000000e+00 | 1.000000e+00 | 1.000000e+00 | 1.000000e+00 | FALSE | FALSE | FALSE | FALSE | FALSE |  TRUE |\n",
       "| 2 | T2 | -2.00062569 |  1.00584236 |  3.0064680500 | 0.06866012 | 0.06751198 | 0.08178809 | -29.1381035 | 14.8987239 | 36.759241928 | ⋯ | 8.507227e-74 | 2.760942e-60 | 1.061862e-29 | 9.676971e-73 |  TRUE |  TRUE |  TRUE |  TRUE |  TRUE |  TRUE |\n",
       "| 3 | T3 | -0.75218203 | -0.71827224 |  0.0339097941 | 0.13448308 | 0.13622569 | 0.09292088 |  -5.5931349 | -5.2726637 |  0.364931927 | ⋯ | 7.164041e-01 | 6.093504e-06 | 1.874949e-05 | 1.000000e+00 |  TRUE |  TRUE | FALSE | FALSE | FALSE | FALSE |\n",
       "| 4 | T4 |  0.02750635 |  0.02336361 | -0.0041427448 | 0.06301810 | 0.06939446 | 0.07585942 |   0.4364834 |  0.3366783 | -0.054610814 | ⋯ | 1.000000e+00 | 1.000000e+00 | 1.000000e+00 | 1.000000e+00 | FALSE | FALSE | FALSE |  TRUE |  TRUE |  TRUE |\n",
       "\n"
      ],
      "text/plain": [
       "  taxon lfc_health_status2 lfc_health_status3 lfc_health_status3_health_status2\n",
       "1 T1    -0.06827952        -0.06796177         0.0003177525                    \n",
       "2 T2    -2.00062569         1.00584236         3.0064680500                    \n",
       "3 T3    -0.75218203        -0.71827224         0.0339097941                    \n",
       "4 T4     0.02750635         0.02336361        -0.0041427448                    \n",
       "  se_health_status2 se_health_status3 se_health_status3_health_status2\n",
       "1 0.09406369        0.09558814        0.08748904                      \n",
       "2 0.06866012        0.06751198        0.08178809                      \n",
       "3 0.13448308        0.13622569        0.09292088                      \n",
       "4 0.06301810        0.06939446        0.07585942                      \n",
       "  W_health_status2 W_health_status3 W_health_status3_health_status2 ⋯\n",
       "1  -0.7258861      -0.7109854        0.003631912                    ⋯\n",
       "2 -29.1381035      14.8987239       36.759241928                    ⋯\n",
       "3  -5.5931349      -5.2726637        0.364931927                    ⋯\n",
       "4   0.4364834       0.3366783       -0.054610814                    ⋯\n",
       "  p_health_status3_health_status2 q_health_status2 q_health_status3\n",
       "1 1.000000e+00                    1.000000e+00     1.000000e+00    \n",
       "2 8.507227e-74                    2.760942e-60     1.061862e-29    \n",
       "3 7.164041e-01                    6.093504e-06     1.874949e-05    \n",
       "4 1.000000e+00                    1.000000e+00     1.000000e+00    \n",
       "  q_health_status3_health_status2 diff_health_status2 diff_health_status3\n",
       "1 1.000000e+00                    FALSE               FALSE              \n",
       "2 9.676971e-73                     TRUE                TRUE              \n",
       "3 1.000000e+00                     TRUE                TRUE              \n",
       "4 1.000000e+00                    FALSE               FALSE              \n",
       "  diff_health_status3_health_status2 passed_ss_health_status2\n",
       "1 FALSE                              FALSE                   \n",
       "2  TRUE                               TRUE                   \n",
       "3 FALSE                              FALSE                   \n",
       "4 FALSE                               TRUE                   \n",
       "  passed_ss_health_status3 passed_ss_health_status3_health_status2\n",
       "1 FALSE                     TRUE                                  \n",
       "2  TRUE                     TRUE                                  \n",
       "3 FALSE                    FALSE                                  \n",
       "4  TRUE                     TRUE                                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(output$res_pair,4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "6c50cca8-85e9-464e-b6c3-ec071890ed13",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "0.000317749999999992"
      ],
      "text/latex": [
       "0.000317749999999992"
      ],
      "text/markdown": [
       "0.000317749999999992"
      ],
      "text/plain": [
       "[1] 0.00031775"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "0.06827952 - 0.06796177"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "56b6cc30-42d1-414d-87f1-73b320149def",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "3.00646805"
      ],
      "text/latex": [
       "3.00646805"
      ],
      "text/markdown": [
       "3.00646805"
      ],
      "text/plain": [
       "[1] 3.006468"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "2.00062569 + 1.00584236"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d780c53-23e9-4a53-97f9-9e4c48829be7",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## 3.3 Differential Abundance Analysis: Continuous covariate (choles level)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0946da0d-fcbd-41ce-84c8-409dcf096a2c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 3 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>stand_bmi</th><th scope=col>stand_choles</th><th scope=col>health_status</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>S1</th><td>-7.046117</td><td>-0.05133271</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>S2</th><td>-8.183509</td><td>-1.30178093</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>S3</th><td>-7.369319</td><td>-1.20561056</td><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 3 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & stand\\_bmi & stand\\_choles & health\\_status\\\\\n",
       "  & <dbl> & <dbl> & <fct>\\\\\n",
       "\\hline\n",
       "\tS1 & -7.046117 & -0.05133271 & 1\\\\\n",
       "\tS2 & -8.183509 & -1.30178093 & 1\\\\\n",
       "\tS3 & -7.369319 & -1.20561056 & 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 3 × 3\n",
       "\n",
       "| <!--/--> | stand_bmi &lt;dbl&gt; | stand_choles &lt;dbl&gt; | health_status &lt;fct&gt; |\n",
       "|---|---|---|---|\n",
       "| S1 | -7.046117 | -0.05133271 | 1 |\n",
       "| S2 | -8.183509 | -1.30178093 | 1 |\n",
       "| S3 | -7.369319 | -1.20561056 | 1 |\n",
       "\n"
      ],
      "text/plain": [
       "   stand_bmi stand_choles health_status\n",
       "S1 -7.046117 -0.05133271  1            \n",
       "S2 -8.183509 -1.30178093  1            \n",
       "S3 -7.369319 -1.20561056  1            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 2 × 150 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>S1</th><th scope=col>S2</th><th scope=col>S3</th><th scope=col>S4</th><th scope=col>S5</th><th scope=col>S6</th><th scope=col>S7</th><th scope=col>S8</th><th scope=col>S9</th><th scope=col>S10</th><th scope=col>⋯</th><th scope=col>S141</th><th scope=col>S142</th><th scope=col>S143</th><th scope=col>S144</th><th scope=col>S145</th><th scope=col>S146</th><th scope=col>S147</th><th scope=col>S148</th><th scope=col>S149</th><th scope=col>S150</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>T1</th><td>  1</td><td>0</td><td>  1</td><td>  3</td><td>  2</td><td>0</td><td> 1</td><td> 0</td><td> 0</td><td>  3</td><td>⋯</td><td>  142</td><td>  2</td><td>  208</td><td>   232</td><td>  105</td><td>   64</td><td>   317</td><td>   41</td><td>  320</td><td>  14</td></tr>\n",
       "\t<tr><th scope=row>T2</th><td>102</td><td>1</td><td>156</td><td>535</td><td>178</td><td>0</td><td>99</td><td>21</td><td>28</td><td>400</td><td>⋯</td><td>41555</td><td>940</td><td>69158</td><td>150981</td><td>68451</td><td>40298</td><td>258659</td><td>23841</td><td>79188</td><td>7491</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 2 × 150 of type dbl\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & S1 & S2 & S3 & S4 & S5 & S6 & S7 & S8 & S9 & S10 & ⋯ & S141 & S142 & S143 & S144 & S145 & S146 & S147 & S148 & S149 & S150\\\\\n",
       "\\hline\n",
       "\tT1 &   1 & 0 &   1 &   3 &   2 & 0 &  1 &  0 &  0 &   3 & ⋯ &   142 &   2 &   208 &    232 &   105 &    64 &    317 &    41 &   320 &   14\\\\\n",
       "\tT2 & 102 & 1 & 156 & 535 & 178 & 0 & 99 & 21 & 28 & 400 & ⋯ & 41555 & 940 & 69158 & 150981 & 68451 & 40298 & 258659 & 23841 & 79188 & 7491\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 2 × 150 of type dbl\n",
       "\n",
       "| <!--/--> | S1 | S2 | S3 | S4 | S5 | S6 | S7 | S8 | S9 | S10 | ⋯ | S141 | S142 | S143 | S144 | S145 | S146 | S147 | S148 | S149 | S150 |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| T1 |   1 | 0 |   1 |   3 |   2 | 0 |  1 |  0 |  0 |   3 | ⋯ |   142 |   2 |   208 |    232 |   105 |    64 |    317 |    41 |   320 |   14 |\n",
       "| T2 | 102 | 1 | 156 | 535 | 178 | 0 | 99 | 21 | 28 | 400 | ⋯ | 41555 | 940 | 69158 | 150981 | 68451 | 40298 | 258659 | 23841 | 79188 | 7491 |\n",
       "\n"
      ],
      "text/plain": [
       "   S1  S2 S3  S4  S5  S6 S7 S8 S9 S10 ⋯ S141  S142 S143  S144   S145  S146 \n",
       "T1   1 0    1   3   2 0   1  0  0   3 ⋯   142   2    208    232   105    64\n",
       "T2 102 1  156 535 178 0  99 21 28 400 ⋯ 41555 940  69158 150981 68451 40298\n",
       "   S147   S148  S149  S150\n",
       "T1    317    41   320   14\n",
       "T2 258659 23841 79188 7491"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Meta data\n",
    "head(smd,3)\n",
    "# Count data\n",
    "head(otu_data,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "19eddc58-a9de-44c6-b093-6547e64751d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the tse object\n",
    "assays = S4Vectors::SimpleList(counts = otu_data) # count data\n",
    "smd_con = S4Vectors::DataFrame(smd) # meta data\n",
    "tse = TreeSummarizedExperiment::TreeSummarizedExperiment(assays = assays, colData = smd_con)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "95a39cb1-b505-4a3a-a2c8-ff60e37fa122",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Obtaining initial estimates ...\n",
      "\n",
      "Estimating sample-specific biases ...\n",
      "\n",
      "Sensitivity analysis for pseudo-count addition to 0s: ...\n",
      "\n",
      "ANCOM-BC2 primary results ...\n",
      "\n"
     ]
    }
   ],
   "source": [
    "set.seed(123)\n",
    "output = ancombc2(data = tse, \n",
    "                  assay_name = \"counts\",\n",
    "                  fix_formula = \"stand_choles\", # formular for linear regression model: response ~ covariate1 + covariate2\n",
    "                  p_adj_method = \"holm\",\n",
    "                  pseudo_sens = TRUE,                 # sensitivity analysis to the pseudo-count addition\n",
    "                  prv_cut = 0.10,                     # (0,1) Taxa with prevalences less than prv_cut will be excluded\n",
    "                  lib_cut = 1000,                     # filter sampling with sampling depth\n",
    "                  s0_perc = 0.05,                     # advoid significant feature due to small standard error, (rare taxa)\n",
    "              #    group = \"health_status\",            # Group to be differencially analyzed (healthy status - normal,A,B             \n",
    "                  alpha = 0.05,                       # Level of significant\n",
    "                  n_cl = 2,                           # CUP cores to compute\n",
    "                  verbose = TRUE,\n",
    "             #     pairwise = TRUE,                  \n",
    "                  iter_control = list(tol = 1e-5, max_iter = 20, # Liner regression model parameter\n",
    "                                      verbose = FALSE),\n",
    "                  em_control = list(tol = 1e-5, max_iter = 100),\n",
    "                  lme_control = NULL, \n",
    "                  mdfdr_control = list(fwer_ctrl_method = \"holm\", B = 100))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0d2ee97b-bcee-4006-973c-79b69529aa42",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 4 × 15</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>taxon</th><th scope=col>lfc_(Intercept)</th><th scope=col>lfc_stand_choles</th><th scope=col>se_(Intercept)</th><th scope=col>se_stand_choles</th><th scope=col>W_(Intercept)</th><th scope=col>W_stand_choles</th><th scope=col>p_(Intercept)</th><th scope=col>p_stand_choles</th><th scope=col>q_(Intercept)</th><th scope=col>q_stand_choles</th><th scope=col>diff_(Intercept)</th><th scope=col>diff_stand_choles</th><th scope=col>passed_ss_(Intercept)</th><th scope=col>passed_ss_stand_choles</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>T1</td><td>-0.67994988</td><td> 0.030468553</td><td>0.02437539</td><td>0.02324272</td><td>-27.8949303</td><td> 1.3108861</td><td>2.856327e-53</td><td>0.19248550</td><td>2.285061e-51</td><td>1</td><td> TRUE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>T2</td><td> 0.01046647</td><td>-0.159244130</td><td>0.09903658</td><td>0.09593896</td><td>  0.1056829</td><td>-1.6598484</td><td>9.159841e-01</td><td>0.09916742</td><td>1.000000e+00</td><td>1</td><td>FALSE</td><td>FALSE</td><td> TRUE</td><td> TRUE</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>T3</td><td>-2.13141147</td><td>-0.007092409</td><td>0.02783238</td><td>0.02778904</td><td>-76.5802796</td><td>-0.2552232</td><td>6.862424e-64</td><td>0.79938209</td><td>6.083494e-62</td><td>1</td><td> TRUE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>T4</td><td>-0.02993724</td><td> 0.005782460</td><td>0.01861777</td><td>0.02042012</td><td> -1.6079927</td><td> 0.2831747</td><td>1.100897e-01</td><td>0.77746125</td><td>1.000000e+00</td><td>1</td><td>FALSE</td><td>FALSE</td><td> TRUE</td><td> TRUE</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 4 × 15\n",
       "\\begin{tabular}{r|lllllllllllllll}\n",
       "  & taxon & lfc\\_(Intercept) & lfc\\_stand\\_choles & se\\_(Intercept) & se\\_stand\\_choles & W\\_(Intercept) & W\\_stand\\_choles & p\\_(Intercept) & p\\_stand\\_choles & q\\_(Intercept) & q\\_stand\\_choles & diff\\_(Intercept) & diff\\_stand\\_choles & passed\\_ss\\_(Intercept) & passed\\_ss\\_stand\\_choles\\\\\n",
       "  & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <lgl> & <lgl> & <lgl> & <lgl>\\\\\n",
       "\\hline\n",
       "\t1 & T1 & -0.67994988 &  0.030468553 & 0.02437539 & 0.02324272 & -27.8949303 &  1.3108861 & 2.856327e-53 & 0.19248550 & 2.285061e-51 & 1 &  TRUE & FALSE & FALSE & FALSE\\\\\n",
       "\t2 & T2 &  0.01046647 & -0.159244130 & 0.09903658 & 0.09593896 &   0.1056829 & -1.6598484 & 9.159841e-01 & 0.09916742 & 1.000000e+00 & 1 & FALSE & FALSE &  TRUE &  TRUE\\\\\n",
       "\t3 & T3 & -2.13141147 & -0.007092409 & 0.02783238 & 0.02778904 & -76.5802796 & -0.2552232 & 6.862424e-64 & 0.79938209 & 6.083494e-62 & 1 &  TRUE & FALSE & FALSE & FALSE\\\\\n",
       "\t4 & T4 & -0.02993724 &  0.005782460 & 0.01861777 & 0.02042012 &  -1.6079927 &  0.2831747 & 1.100897e-01 & 0.77746125 & 1.000000e+00 & 1 & FALSE & FALSE &  TRUE &  TRUE\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 4 × 15\n",
       "\n",
       "| <!--/--> | taxon &lt;chr&gt; | lfc_(Intercept) &lt;dbl&gt; | lfc_stand_choles &lt;dbl&gt; | se_(Intercept) &lt;dbl&gt; | se_stand_choles &lt;dbl&gt; | W_(Intercept) &lt;dbl&gt; | W_stand_choles &lt;dbl&gt; | p_(Intercept) &lt;dbl&gt; | p_stand_choles &lt;dbl&gt; | q_(Intercept) &lt;dbl&gt; | q_stand_choles &lt;dbl&gt; | diff_(Intercept) &lt;lgl&gt; | diff_stand_choles &lt;lgl&gt; | passed_ss_(Intercept) &lt;lgl&gt; | passed_ss_stand_choles &lt;lgl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | T1 | -0.67994988 |  0.030468553 | 0.02437539 | 0.02324272 | -27.8949303 |  1.3108861 | 2.856327e-53 | 0.19248550 | 2.285061e-51 | 1 |  TRUE | FALSE | FALSE | FALSE |\n",
       "| 2 | T2 |  0.01046647 | -0.159244130 | 0.09903658 | 0.09593896 |   0.1056829 | -1.6598484 | 9.159841e-01 | 0.09916742 | 1.000000e+00 | 1 | FALSE | FALSE |  TRUE |  TRUE |\n",
       "| 3 | T3 | -2.13141147 | -0.007092409 | 0.02783238 | 0.02778904 | -76.5802796 | -0.2552232 | 6.862424e-64 | 0.79938209 | 6.083494e-62 | 1 |  TRUE | FALSE | FALSE | FALSE |\n",
       "| 4 | T4 | -0.02993724 |  0.005782460 | 0.01861777 | 0.02042012 |  -1.6079927 |  0.2831747 | 1.100897e-01 | 0.77746125 | 1.000000e+00 | 1 | FALSE | FALSE |  TRUE |  TRUE |\n",
       "\n"
      ],
      "text/plain": [
       "  taxon lfc_(Intercept) lfc_stand_choles se_(Intercept) se_stand_choles\n",
       "1 T1    -0.67994988      0.030468553     0.02437539     0.02324272     \n",
       "2 T2     0.01046647     -0.159244130     0.09903658     0.09593896     \n",
       "3 T3    -2.13141147     -0.007092409     0.02783238     0.02778904     \n",
       "4 T4    -0.02993724      0.005782460     0.01861777     0.02042012     \n",
       "  W_(Intercept) W_stand_choles p_(Intercept) p_stand_choles q_(Intercept)\n",
       "1 -27.8949303    1.3108861     2.856327e-53  0.19248550     2.285061e-51 \n",
       "2   0.1056829   -1.6598484     9.159841e-01  0.09916742     1.000000e+00 \n",
       "3 -76.5802796   -0.2552232     6.862424e-64  0.79938209     6.083494e-62 \n",
       "4  -1.6079927    0.2831747     1.100897e-01  0.77746125     1.000000e+00 \n",
       "  q_stand_choles diff_(Intercept) diff_stand_choles passed_ss_(Intercept)\n",
       "1 1               TRUE            FALSE             FALSE                \n",
       "2 1              FALSE            FALSE              TRUE                \n",
       "3 1               TRUE            FALSE             FALSE                \n",
       "4 1              FALSE            FALSE              TRUE                \n",
       "  passed_ss_stand_choles\n",
       "1 FALSE                 \n",
       "2  TRUE                 \n",
       "3 FALSE                 \n",
       "4  TRUE                 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(output$res, 4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff1201ac-6801-47c7-a597-ff95cab5a4a6",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## 3.4 Differential Abundance Analysis: Continuous covariate (choles level) + Categorical variable (Health status 1,2,3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "fe7c3891-9476-496a-bfe4-865a0d089ad2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 3 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>stand_bmi</th><th scope=col>stand_choles</th><th scope=col>health_status</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>S1</th><td>-7.046117</td><td>-0.05133271</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>S2</th><td>-8.183509</td><td>-1.30178093</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>S3</th><td>-7.369319</td><td>-1.20561056</td><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 3 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & stand\\_bmi & stand\\_choles & health\\_status\\\\\n",
       "  & <dbl> & <dbl> & <fct>\\\\\n",
       "\\hline\n",
       "\tS1 & -7.046117 & -0.05133271 & 1\\\\\n",
       "\tS2 & -8.183509 & -1.30178093 & 1\\\\\n",
       "\tS3 & -7.369319 & -1.20561056 & 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 3 × 3\n",
       "\n",
       "| <!--/--> | stand_bmi &lt;dbl&gt; | stand_choles &lt;dbl&gt; | health_status &lt;fct&gt; |\n",
       "|---|---|---|---|\n",
       "| S1 | -7.046117 | -0.05133271 | 1 |\n",
       "| S2 | -8.183509 | -1.30178093 | 1 |\n",
       "| S3 | -7.369319 | -1.20561056 | 1 |\n",
       "\n"
      ],
      "text/plain": [
       "   stand_bmi stand_choles health_status\n",
       "S1 -7.046117 -0.05133271  1            \n",
       "S2 -8.183509 -1.30178093  1            \n",
       "S3 -7.369319 -1.20561056  1            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 2 × 150 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>S1</th><th scope=col>S2</th><th scope=col>S3</th><th scope=col>S4</th><th scope=col>S5</th><th scope=col>S6</th><th scope=col>S7</th><th scope=col>S8</th><th scope=col>S9</th><th scope=col>S10</th><th scope=col>⋯</th><th scope=col>S141</th><th scope=col>S142</th><th scope=col>S143</th><th scope=col>S144</th><th scope=col>S145</th><th scope=col>S146</th><th scope=col>S147</th><th scope=col>S148</th><th scope=col>S149</th><th scope=col>S150</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>T1</th><td>  1</td><td>0</td><td>  1</td><td>  3</td><td>  2</td><td>0</td><td> 1</td><td> 0</td><td> 0</td><td>  3</td><td>⋯</td><td>  142</td><td>  2</td><td>  208</td><td>   232</td><td>  105</td><td>   64</td><td>   317</td><td>   41</td><td>  320</td><td>  14</td></tr>\n",
       "\t<tr><th scope=row>T2</th><td>102</td><td>1</td><td>156</td><td>535</td><td>178</td><td>0</td><td>99</td><td>21</td><td>28</td><td>400</td><td>⋯</td><td>41555</td><td>940</td><td>69158</td><td>150981</td><td>68451</td><td>40298</td><td>258659</td><td>23841</td><td>79188</td><td>7491</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 2 × 150 of type dbl\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & S1 & S2 & S3 & S4 & S5 & S6 & S7 & S8 & S9 & S10 & ⋯ & S141 & S142 & S143 & S144 & S145 & S146 & S147 & S148 & S149 & S150\\\\\n",
       "\\hline\n",
       "\tT1 &   1 & 0 &   1 &   3 &   2 & 0 &  1 &  0 &  0 &   3 & ⋯ &   142 &   2 &   208 &    232 &   105 &    64 &    317 &    41 &   320 &   14\\\\\n",
       "\tT2 & 102 & 1 & 156 & 535 & 178 & 0 & 99 & 21 & 28 & 400 & ⋯ & 41555 & 940 & 69158 & 150981 & 68451 & 40298 & 258659 & 23841 & 79188 & 7491\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 2 × 150 of type dbl\n",
       "\n",
       "| <!--/--> | S1 | S2 | S3 | S4 | S5 | S6 | S7 | S8 | S9 | S10 | ⋯ | S141 | S142 | S143 | S144 | S145 | S146 | S147 | S148 | S149 | S150 |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| T1 |   1 | 0 |   1 |   3 |   2 | 0 |  1 |  0 |  0 |   3 | ⋯ |   142 |   2 |   208 |    232 |   105 |    64 |    317 |    41 |   320 |   14 |\n",
       "| T2 | 102 | 1 | 156 | 535 | 178 | 0 | 99 | 21 | 28 | 400 | ⋯ | 41555 | 940 | 69158 | 150981 | 68451 | 40298 | 258659 | 23841 | 79188 | 7491 |\n",
       "\n"
      ],
      "text/plain": [
       "   S1  S2 S3  S4  S5  S6 S7 S8 S9 S10 ⋯ S141  S142 S143  S144   S145  S146 \n",
       "T1   1 0    1   3   2 0   1  0  0   3 ⋯   142   2    208    232   105    64\n",
       "T2 102 1  156 535 178 0  99 21 28 400 ⋯ 41555 940  69158 150981 68451 40298\n",
       "   S147   S148  S149  S150\n",
       "T1    317    41   320   14\n",
       "T2 258659 23841 79188 7491"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Meta data\n",
    "head(smd,3)\n",
    "# Count data\n",
    "head(otu_data,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61d3062e-1592-4783-bbe3-65906808b92c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the tse object\n",
    "assays = S4Vectors::SimpleList(counts = otu_data) # count data\n",
    "smd_3H = S4Vectors::DataFrame(smd) # meta data\n",
    "tse = TreeSummarizedExperiment::TreeSummarizedExperiment(assays = assays, colData = smd_3H)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1b044956-a747-4866-9b46-bdd0caa4aadc",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Obtaining initial estimates ...\n",
      "\n",
      "Estimating sample-specific biases ...\n",
      "\n",
      "Sensitivity analysis for pseudo-count addition to 0s: ...\n",
      "\n",
      "ANCOM-BC2 primary results ...\n",
      "\n",
      "ANCOM-BC2 multiple pairwise comparisons ...\n",
      "\n"
     ]
    }
   ],
   "source": [
    "set.seed(123)\n",
    "output = ancombc2(data = tse, \n",
    "                  assay_name = \"counts\",\n",
    "                  fix_formula = \"stand_choles + health_status\", # formular for linear regression model: response ~ covariate1 + covariate2\n",
    "                  p_adj_method = \"holm\",\n",
    "                  pseudo_sens = TRUE,                 # sensitivity analysis to the pseudo-count addition\n",
    "                  prv_cut = 0.10,                     # (0,1) Taxa with prevalences less than prv_cut will be excluded\n",
    "                  lib_cut = 1000,                     # filter sampling with sampling depth\n",
    "                  s0_perc = 0.05,                     # advoid significant feature due to small standard error, (rare taxa)\n",
    "                  group = \"health_status\",            # Group to be differencially analyzed (healthy status - normal,A,B             \n",
    "                  alpha = 0.05,                       # Level of significant\n",
    "                  n_cl = 2,                           # CUP cores to compute\n",
    "                  verbose = TRUE,\n",
    "                  pairwise = TRUE,                   # CUP cores to compute\n",
    "                  iter_control = list(tol = 1e-5, max_iter = 20, # Liner regression model parameter\n",
    "                                      verbose = FALSE),\n",
    "                  em_control = list(tol = 1e-5, max_iter = 100),\n",
    "                  lme_control = NULL, \n",
    "                  mdfdr_control = list(fwer_ctrl_method = \"holm\", B = 100))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c99fc858-3a06-4788-a9d7-e2c9f54c7196",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 4 × 29</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>taxon</th><th scope=col>lfc_(Intercept)</th><th scope=col>lfc_stand_choles</th><th scope=col>lfc_health_status2</th><th scope=col>lfc_health_status3</th><th scope=col>se_(Intercept)</th><th scope=col>se_stand_choles</th><th scope=col>se_health_status2</th><th scope=col>se_health_status3</th><th scope=col>W_(Intercept)</th><th scope=col>⋯</th><th scope=col>q_health_status2</th><th scope=col>q_health_status3</th><th scope=col>diff_(Intercept)</th><th scope=col>diff_stand_choles</th><th scope=col>diff_health_status2</th><th scope=col>diff_health_status3</th><th scope=col>passed_ss_(Intercept)</th><th scope=col>passed_ss_stand_choles</th><th scope=col>passed_ss_health_status2</th><th scope=col>passed_ss_health_status3</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>T1</td><td>-0.6597409</td><td> 0.030997815</td><td>-0.03692064</td><td>-0.02252950</td><td>0.05046109</td><td>0.02209994</td><td>0.08283890</td><td>0.08287436</td><td>-13.074251</td><td>⋯</td><td>1.000000e+00</td><td>1.000000e+00</td><td> TRUE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>T2</td><td> 0.3163284</td><td>-0.015585963</td><td>-2.00692293</td><td> 0.98887925</td><td>0.02811085</td><td>0.01591812</td><td>0.03515360</td><td>0.03680906</td><td> 11.252893</td><td>⋯</td><td>2.018298e-96</td><td>5.928845e-55</td><td> TRUE</td><td>FALSE</td><td> TRUE</td><td> TRUE</td><td> TRUE</td><td> TRUE</td><td> TRUE</td><td> TRUE</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>T3</td><td>-1.2393678</td><td>-0.028958697</td><td>-0.92761498</td><td>-0.89656130</td><td>0.06904852</td><td>0.02909475</td><td>0.12633810</td><td>0.13062341</td><td>-17.949231</td><td>⋯</td><td>5.015500e-08</td><td>3.335914e-07</td><td> TRUE</td><td>FALSE</td><td> TRUE</td><td> TRUE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>T4</td><td>-0.0420827</td><td> 0.009269526</td><td> 0.01861696</td><td> 0.01487885</td><td>0.03146651</td><td>0.01529956</td><td>0.03787916</td><td>0.03601447</td><td> -1.337381</td><td>⋯</td><td>1.000000e+00</td><td>1.000000e+00</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td> TRUE</td><td> TRUE</td><td> TRUE</td><td> TRUE</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 4 × 29\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & taxon & lfc\\_(Intercept) & lfc\\_stand\\_choles & lfc\\_health\\_status2 & lfc\\_health\\_status3 & se\\_(Intercept) & se\\_stand\\_choles & se\\_health\\_status2 & se\\_health\\_status3 & W\\_(Intercept) & ⋯ & q\\_health\\_status2 & q\\_health\\_status3 & diff\\_(Intercept) & diff\\_stand\\_choles & diff\\_health\\_status2 & diff\\_health\\_status3 & passed\\_ss\\_(Intercept) & passed\\_ss\\_stand\\_choles & passed\\_ss\\_health\\_status2 & passed\\_ss\\_health\\_status3\\\\\n",
       "  & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <dbl> & <dbl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl>\\\\\n",
       "\\hline\n",
       "\t1 & T1 & -0.6597409 &  0.030997815 & -0.03692064 & -0.02252950 & 0.05046109 & 0.02209994 & 0.08283890 & 0.08287436 & -13.074251 & ⋯ & 1.000000e+00 & 1.000000e+00 &  TRUE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE\\\\\n",
       "\t2 & T2 &  0.3163284 & -0.015585963 & -2.00692293 &  0.98887925 & 0.02811085 & 0.01591812 & 0.03515360 & 0.03680906 &  11.252893 & ⋯ & 2.018298e-96 & 5.928845e-55 &  TRUE & FALSE &  TRUE &  TRUE &  TRUE &  TRUE &  TRUE &  TRUE\\\\\n",
       "\t3 & T3 & -1.2393678 & -0.028958697 & -0.92761498 & -0.89656130 & 0.06904852 & 0.02909475 & 0.12633810 & 0.13062341 & -17.949231 & ⋯ & 5.015500e-08 & 3.335914e-07 &  TRUE & FALSE &  TRUE &  TRUE & FALSE & FALSE & FALSE & FALSE\\\\\n",
       "\t4 & T4 & -0.0420827 &  0.009269526 &  0.01861696 &  0.01487885 & 0.03146651 & 0.01529956 & 0.03787916 & 0.03601447 &  -1.337381 & ⋯ & 1.000000e+00 & 1.000000e+00 & FALSE & FALSE & FALSE & FALSE &  TRUE &  TRUE &  TRUE &  TRUE\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 4 × 29\n",
       "\n",
       "| <!--/--> | taxon &lt;chr&gt; | lfc_(Intercept) &lt;dbl&gt; | lfc_stand_choles &lt;dbl&gt; | lfc_health_status2 &lt;dbl&gt; | lfc_health_status3 &lt;dbl&gt; | se_(Intercept) &lt;dbl&gt; | se_stand_choles &lt;dbl&gt; | se_health_status2 &lt;dbl&gt; | se_health_status3 &lt;dbl&gt; | W_(Intercept) &lt;dbl&gt; | ⋯ ⋯ | q_health_status2 &lt;dbl&gt; | q_health_status3 &lt;dbl&gt; | diff_(Intercept) &lt;lgl&gt; | diff_stand_choles &lt;lgl&gt; | diff_health_status2 &lt;lgl&gt; | diff_health_status3 &lt;lgl&gt; | passed_ss_(Intercept) &lt;lgl&gt; | passed_ss_stand_choles &lt;lgl&gt; | passed_ss_health_status2 &lt;lgl&gt; | passed_ss_health_status3 &lt;lgl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | T1 | -0.6597409 |  0.030997815 | -0.03692064 | -0.02252950 | 0.05046109 | 0.02209994 | 0.08283890 | 0.08287436 | -13.074251 | ⋯ | 1.000000e+00 | 1.000000e+00 |  TRUE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE |\n",
       "| 2 | T2 |  0.3163284 | -0.015585963 | -2.00692293 |  0.98887925 | 0.02811085 | 0.01591812 | 0.03515360 | 0.03680906 |  11.252893 | ⋯ | 2.018298e-96 | 5.928845e-55 |  TRUE | FALSE |  TRUE |  TRUE |  TRUE |  TRUE |  TRUE |  TRUE |\n",
       "| 3 | T3 | -1.2393678 | -0.028958697 | -0.92761498 | -0.89656130 | 0.06904852 | 0.02909475 | 0.12633810 | 0.13062341 | -17.949231 | ⋯ | 5.015500e-08 | 3.335914e-07 |  TRUE | FALSE |  TRUE |  TRUE | FALSE | FALSE | FALSE | FALSE |\n",
       "| 4 | T4 | -0.0420827 |  0.009269526 |  0.01861696 |  0.01487885 | 0.03146651 | 0.01529956 | 0.03787916 | 0.03601447 |  -1.337381 | ⋯ | 1.000000e+00 | 1.000000e+00 | FALSE | FALSE | FALSE | FALSE |  TRUE |  TRUE |  TRUE |  TRUE |\n",
       "\n"
      ],
      "text/plain": [
       "  taxon lfc_(Intercept) lfc_stand_choles lfc_health_status2 lfc_health_status3\n",
       "1 T1    -0.6597409       0.030997815     -0.03692064        -0.02252950       \n",
       "2 T2     0.3163284      -0.015585963     -2.00692293         0.98887925       \n",
       "3 T3    -1.2393678      -0.028958697     -0.92761498        -0.89656130       \n",
       "4 T4    -0.0420827       0.009269526      0.01861696         0.01487885       \n",
       "  se_(Intercept) se_stand_choles se_health_status2 se_health_status3\n",
       "1 0.05046109     0.02209994      0.08283890        0.08287436       \n",
       "2 0.02811085     0.01591812      0.03515360        0.03680906       \n",
       "3 0.06904852     0.02909475      0.12633810        0.13062341       \n",
       "4 0.03146651     0.01529956      0.03787916        0.03601447       \n",
       "  W_(Intercept) ⋯ q_health_status2 q_health_status3 diff_(Intercept)\n",
       "1 -13.074251    ⋯ 1.000000e+00     1.000000e+00      TRUE           \n",
       "2  11.252893    ⋯ 2.018298e-96     5.928845e-55      TRUE           \n",
       "3 -17.949231    ⋯ 5.015500e-08     3.335914e-07      TRUE           \n",
       "4  -1.337381    ⋯ 1.000000e+00     1.000000e+00     FALSE           \n",
       "  diff_stand_choles diff_health_status2 diff_health_status3\n",
       "1 FALSE             FALSE               FALSE              \n",
       "2 FALSE              TRUE                TRUE              \n",
       "3 FALSE              TRUE                TRUE              \n",
       "4 FALSE             FALSE               FALSE              \n",
       "  passed_ss_(Intercept) passed_ss_stand_choles passed_ss_health_status2\n",
       "1 FALSE                 FALSE                  FALSE                   \n",
       "2  TRUE                  TRUE                   TRUE                   \n",
       "3 FALSE                 FALSE                  FALSE                   \n",
       "4  TRUE                  TRUE                   TRUE                   \n",
       "  passed_ss_health_status3\n",
       "1 FALSE                   \n",
       "2  TRUE                   \n",
       "3 FALSE                   \n",
       "4  TRUE                   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(output$res, 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed7700ce-1d4f-481f-bc26-c45ba9d9253d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# log(count) = intercept + log_fold_change1 * healthy_status1 \n",
    "#                        + log_fold_change2 * healthy_status2 \n",
    "#                        + log_fold_change3 * healthy_status3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c81c8975-0a7e-4eab-83f1-3190a4879305",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 4 × 22</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>taxon</th><th scope=col>lfc_health_status2</th><th scope=col>lfc_health_status3</th><th scope=col>lfc_health_status3_health_status2</th><th scope=col>se_health_status2</th><th scope=col>se_health_status3</th><th scope=col>se_health_status3_health_status2</th><th scope=col>W_health_status2</th><th scope=col>W_health_status3</th><th scope=col>W_health_status3_health_status2</th><th scope=col>⋯</th><th scope=col>p_health_status3_health_status2</th><th scope=col>q_health_status2</th><th scope=col>q_health_status3</th><th scope=col>q_health_status3_health_status2</th><th scope=col>diff_health_status2</th><th scope=col>diff_health_status3</th><th scope=col>diff_health_status3_health_status2</th><th scope=col>passed_ss_health_status2</th><th scope=col>passed_ss_health_status3</th><th scope=col>passed_ss_health_status3_health_status2</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>T1</td><td>-0.03692064</td><td>-0.02252950</td><td> 0.014391146</td><td>0.08283890</td><td>0.08287436</td><td>0.05071471</td><td> -0.4456921</td><td>-0.2718513</td><td> 0.2837667</td><td>⋯</td><td> 1.000000e+00</td><td>1.000000e+00</td><td>1.000000e+00</td><td> 1.00000e+00</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td> TRUE</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>T2</td><td>-2.00692293</td><td> 0.98887925</td><td> 2.995802184</td><td>0.03515360</td><td>0.03680906</td><td>0.03928951</td><td>-57.0901149</td><td>26.8651029</td><td>76.2494187</td><td>⋯</td><td>2.692934e-115</td><td>3.900530e-97</td><td>1.103992e-55</td><td>4.90114e-114</td><td> TRUE</td><td> TRUE</td><td> TRUE</td><td> TRUE</td><td> TRUE</td><td> TRUE</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>T3</td><td>-0.92761498</td><td>-0.89656130</td><td> 0.031053674</td><td>0.12633810</td><td>0.13062341</td><td>0.05287066</td><td> -7.3423215</td><td>-6.8637108</td><td> 0.5873517</td><td>⋯</td><td> 5.591354e-01</td><td>1.086692e-08</td><td>6.830681e-08</td><td> 1.00000e+00</td><td> TRUE</td><td> TRUE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>T4</td><td> 0.01861696</td><td> 0.01487885</td><td>-0.003738113</td><td>0.03787916</td><td>0.03601447</td><td>0.03230784</td><td>  0.4914830</td><td> 0.4131353</td><td>-0.1157030</td><td>⋯</td><td> 1.000000e+00</td><td>1.000000e+00</td><td>1.000000e+00</td><td> 1.00000e+00</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td> TRUE</td><td> TRUE</td><td> TRUE</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 4 × 22\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & taxon & lfc\\_health\\_status2 & lfc\\_health\\_status3 & lfc\\_health\\_status3\\_health\\_status2 & se\\_health\\_status2 & se\\_health\\_status3 & se\\_health\\_status3\\_health\\_status2 & W\\_health\\_status2 & W\\_health\\_status3 & W\\_health\\_status3\\_health\\_status2 & ⋯ & p\\_health\\_status3\\_health\\_status2 & q\\_health\\_status2 & q\\_health\\_status3 & q\\_health\\_status3\\_health\\_status2 & diff\\_health\\_status2 & diff\\_health\\_status3 & diff\\_health\\_status3\\_health\\_status2 & passed\\_ss\\_health\\_status2 & passed\\_ss\\_health\\_status3 & passed\\_ss\\_health\\_status3\\_health\\_status2\\\\\n",
       "  & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl>\\\\\n",
       "\\hline\n",
       "\t1 & T1 & -0.03692064 & -0.02252950 &  0.014391146 & 0.08283890 & 0.08287436 & 0.05071471 &  -0.4456921 & -0.2718513 &  0.2837667 & ⋯ &  1.000000e+00 & 1.000000e+00 & 1.000000e+00 &  1.00000e+00 & FALSE & FALSE & FALSE & FALSE & FALSE &  TRUE\\\\\n",
       "\t2 & T2 & -2.00692293 &  0.98887925 &  2.995802184 & 0.03515360 & 0.03680906 & 0.03928951 & -57.0901149 & 26.8651029 & 76.2494187 & ⋯ & 2.692934e-115 & 3.900530e-97 & 1.103992e-55 & 4.90114e-114 &  TRUE &  TRUE &  TRUE &  TRUE &  TRUE &  TRUE\\\\\n",
       "\t3 & T3 & -0.92761498 & -0.89656130 &  0.031053674 & 0.12633810 & 0.13062341 & 0.05287066 &  -7.3423215 & -6.8637108 &  0.5873517 & ⋯ &  5.591354e-01 & 1.086692e-08 & 6.830681e-08 &  1.00000e+00 &  TRUE &  TRUE & FALSE & FALSE & FALSE & FALSE\\\\\n",
       "\t4 & T4 &  0.01861696 &  0.01487885 & -0.003738113 & 0.03787916 & 0.03601447 & 0.03230784 &   0.4914830 &  0.4131353 & -0.1157030 & ⋯ &  1.000000e+00 & 1.000000e+00 & 1.000000e+00 &  1.00000e+00 & FALSE & FALSE & FALSE &  TRUE &  TRUE &  TRUE\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 4 × 22\n",
       "\n",
       "| <!--/--> | taxon &lt;chr&gt; | lfc_health_status2 &lt;dbl&gt; | lfc_health_status3 &lt;dbl&gt; | lfc_health_status3_health_status2 &lt;dbl&gt; | se_health_status2 &lt;dbl&gt; | se_health_status3 &lt;dbl&gt; | se_health_status3_health_status2 &lt;dbl&gt; | W_health_status2 &lt;dbl&gt; | W_health_status3 &lt;dbl&gt; | W_health_status3_health_status2 &lt;dbl&gt; | ⋯ ⋯ | p_health_status3_health_status2 &lt;dbl&gt; | q_health_status2 &lt;dbl&gt; | q_health_status3 &lt;dbl&gt; | q_health_status3_health_status2 &lt;dbl&gt; | diff_health_status2 &lt;lgl&gt; | diff_health_status3 &lt;lgl&gt; | diff_health_status3_health_status2 &lt;lgl&gt; | passed_ss_health_status2 &lt;lgl&gt; | passed_ss_health_status3 &lt;lgl&gt; | passed_ss_health_status3_health_status2 &lt;lgl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | T1 | -0.03692064 | -0.02252950 |  0.014391146 | 0.08283890 | 0.08287436 | 0.05071471 |  -0.4456921 | -0.2718513 |  0.2837667 | ⋯ |  1.000000e+00 | 1.000000e+00 | 1.000000e+00 |  1.00000e+00 | FALSE | FALSE | FALSE | FALSE | FALSE |  TRUE |\n",
       "| 2 | T2 | -2.00692293 |  0.98887925 |  2.995802184 | 0.03515360 | 0.03680906 | 0.03928951 | -57.0901149 | 26.8651029 | 76.2494187 | ⋯ | 2.692934e-115 | 3.900530e-97 | 1.103992e-55 | 4.90114e-114 |  TRUE |  TRUE |  TRUE |  TRUE |  TRUE |  TRUE |\n",
       "| 3 | T3 | -0.92761498 | -0.89656130 |  0.031053674 | 0.12633810 | 0.13062341 | 0.05287066 |  -7.3423215 | -6.8637108 |  0.5873517 | ⋯ |  5.591354e-01 | 1.086692e-08 | 6.830681e-08 |  1.00000e+00 |  TRUE |  TRUE | FALSE | FALSE | FALSE | FALSE |\n",
       "| 4 | T4 |  0.01861696 |  0.01487885 | -0.003738113 | 0.03787916 | 0.03601447 | 0.03230784 |   0.4914830 |  0.4131353 | -0.1157030 | ⋯ |  1.000000e+00 | 1.000000e+00 | 1.000000e+00 |  1.00000e+00 | FALSE | FALSE | FALSE |  TRUE |  TRUE |  TRUE |\n",
       "\n"
      ],
      "text/plain": [
       "  taxon lfc_health_status2 lfc_health_status3 lfc_health_status3_health_status2\n",
       "1 T1    -0.03692064        -0.02252950         0.014391146                     \n",
       "2 T2    -2.00692293         0.98887925         2.995802184                     \n",
       "3 T3    -0.92761498        -0.89656130         0.031053674                     \n",
       "4 T4     0.01861696         0.01487885        -0.003738113                     \n",
       "  se_health_status2 se_health_status3 se_health_status3_health_status2\n",
       "1 0.08283890        0.08287436        0.05071471                      \n",
       "2 0.03515360        0.03680906        0.03928951                      \n",
       "3 0.12633810        0.13062341        0.05287066                      \n",
       "4 0.03787916        0.03601447        0.03230784                      \n",
       "  W_health_status2 W_health_status3 W_health_status3_health_status2 ⋯\n",
       "1  -0.4456921      -0.2718513        0.2837667                      ⋯\n",
       "2 -57.0901149      26.8651029       76.2494187                      ⋯\n",
       "3  -7.3423215      -6.8637108        0.5873517                      ⋯\n",
       "4   0.4914830       0.4131353       -0.1157030                      ⋯\n",
       "  p_health_status3_health_status2 q_health_status2 q_health_status3\n",
       "1  1.000000e+00                   1.000000e+00     1.000000e+00    \n",
       "2 2.692934e-115                   3.900530e-97     1.103992e-55    \n",
       "3  5.591354e-01                   1.086692e-08     6.830681e-08    \n",
       "4  1.000000e+00                   1.000000e+00     1.000000e+00    \n",
       "  q_health_status3_health_status2 diff_health_status2 diff_health_status3\n",
       "1  1.00000e+00                    FALSE               FALSE              \n",
       "2 4.90114e-114                     TRUE                TRUE              \n",
       "3  1.00000e+00                     TRUE                TRUE              \n",
       "4  1.00000e+00                    FALSE               FALSE              \n",
       "  diff_health_status3_health_status2 passed_ss_health_status2\n",
       "1 FALSE                              FALSE                   \n",
       "2  TRUE                               TRUE                   \n",
       "3 FALSE                              FALSE                   \n",
       "4 FALSE                               TRUE                   \n",
       "  passed_ss_health_status3 passed_ss_health_status3_health_status2\n",
       "1 FALSE                     TRUE                                  \n",
       "2  TRUE                     TRUE                                  \n",
       "3 FALSE                    FALSE                                  \n",
       "4  TRUE                     TRUE                                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(output$res_pair,4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ebb21cb-3825-4360-8c49-bc499d8d307f",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## 4. Barplot of taxa with differentially abundance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b8d97860-d6d1-45d7-aede-1b0267cdddcc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 3 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>stand_bmi</th><th scope=col>stand_choles</th><th scope=col>health_status</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>S1</th><td>-7.046117</td><td>-0.05133271</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>S2</th><td>-8.183509</td><td>-1.30178093</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>S3</th><td>-7.369319</td><td>-1.20561056</td><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 3 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & stand\\_bmi & stand\\_choles & health\\_status\\\\\n",
       "  & <dbl> & <dbl> & <fct>\\\\\n",
       "\\hline\n",
       "\tS1 & -7.046117 & -0.05133271 & 1\\\\\n",
       "\tS2 & -8.183509 & -1.30178093 & 1\\\\\n",
       "\tS3 & -7.369319 & -1.20561056 & 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 3 × 3\n",
       "\n",
       "| <!--/--> | stand_bmi &lt;dbl&gt; | stand_choles &lt;dbl&gt; | health_status &lt;fct&gt; |\n",
       "|---|---|---|---|\n",
       "| S1 | -7.046117 | -0.05133271 | 1 |\n",
       "| S2 | -8.183509 | -1.30178093 | 1 |\n",
       "| S3 | -7.369319 | -1.20561056 | 1 |\n",
       "\n"
      ],
      "text/plain": [
       "   stand_bmi stand_choles health_status\n",
       "S1 -7.046117 -0.05133271  1            \n",
       "S2 -8.183509 -1.30178093  1            \n",
       "S3 -7.369319 -1.20561056  1            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 2 × 150 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>S1</th><th scope=col>S2</th><th scope=col>S3</th><th scope=col>S4</th><th scope=col>S5</th><th scope=col>S6</th><th scope=col>S7</th><th scope=col>S8</th><th scope=col>S9</th><th scope=col>S10</th><th scope=col>⋯</th><th scope=col>S141</th><th scope=col>S142</th><th scope=col>S143</th><th scope=col>S144</th><th scope=col>S145</th><th scope=col>S146</th><th scope=col>S147</th><th scope=col>S148</th><th scope=col>S149</th><th scope=col>S150</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>T1</th><td>  1</td><td>0</td><td>  1</td><td>  3</td><td>  2</td><td>0</td><td> 1</td><td> 0</td><td> 0</td><td>  3</td><td>⋯</td><td>  142</td><td>  2</td><td>  208</td><td>   232</td><td>  105</td><td>   64</td><td>   317</td><td>   41</td><td>  320</td><td>  14</td></tr>\n",
       "\t<tr><th scope=row>T2</th><td>102</td><td>1</td><td>156</td><td>535</td><td>178</td><td>0</td><td>99</td><td>21</td><td>28</td><td>400</td><td>⋯</td><td>41555</td><td>940</td><td>69158</td><td>150981</td><td>68451</td><td>40298</td><td>258659</td><td>23841</td><td>79188</td><td>7491</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 2 × 150 of type dbl\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & S1 & S2 & S3 & S4 & S5 & S6 & S7 & S8 & S9 & S10 & ⋯ & S141 & S142 & S143 & S144 & S145 & S146 & S147 & S148 & S149 & S150\\\\\n",
       "\\hline\n",
       "\tT1 &   1 & 0 &   1 &   3 &   2 & 0 &  1 &  0 &  0 &   3 & ⋯ &   142 &   2 &   208 &    232 &   105 &    64 &    317 &    41 &   320 &   14\\\\\n",
       "\tT2 & 102 & 1 & 156 & 535 & 178 & 0 & 99 & 21 & 28 & 400 & ⋯ & 41555 & 940 & 69158 & 150981 & 68451 & 40298 & 258659 & 23841 & 79188 & 7491\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 2 × 150 of type dbl\n",
       "\n",
       "| <!--/--> | S1 | S2 | S3 | S4 | S5 | S6 | S7 | S8 | S9 | S10 | ⋯ | S141 | S142 | S143 | S144 | S145 | S146 | S147 | S148 | S149 | S150 |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| T1 |   1 | 0 |   1 |   3 |   2 | 0 |  1 |  0 |  0 |   3 | ⋯ |   142 |   2 |   208 |    232 |   105 |    64 |    317 |    41 |   320 |   14 |\n",
       "| T2 | 102 | 1 | 156 | 535 | 178 | 0 | 99 | 21 | 28 | 400 | ⋯ | 41555 | 940 | 69158 | 150981 | 68451 | 40298 | 258659 | 23841 | 79188 | 7491 |\n",
       "\n"
      ],
      "text/plain": [
       "   S1  S2 S3  S4  S5  S6 S7 S8 S9 S10 ⋯ S141  S142 S143  S144   S145  S146 \n",
       "T1   1 0    1   3   2 0   1  0  0   3 ⋯   142   2    208    232   105    64\n",
       "T2 102 1  156 535 178 0  99 21 28 400 ⋯ 41555 940  69158 150981 68451 40298\n",
       "   S147   S148  S149  S150\n",
       "T1    317    41   320   14\n",
       "T2 258659 23841 79188 7491"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Meta data\n",
    "head(smd,3)\n",
    "# Count data\n",
    "head(otu_data,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "18bdfb7b-ec62-4245-9822-808c42ae31af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 2 × 2 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>count</th><th scope=col>health_status</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>S1</th><td>102</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>S2</th><td>  1</td><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 2 × 2 of type dbl\n",
       "\\begin{tabular}{r|ll}\n",
       "  & count & health\\_status\\\\\n",
       "\\hline\n",
       "\tS1 & 102 & 1\\\\\n",
       "\tS2 &   1 & 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 2 × 2 of type dbl\n",
       "\n",
       "| <!--/--> | count | health_status |\n",
       "|---|---|---|\n",
       "| S1 | 102 | 1 |\n",
       "| S2 |   1 | 1 |\n",
       "\n"
      ],
      "text/plain": [
       "   count health_status\n",
       "S1 102   1            \n",
       "S2   1   1            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "taxa2 = cbind(otu_data['T2',], smd[,'health_status'])\n",
    "colnames(taxa2) = c('count', 'health_status')\n",
    "head(taxa2,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "5c119868-3be4-4fbb-a6d3-0f9fb5dc8a49",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAIAAAByhViMAAAABmJLR0QA/wD/AP+gvaeTAAAg\nAElEQVR4nO3db4xj530f+odau5Fc1459i3uT3tyrOiVtdO8mTRukDQ6N1oKKBORIihNAU8Ct\nMQaKkkAbg7wIFinaCfqi47zoIA2ZWirIokA2dQJ0UrSGnCEhG4YExDwxagRF7M0gNs+N5Dqu\ngdZubKWw5Mi7vC+Olsv5u5zdmTnzHH4+LwTOw2fJ38xQ2q+ec57fU5nNZgEAgPg9VHQBAACc\nDcEOAKAkBDsAgJIQ7AAASkKwAwAoCcEOAKAkBDsAgJIQ7AAASkKwAwAoCcEOAKAkBDsAgJIQ\n7AAASkKwAwAoCcEOAKAkBDsAgJIQ7AAASkKwAwAoCcEOAKAkBDsAgJIQ7AAASkKwAwAoCcEO\nAKAkBDsAgJIQ7AAASkKwAwAoCcEOAKAkBDsAgJIQ7AAASkKwAwAoCcEOAKAkBDsAgJIQ7AAA\nSkKwAwAoCcEOAKAkBDsAgJIQ7AAASkKwAwAoCcEOAKAkBDsAgJIQ7KCEsn69spx6Pyu62ONl\n2bhdr9/9Vur1dn98rvUe+MG1x/eadcyMgo3bUfx+gXMg2AGXUtav12rNYZqm86E0HXabtfr5\nhalsdydd/Hr48UuZ2wCOJdgBl9C4XeumRz+VDpvntA51MNeFMNyy4gVERbCDEqp2JrNFo9ad\nZ5LedN8zk061yEKPkfW3hncezyue9pI7Y2l3+xyW0u7mulbrzg8s3dmV7ICICHZACCFk4367\nvniDWb1eb++/oe3unVv77y07bnyZ1zy6lrsrZ0nvxp3oWe3cuBvtbn7pzPPW3XdNrl5//ymS\nXTa++10e8Q0ed8Pb0XfqzWfX+9n+H+BxP7vFt6/c4+e73G/k9DW8cTvk4usedzvk/onnf9sk\nrJwZUHrHr9jlFhbDDmqNjnydu08s/tnFycu+5lHljFpJkiTJoXLv9X08iLv1Jr3pyW+18K21\nWq3D32ZyzA9t/ystvsro8OzkqFc+VMzoqEl3Fxz3zV/6N3LKGo4u4o3X3Tf1uAqSe30igKUJ\ndrACTg5EByLNbDabjo5Ja4vRLulNl4l1937NpR0dhc7GoYJP+JmdkJCOqO8+g90Sr3zvyQtv\neYrfyKlquMcPY+F7PvFlzz6pw4pyKRZW3Xj7jW0Kdy97Vhtr60df9Wxcn/89nna3+/2N+RaH\n1mjQuL/XXFK2+Gbvb5w49/Qvfvc67PpaNYQQGstejU16oztB6W5yOZtdF3deejE83d2pu3gn\n4nxtbF9aW3Dfv5GTa7j7usfUML8d8qj7Jhcmns9tk7CKLj5LAhftPi5hHv9HjlyiWWoF7UGu\npC4u95znddiF7+TYchd/Aseuws2fuO8Vu8XpC8NHXAM/9or1vX5UR9e2fA0nfnNJq9cbTQ9v\nfdn3WTnPVVhYSVbsgAVZlo37/Xa93hweN2XfJobcvtW6+3nNexQ1bleaC8s9N856K+/Cbo2F\ntcBl1uzurO+9ofrua3cepnvTBytq30vXrh5K09O99KiZR0/eb+nfyL1qyL50887Da+/e94Po\nTGaTQafTqFYPVrt/tbV6d9XwHDbEwAoS7IA3dktWKpVKrVZrdrvDha7ARzgQ7ZLe9aNi3ele\n8/jS+vXaYqqb3rtDyxEHb5zY1PjoXLdUstsfZ5bIVBdhIV7uc0a/kQUL4fJq7aSJCwlw2Nz3\nm7nbr/CBozAQBDsgjNv1WrM7vHOPWdLqjabTE+9039/J96jUc/rXPPqN+vW7f/EnrdESqe70\nFr+b/anjbqCMv5/dGf1GgEtOsIMVl/W35n/b96az2WQyuHsB7Zg/sbH/VIi0u7F/q8DpX/Po\n91lMdb3pZNA4j27Kh8+bOMoxN/cfvHx4dwnrnk4x9R4OLnUtLI/Nh87kN3LI3QXKeyy3LSwi\nHn8r3UnX84ElCXaw2hYvQ27eXQ47IhvceeZgrAvhYLQ79WseZeFQsaQ1OtURGQcP3jgxMyyX\n68IxB8ceWMlb+BaP2Li7L/yc7qdxhIWrvvvj5RHf0Zn8Ro6wkNcO/HjG7Xq93u7Puw/frdYB\nvHCuBDvgCMctJ+3rOTKaThfaVWzcq8HHKZaosv7d+/pbo8k5LuXsO+XiiB2kC7s2j0523dqd\nwxiyxQNu795zthDAhltvBJ1s3D4iHp/OwraD/UUs/8oPvmh49zbEsHCEbzZubw3TNB12u81a\nPrpQ7bBZ78+r7dcrlcUACDyos91kC1xGJ/UZ2X+axHR2qBXawh85os/Hka0/TvOaJ9d7nLPq\neHJ8z5Cjvuc3riKerkHxEt/NvdsZ32c34/mLnOo3cqoa7vXDuPsS96hWi2I4E1bsYMU1BqOF\nRZdapVKpNfct+dy9gLjQjHbec6Ta2Zz/8fmq3Sle8yjjj99nX5RTW7xqeXCD6xsWlpoOr9nl\nB58dlPSmi2uMiz+NxVlHntd1Ko3B0bFq4UyxI2q4j9/ICaqdyfFHiiWt0d1r6MdV+8bEM29i\nA6tJsIOV1xhMR72Fv5uTpDVaPC31jVMUxu351dF9reQWM8N8i8GSr3mkYnLdsYdZnJjs1m9M\npqPW3XSXtHpHbN098NNIktZoOp1cv/pg1YcQQrUzOfzSs8H7D898kN/IPTQGk+m+n0IIyRvf\n4/6L6NXOGxOTfRN7o/PaGAMrqDKbzYquAQCAM2DFDgCgJAQ7AICSEOwAAEpCsAMAKAnBDgCg\nJAQ7AICSEOwAAEpCsAMAKAnBDgCgJAQ7AICSEOwAAEpCsAMAKAnBDgCgJAQ7AICSEOwAAEpC\nsAMAKAnBDgCgJAQ7AICSEOwAAEpCsAMAKAnBDgCgJAQ7AICSEOwAAEpCsAMAKAnBDgCgJAQ7\nAICSEOwAAEpCsAMAKAnBDgCgJAQ7AICSEOwAAEpCsAMAKAnBDgCgJKIPdlmWtdvter1er9f7\n/X7R5QAAFKYym82KrmFZlUolhLBYcL/f73a7B6aNRqNGo3GhlQEAXAIRB7vxeNxsNkMIo9Go\nVquFEHZ3d/OcN51Oq9VqcZUCABQg4kuxW1tb4c76XLVarVarnU6n1+uFELa3t4uuDgDgokW8\nYnf4ymwIIcuyfPUuou8LAOBMlC3YnTB+3771rW/duHHj1VdfPasXBACi9sgjj2xsbLz97W8v\nupBDZvE4UHCr1QohTKfTxTnT6TSEkCTJGb7vM888U+RvCAC4fJ555pkzDBtn5U1F/1hOrVKp\nJEly7dq1q1evhhC2t7cHg8H82fzuumvXrp3hO77++ushhH/7b//tX/krf+UMXxYAlnf79u3/\n/J//81//63/9oYcivj++HH7v937v7//9v5/Hg8smpmA3Go2+9KUv7ezspGmapmk+OBwO58Eu\nvwgbQliMemflPe95z4/+6I+e+csCwJJ+7Md+rOgSCCGE1157regSjhVT6m80Gp1OZzKZ5IuN\no9Go1+slSbI4p9VqzeK5axAAlnTr1q1PfvKTt27dKroQLrWYVuwOaDQaedSbj4h0AJTVeDx+\n8sknP/GJTzzxxBNF18LlFdOKHQCsrLw5gxYNnCzKFbssy3Z3d/f29m7evJmP5Hsp1tbWHDgB\nAKysyIJdlmUbGxvznRNz+Ui3202S5MaNG+IdALCCYgp281MlkiTZ3NzMH89Np9Otra00TWu1\nmrNiASiZRx55ZP5POE5MwS7vUddqtY7sZlKtVhuNRrvdHg6HB5rbAUDsGo3G888///jjjxdd\nCJdaTJsnhsNhuFePuvzZfCYAlMaVK1d+4id+4sqVK0UXwqUWU7ADgJWljx3LiCnY5b2Ix+Px\nCXPyZw90LQaA2I3H45/8yZ88+S9BiCnYbW5uhhCazeZxH+vxeNxsNuczAaA09LFjGTFtnmg0\nGr1er9vt5uktLKzMLTZA6fV6jUajgPoAAAoVU7ALIXQ6nbW1td3d3Z2dnTRNF/NckiTr6+t6\nFAMAKyuyYBdCqFarnU5n8YhYACg9fexYRnzBDgBWkD52LCP6YFepVEIIs9ms6EIA4BzlfeyK\nroLLLqZdsQCwsvSxYxmCHQBEQB87lhHlpdgsy04YsSsWgPLRx45lRBnsarXaCSOnut/u1q1b\no9HotddeO2HOf/kv/yWE8Prrr5+mRgCAixZlsGu1WvPHw+HwwMipvPDCC0899dQyM3/jN37j\nfe973/29CwDABYgy2A0Gg/njPNgtjpzKY4899txzz528Yvfss8+++OKLP/ADP3B/bwEAD04f\nO5YRZbA7Q1euXHnyySdPnjMajUIIDz1kowkAhdHHjmWserADgCjoY8cyol+FSpIkSZKiqwCA\n86WPHcuIPthNJpPJZFJ0FQBwvvSxYxnRBzsAWAX62LGM6INdv99vt9vtdjv/0v/KAAArK+LN\nE+PxuNlszr/MO540m80kSVycBQBWUKwrdvNUl/cimWu1Wmma9vv9guoCgHOhjx3LiDXYbW1t\nhRBGo1Gj0Vgcv379eghhZ2enmLIA4HzkfewO/K0HB8R6KTZN0xDC4c93tVqdPwsApaGPHcuI\ndcUOAFaKPnYsI9ZglzclzrLswHi+K1bLYgBKRh87lhFrsFtfXw8hbGxsLA7Od1TkzwJAaehj\nxzJivceu0+ns7e0Nh8NKpZKPzB+0Wq1Op1NcaQAAxYh1xS6EMBgMRqPR4lXXJElGo1He0A4A\nYNXEumKXazQaNn4DsAr0sWMZcQc7AFgReR+7xx9/vOhCuNQEOwCIgD52LCPie+wAYHXoY8cy\nBDsAiIA+diwjpkux9Xp9+cmTyeT8KgGAC6aPHcuIKdhdu3ZtOBwWXQUAwCUV06XYwWDQ6/VC\nCK1Wa3YvRRcLAHDRYgp2IYROp9NqtYbDoZsMAFgp+tixjJguxeYGg8HNmzebzaZlOQBWhz52\nLCOyFbvc5uZmq9WyaAfA6sj72F25cqXoQrjU4luxC04SA2D13Lp169Of/vTjjz8u23GCKFfs\nAGDV6GPHMgQ7AIiAPnYsQ7ADACgJwQ4AoCQEOwCIgD52LCPKXbEAsGr0sWMZgh0ARCDvY1d0\nFVx2LsUCQARu3br1yU9+8tatW0UXwqUm2AFABPSxYxmCHQBEQB87liHYAQCUhGAHAFASgh0A\nREAfO5ah3QkAREAfO5Yh2AFABPSxYxkuxQJABPSxYxmCHQBEQB87liHYAUAE9LFjGYIdAEBJ\nCHYAACUh2AFABPSxYxnanQBABPSxYxmCHQBEQB87luFSLABEQB87liHYAUAE9LFjGYIdAERA\nHzuWIdgBAJSEYAcAUBKCHQBEQB87lqHdCQBEQB87liHYAUAE9LFjGS7FAkAE9LFjGYIdAERA\nHzuWIdgBQAT0sWMZgh0AQEkIdgAAJSHYAUAE9LFjGdqdAEAE9LFjGYIdAERAHzuW4VIsAERA\nHzuWIdgBQAT0sWMZgh0AREAfO5Yh2AEAlIRgBwBQEoIdAERAHzuWod0JAERAHzuWIdgBQAT0\nsWMZLsUCQAT0sWMZgh0AREAfO5Yh2AFABPSxYxmCHQBASQh2AAAlIdgBQAT0sWMZ2p0AQAT0\nsWMZgh0AREAfO5bhUiwAREAfO5Yh2AFABPSxYxklCXb1er1erxddBQCcF33sWEZJ7rFL07To\nEgAAChZTsLvnmtzihMlkcs7lAABcLjEFu3CvlTnrdgCUlT52LCOme+wmk0mSJCGEXq832y+f\ncHgEAMoh72PXaDSKLoRLLaZgF0KYTCa9Xq/b7doqAcBKyfvYXblypehCuNQiC3YhhE6nM51O\n0zStVCpZlhVdDgBcBH3sWEZ8wS6EUK1WZ7NZkiS1Wq3f7xddDgCcO33sWEaUwS43mUxarZbL\nsgCsAn3sWEbEwS6EMBgMRqORzbAAACG6dieHNRoNe2ABAELsK3YAsCL0sWMZUa7YZVm2u7u7\nt7d38+bNfOTatWtXr15dW1urVqvF1gYA5yHvY/f4448XXQiXWmTBLsuyjY2NwzfV5SPdbjdJ\nkhs3boh3AJRM3seu6Cq47GIKdlmW1Wq1EEKSJJubm/njuel0urW1laZprVabTqeyHQBlcuvW\nrU9/+tOPP/64HsWcIKZ77La3t0MIrVZrMpk0Go3qfo1GI2+AMp8JAKWhjx3LiCnYDYfDEMJg\nMDhhTv5sPhMASkMfO5YRU7ADAOAEMQW7JElCCCevQufP5jMBAFZKTMFuc3MzhNBsNo/LduPx\nuNlszmcCQGnoY8cyYtoV22g0er1et9vN01tYWJlbbIDS6/UajUYB9QHAudHHjmXEFOxCCJ1O\nZ21tbXd3d2dnJ03TxTyXJMn6+roexQCUkj52LCOyYBdCqFarnU6n0+kUXQgAXBx97FhGTPfY\nAcDK0seOZUQf7CqVSqVSKboKADhf+tixjOiDHQAAOcEOAKAk4ts8EULIsuyEkVPtir1169Zo\nNHrttddOmPPyyy+HEG7fvn2KEgHgTOljxzKiDHa1Wu2EkdlstvxLvfDCC0899dQyM1966aXl\nXxYAzpY+diwjymDXarXmj4fD4YGRU3nssceee+65k1fsnn322RdffPFd73rX/b0FADw4fexY\nRpTBbjAYzB/nwW5x5FSuXLny5JNPnjxnNBqFEB56yP2IABRGHzuWIawAQAT0sWMZUa7YAbCC\nXnnllel0WnQVhfn93//9/J/f//3fX3QtxXj44Yf/8l/+yy6gnSz6YJckSdElAHARnnjiid/+\n7d8uuoqC/eN//I+LLqFIv/Irv/LhD3+46CouteiD3WQyKboEAC7C//yf//P/eMc7/9+nP1B0\nIcW4PZv9f1/9yl/6P/+vh1byvKX//s0//pc7v/6Nb3yj6EIuu+iDHQCr48+//Xt//gMbRVdB\nAfZefulf7vx60VVEIPoL1f1+v91ut9vt/Et3lQIAKyviFbvxeNxsNudf5h1Pms1mkiSuzwIA\nKyjWFbt5qsubzM21Wq00Tfv9fkF1AQAUJtZgt7W1FUIYjUaNRmNx/Pr16yGEnZ2dYsoCAChO\nrMEuTdMQwoFUF0KoVqvzZwEAVkqswQ4AgANiDXZ5X+Isyw6M57tidS0GAFZQrMFufX09hLCx\nsa+b0XxHRf4sAMBKibXdSafT2dvbGw6HlTsNuOcPWq1Wp9MprjQAgGLEumIXQhgMBqPRaPGq\na5Iko9Eob2gHALBqYl2xyzUajcMbYwEAVlPEK3YAACyKcsUuy7K8X938y93d3b29vRDC9evX\nF58CAFgdkQW7xfNh8zNhsyyr1WrzCcPhsNVquc0OAFhBMV2KzbIsT3VJkiRJkqZpu93OO56M\nRqPpdJqfGzscDvNudgAAKyWmYLe9vR1C6PV6k8lkMpn0er3hcJimaX5ibLVabTQaebbLT5IF\nAFgpMQW74XAYQpj3qJs/WNwYmz92ViwAsIJiCnYAAJwgpmCX9yKe3z83f7B4Ymz+2FmxAMAK\niinY5SfANpvN8Xicb4/NA9ziibH5Y2fFAgArKKZ2J51OZ2dnJ03TeceTzc3NL33pS91ut1Kp\n5PtkQwhJkjgrFgBYQTEFuxBC3rhue3v76tWra2tr+U7YEEK3281TnSZ2AMDKiizYhRCq1eqB\n6NbpdCzRAQDEdI8dAAAnEOwAAEpCsAMAKAnBDgCgJAQ7AICSEOwAAEpCsAMAKAnBDgCgJAQ7\nAICSEOwAAEpCsAMAKAnBDgCgJAQ7AICSEOwAAEpCsAMAKAnBDgCgJAQ7AICSEOwAAEpCsAMA\nKAnBDgCgJAQ7AICSEOwAAEpCsAMAKAnBDgCgJAQ7AICSEOwAAEpCsAMAKAnBDgCgJAQ7AICS\nEOwAAEpCsAMAKAnBDgCgJAQ7AICSEOwAAEpCsAMAKAnBDgCgJAQ7AICSEOwAAEpCsAMAKAnB\nDgCgJAQ7AICSEOwAAEpCsAMAKAnBDgCgJAQ7AICSEOwAAEpCsAMAKAnBDgCgJAQ7AICSEOwA\nAEpCsAMAKAnBDgCgJAQ7AICSEOwAAEpCsAMAKAnBDgCgJEoS7Or1er1eL7oKAIAivanoAs5G\nmqZFlwAAULCYgt091+QWJ0wmk3MuBwDgcokp2IV7rcxZtwMAVllM99hNJpMkSUIIvV5vtl8+\n4fAIAMDqiCnYhRAmk0mv1+t2u7ZKAAAcEFmwCyF0Op3pdJqmaaVSybKs6HIAAC6L+IJdCKFa\nrc5msyRJarVav98vuhwAgEshymCXm0wmrVbLZVkAgFzEwS6EMBgMRqORzbAAACG6dieHNRoN\ne2ABAELsK3YAAMxFuWKXZdnu7u7e3t7NmzfzkWvXrl29enVtba1arRZbGwBAUSILdlmWbWxs\nHL6pLh/pdrtJkty4cUO8AwBWUEzBLsuyWq0WQkiSZHNzM388N51Ot7a20jSt1WrT6VS2AwBW\nTUzBbnt7O4TQarUGg8HhZ6vVaqPRaLfbw+Fwe3v7yDkAACUW0+aJ4XAYQjg5seXP5jMBAFZK\nTMEOAIATxBTskiQJIYzH4xPm5M/mMwEAVkpMwW5zczOE0Gw2j8t24/G42WzOZwIArJSYNk80\nGo1er9ftdvP0FhZW5hYboPR6vUajUUB9AACFiinYhRA6nc7a2tru7u7Ozk6apot5LkmS9fV1\nPYoBgJUVWbALIVSr1U6n0+l0ii4EAOByiekeOwAAThB9sKtUKpVKpegqAACKF32wAwAgJ9gB\nAJREfJsnQghZlp0wcqpdsbdu3RqNRq+99toJc15++eUQwu3bt09RIgDAhYsy2NVqtRNGZrPZ\n8i/1wgsvPPXUU8vMfOmll5Z/WQCAixdlsGu1WvPHw+HwwMipPPbYY88999zJK3bPPvvsiy++\n+K53vev+3gIA4GJEGewGg8H8cR7sFkdO5cqVK08++eTJc0ajUQjhoYfcjwgAXGrCCgBASQh2\nAAAlEeWl2EVJkhRdAgDApRB9sJtMJkWXAABwKbgUCwBQEtEHu36/32632+12/uV4PC62HgCA\nokR8KXY8HjebzfmXeceTZrOZJInrswDACop1xW6e6vImc3OtVitN036/X1BdAACFiTXYbW1t\nhRBGo1Gj0Vgcv379eghhZ2enmLIAAIoTa7BL0zSEcCDVhRCq1er8WQCAlRJrsAMA4IBYg13e\nlzjLsgPj+a5YXYsBgBUUa7BbX18PIWxsbCwOzndU5M8CAKyUWNuddDqdvb294XBYqVTykfmD\nVqvV6XSKKw0AoBixrtiFEAaDwWg0WrzqmiTJaDTKG9oBAKyaWFfsco1G4/DGWACA1RTxih0A\nAIuiDHaLB8JmWdZut+v1er1e7/f7h/fJAgCsiMiCXb/fr1Qq8yNi+/1+rVYbDodpmqZp2u12\na7Wa88QAgNUUU7Abj8fdbjeE0Gq1Fr/s9XrT6XQ6nfZ6vRBCt9tdXNIDAFgRMW2eOHA+7OHj\nYvMuJ91ud2try6YKAGDVxLRid+B82COPi82znbNiAYAVFFOwAwDgBDEFu7wX8fz+ufmddotz\nnBULAKysmILd5uZmCKHZbOY9Ta5fvx5C2Nramrc4ybLMWbEAwMqKafNEo9Ho9Xp5T5MkSa5d\nu5YkSZqm+Zfhzq11zooFAFZTTMEuhNDpdN797ndvbW3ljevm4/PHvV5PqgMAVlNkwS44HxYA\n4Bgx3WMHAMAJlg12lUqlUqmc9ikAAC7MSZdi+/3+3t7e4ki73T7negAAuE8nBbu1tbX8MNa5\n4XB45My8pRwAAAU6KdhVq9XRaJQ/zvvDzb+cq9Vq+czzKQ8AgGXdY1fsfP9pviZnOyoAwKW1\nbLuTwWBwrnUAAPCAtDsBACiJUwS7drtdOd75lQgAwDKWvRTbbrfzLbH5Ia3nWRIAAPdj2WCX\np7rRaGT/BADA5XS6e+ykOgCAS2vZYJckybnWAQDAA1o22K2vr4cQ+v3+eRYDAMD9W/Yeu7W1\ntZ2dnW63u7e39/73v//wBFdpAQCKtWywy48OCyEMh8MjT4ydzWZnVhQAAKe3bLDLjxQDAODS\ncqQYAEBJOFIMAKAkBDsAgJJY9lLsPU+DtXkCAKBYD7R5It8ea18FAMBl8ECbJwaDQbvdvnnz\n5mQyOdOqAAA4tQe9x+769etpmrbb7TOpBgCA+/agwa5arYY712QBACjQgwa7LMvOpA4AAB7Q\nsvfYjcfjI8ebzWawfwIA4BJYNtjlAe44169fP4tiAAC4fw96Vuz73//+RqNxdvUAAHCfnBUL\nAFASjhQDACiJZVfsclmWbW9v37x5M//y2rVr169fzzueAABQrFMEu36/3+12F0fSNB0Oh71e\nr9PpnHVhAACczrKXYsfjcZ7qer3edDqdzWbT6bTX64UQut3ucc1QAAC4MMsGu62trRBCvjiX\nX3utVqudTifPdh//+MfPr0QAAJaxbLBL0zSEcPiSaz7iSDEAgMLZFQsAUBLLBrskScJRB4vl\nI/mzAAAUaNlgt76+HkJoNptZls0HsyzLjxrLnwUAoEDLtjvpdDo7OztpmtZqtfn6XH7jXZIk\n2p0AABTuFPfYTSaTfA9sekcIodfrTSaT86oOAIClne7kiU6nY3EOAOBysisWAKAkHijYLW6k\nAACgWKcIdlmW1ev1SqUyH6nVapVKxXliAACXwSnOiq3VavmGiblWqxVCaDabsh0AQOFOfVbs\nbDabDw4Gg9FoNH8WAIACPehZsY1GY/4sAAAFOt3micO7JeyfAAC4JJYNdvntdBsbGwfG85H8\nWQAACrRsg+LBYDAcDtM0rVQqB44Uy589l+oAAFjaKS7FzmazfGVu8UixVqu1uJ0CAICinO5I\nscFgYHEOAOBycqQYAEBJCHYAACVRkmBXr9fr9XrRVQAAFOl099hdWjokAwDEFOzuuSa3OGEy\nmZxzOQAAl0tMwS7ca2XOuh0AsMpiusduMpnkvZF7vd5sv3zC4REAgNURU7ALIUwmk16v1+12\nbZUAADggsmAXQuh0OtPpND/cLMuyossBALgs4gt2IYRqtTqbzZIkqdVq/X6/6HIAAC6FKINd\nbjKZtFotl2UBAHIRB7sQwmAwGI1GNsMCAITo2p0c1mg07IEFAAixr9gBAOSk1nsAAB1NSURB\nVDAX5YpdlmW7u7t7e3s3b97MR65du3b16tW1tbVqtVpsbQAARYks2GVZtrGxcfimunyk2+0m\nSXLjxg3xDgBYQTEFuyzLarVaCCFJks3Nzfzx3HQ63draStO0VqtNp1PZDgBYNTEFu+3t7RBC\nq9UaDAaHn61Wq41Go91uD4fD7e3tI+cAAJRYTJsnhsNhCOHkxJY/m88EAFgpMQU7AABOEFOw\nS5IkhDAej0+Ykz+bzwQAWCkxBbvNzc0QQrPZPC7bjcfjZrM5nwkAsFJi2jzRaDR6vV63283T\nW1hYmVtsgNLr9RqNRgH1AQAUKqZgF0LodDpra2u7u7s7Oztpmi7muSRJ1tfX9SgGAFZWZMEu\nhFCtVjudTqfTKboQAIDLJaZ77AAAOEH0wa5SqVQqlaKrAAAoXvTBDgCAnGAHAFAS8W2eCCFk\nWXbCiF2xAMBqijLY1Wq1E0Zms9nyL3Xr1q3RaPTaa6+dMOfll18OIdy+ffsUJQIAXLgog12r\n1Zo/Hg6HB0ZO5YUXXnjqqaeWmfnSSy/d31sAAFyMKIPdYDCYP86D3eLIqTz22GPPPffcySt2\nzz777Isvvviud73r/t4CAOBiRBnsztCVK1eefPLJk+eMRqMQwkMP2WgCAFxqwgoAQElEv2KX\nJEnRJQAAXArRB7vJZFJ0CQAAl4JLsQAAJRF9sOv3++12u91u51+Ox+Ni6wEAKErEl2LH43Gz\n2Zx/mXc8aTabSZK4PgsArKBYV+zmqS7vRTLXarXSNO33+wXVBQBQmFiD3dbWVghhNBo1Go3F\n8evXr4cQdnZ2iikLAKA4sQa7NE1DCAdSXQihWq3OnwUAWCmxBjsAAA6INdjlfYmzLDswnu+K\n1bUYAFhBsQa79fX1EMLGxsbi4HxHRf4sAMBKibXdSafT2dvbGw6HlUolH5k/aLVanU6nuNIA\nAIoR64pdCGEwGIxGo8WrrkmSjEajvKEdAMCqiXXFLtdoNA5vjAUAWE0Rr9gBALCoJMGuXq/X\n6/WiqwAAKFLcl2LndCQGAIgp2N1zTW5xwmQyOedyAAAul5iCXbjXypx1OwBglcV0j91kMsmb\nm/R6vdl++YTDIwAAqyOmYBdCmEwmvV6v2+3aKgEAcEBkwS6E0Ol0ptNpmqaVSuXwWbEAACsr\nvmAXQqhWq7PZLEmSWq3W7/eLLgcA4FKIMtjlJpNJq9VyWRYAIBdxsAt3jou1GRYAIETX7uSw\nRqNhDywAQIh9xQ4AgDnBDgCgJAQ7AICSEOwAAEpCsAMAKAnBDgCgJAQ7AICSEOwAAEpCsAMA\nKAnBDgCgJAQ7AICSEOwAAEpCsAMAKAnBDgCgJAQ7AICSEOwAAEpCsAMAKAnBDgCgJAQ7AICS\nEOwAAEpCsAMAKAnBDgCgJAQ7AICSEOwAAEpCsAMAKAnBDgCgJAQ7AICSEOwAAEpCsAMAKAnB\nDgCgJAQ7AICSEOwAAEpCsAMAKAnBDgCgJAQ7AICSEOwAAEpCsAMAKAnBDgCgJAQ7AICSEOwA\nAEpCsAMAKAnBDgCgJAQ7AICSEOwAAEriTUUXAADL+u/f/OPK+36s6Crg8rJiBwBQEoIdAEBJ\nCHYAACUh2AEAlITNEwBE43//3nf8949/sugqKMDeyy/9Px9aL7qKCFixAwAoCcEOAKAkBDsA\ngJIQ7AAASkKwAwAoCcEOAKAkShLs6vV6vV4vugoAgCKVpI9dmqZFlwAAULCYgt091+QWJ0wm\nk3MuBwDgcokp2IV7rcxZtwMAVllM99hNJpMkSUIIvV5vtl8+4fAIAMDqiCnYhRAmk0mv1+t2\nu7ZKAAAcEFmwCyF0Op3pdJqmaaVSybKs6HIAAC6L+IJdCKFarc5msyRJarVav98vuhwAgEsh\nymCXm0wmrVbLZVkAgFzEwS6EMBgMRqORzbAAACG6dieHNRoNe2ABAELsK3YAAMxFuWKXZdnu\n7u7e3t7NmzfzkWvXrl29enVtba1arRZbGwBAUSILdlmWbWxsHL6pLh/pdrtJkty4cUO8AwBW\nUEzBLsuyWq0WQkiSZHNzM388N51Ot7a20jSt1WrT6VS2AwBWTUzBbnt7O4TQarUGg8HhZ6vV\naqPRaLfbw+Fwe3v7yDkAACUW0+aJ4XAYQjg5seXP5jMBAFZKTMEOAIATxBTskiQJIYzH4xPm\n5M/mMwEAVkpMwW5zczOE0Gw2j8t24/G42WzOZwIArJSYNk80Go1er9ftdvP0FhZW5hYboPR6\nvUajUUB9AACFiinYhRA6nc7a2tru7u7Ozk6apot5LkmS9fV1PYoBgJUVWbALIVSr1U6n0+l0\nii4EAOByiekeOwAAThB9sKtUKpVKpegqAACKF32wAwAgJ9gBAJREfJsnQghZlp0wYlcsALCa\nogx2tVrthJHZbLb8S926dWs0Gr322msnzHn55ZdDCLdv3z5FiQAAFy7KYNdqteaPh8PhgZFT\neeGFF5566qllZr700kv39xYAABcjymA3GAzmj/NgtzhyKo899thzzz138ords88+++KLL77r\nXe+6v7cAALgYUQa7M3TlypUnn3zy5Dmj0SiE8NBDNpoAAJeasAIAUBLRr9glSVJ0CQAAl0L0\nwW4ymRRdAgDApeBSLABASUQf7Pr9frvdbrfb+Zfj8bjYegAAihLxpdjxeNxsNudf5h1Pms1m\nkiSuzwIAKyjWFbt5qst7kcy1Wq00Tfv9fkF1AQAUJtZgt7W1FUIYjUaNRmNx/Pr16yGEnZ2d\nYsoCAChOrMEuTdMQwoFUF0KoVqvzZwEAVkqswQ4AgANiDXZ5X+Isyw6M57tidS0GAFZQrMFu\nfX09hLCxsbE4ON9RkT8LALBSYm130ul09vb2hsNhpVLJR+YPWq1Wp9MprjQAgGLEumIXQhgM\nBqPRaPGqa5Iko9Eob2gHALBqYl2xyzUajcMbYwEAVlPEK3YAACyKMtiNx+MTzoRdPDoWAGB1\nRBbsxuNxpVJpNpvNZrNSqdTr9cMdT4bD4XA4LKQ8AIACxXSP3bybSZIk165du3nzZpqmtVpt\nOp3mB04AAKyymFbsPv7xj4cQer3eZDIZDAaTyWQ0GoUQarXa4XU7AIBVE1Owyy+wrq2tzUca\njcY82xVWFgDA5RBTsMsduOraaDR6vV4IoV6vF1QRAMClEFOwO+582E6n02q10jTt9/tF1AUA\ncCnEFOyuXbsWQtjd3T381GAwSJKk2+3KdgDAyoop2F2/fj2E0O1258fCLppMJnm2u/C6AAAu\nhZiCXbVanU6ni4fDHjCZTFqt1kWWBABwecQU7EII1Wp1MpnMZrPjJgwGg9lsdsIEAICyiizY\nAQBwHMEOAKAkBDsAgJIQ7AAASkKwAwAoCcEOAKAkBDsAgJIQ7AAASkKwAwAoCcEOAKAkBDsA\ngJIQ7AAASkKwAwAoCcEOAKAkBDsAgJIQ7AAASkKwAwAoCcEOAKAkBDsAgJIQ7AAASkKwAwAo\nCcEOAKAkBDsAgJIQ7AAASkKwAwAoCcEOAKAkBDsAgJIQ7AAASkKwAwAoCcEOAKAkBDsAgJIQ\n7AAASkKwAwAoCcEOAKAkBDsAgJIQ7AAASkKwAwAoCcEOAKAkBDsAgJIQ7AAASkKwAwAoCcEO\nAKAkBDsAgJIQ7AAASkKwAwAoCcEOAKAkBDsAgJIQ7AAASkKwAwAoCcEOAKAkBDsAgJIQ7AAA\nSkKwAwAoCcEOAKAkBDsAgJIQ7AAASuJNRRdwP7Is293d3dvbu3nzZj5y7dq1q1evrq2tVavV\nYmsDAChKZMEuy7KNjY00TQ+M5yPdbjdJkhs3boh3AMAKiinYZVlWq9VCCEmSbG5u5o/nptPp\n1tZWmqa1Wm06ncp2AMCqiSnYbW9vhxBardZgMDj8bLVabTQa7XZ7OBxub28fOQcAoMRi2jwx\nHA5DCCcntvzZfCYAwEqJKdgBAHCCmIJdkiQhhPF4fMKc/Nl8JgDASokp2G1uboYQms3mcdlu\nPB43m835TACAlRLT5olGo9Hr9brdbp7ewsLK3GIDlF6v12g0CqgPAKBQMQW7EEKn01lbW9vd\n3d3Z2UnTdDHPJUmyvr6uRzFwTr773e/+t//230IIf+Ev/IU3vSmy/3gCKyKmS7G5arXa6XQm\nk8lsv8lk0ul0pDrgzL388ssf+tCH3vnOdz766KOPPvroO9/5zg996EMvv/xy0XWxWm7dvv3J\nz3321u3bRRfCpeZ/OgFO8tnPfvYnf/InX3nllfkq3auvvnrjxo3/9J/+0/PPP//jP/7jxZa3\naqZf/cpf+sD7i66iGK9+5ztf+8bXv/9/+/OPfM/3FF1LAf709deLLiEOUQY7Z8UCF+OVV155\n4okn/uRP/iSE8D3f8z2VSiWE8Gf+zJ/57ne/+yd/8idPPPHEH/7hH77tbW8rusxV8VM/9VPP\nP/980VUU5ntns7e+8x1/7s/9ufxzuIK+79H/+2/+zb9ZdBWXXWTBzlmxwEV69tlnv/GNb+R/\nj37729+ezWYhhO9+97v5s9/4xjf+9b/+1z//8z9fZImr5CMf+chHPvKRoquASy2mYOesWOCC\n/dqv/VoIIc9zf+2v/bV8J36apr/7u7+bD964cUOw42LcunXr05/+9OOPP37lypWia+HyiinY\nOSt2lX3lK1/56Ec/+vzzz3/1q19929ve9jf+xt9ot9t/62/9raLrouTyHRJvfvOb/+N//I9P\nPPHEfPy3fuu3fuZnfub111+3hYILMx6Pn3zyyU984hOLH0U4IKZdsc6KXVmf+MQnrl279i/+\nxb/4+te//p73vOeRRx759//+37/vfe/78Ic/nK+awDn5zne+E0L4pV/6pQN/lT7xxBO/9Eu/\nNJ8AF+DVV1+d/xOOE9OKHavpC1/4wvr6+lve8pYf/dEf/f3f//2vfvWrIYR3v/vdlUrlox/9\n6Pd93/f903/6T4uukZL7q3/1rx4e/JEf+ZGLrwTgZDGt2DkrdjX9wi/8wp/+6Z9+85vf/Pzn\nP//YY499+MMf/sAHPvDNb37zi1/84lvf+tZf/MVf/OY3v1l0jZTWQw89FEL4B//gH3zrW99a\nHP/Wt77VarXmEwAuiZj+k+Ss2BX0ne98ZzQa3b59+9q1a3/wB38wGo1+5Vd+5dd//de//OUv\n/8N/+A//1//6X9/+9rc/9alPFV0mpfW93/u9IYQ/+IM/+KEf+qF+v/+5z33uc5/7XL/fzz+Q\nIYR3vOMdRdfIqnjkkUfm/4TjxHQp1lmxK+hrX/va66+//uY3v3l3d/cHfuAH5uMPP/zwRz/6\n0fxv2c9//vNPP/10gUVyYW7dujUajV577bULe8darfb1r389hPDVr3612+3Ox+cLddVq9Td/\n8zcvrJ6HH3642WzaFLmaGo3G888///jjjxddCJdaTMEuOCt29bz++ushhL/4F//iYqrLVSqV\nn/qpn/rc5z73h3/4h0WURgFeeOGFp556qpC3vr3/HKf5l7/zO7/zO7/zOxdZyac+9am//bf/\n9kW+I5fElStXfuInfqLoKrjsIgt24c5ZsZ1Op+hCVtHFr5d87WtfCyF8/etfzxdFbt++/YUv\nfOGHfuiH8vWS/C/Ur33ta5ZMVsRjjz323HPPXeQnMISws7PzH/7DfzjyqaeffvqCV4sffvjh\nxx577CLfkctDHzuWEV+wo0BFrZf88R//8fr6+nHPvvDCCy+88MJF1mPJpChXrlx58sknL/hN\nn3766Weeeeaf/JN/8sorr8wH3/a2t/3iL/7iP/pH/+iCi2GV6WPHMqIPdvlRP5qZXYyLXy+Z\nzWbtdjs/qfOnf/qn3/GOd/ybf/NvPvzhD7/pTW/62Mc+9j/+x/8IIfzzf/7P3/Oe91xYSSu+\nZJJl2YH9oavgx3/8x3/rt37rs5/97Kc+9alPfepTf+/v/b1Wq/WWt7zld3/3d4su7aK9/e1v\nd7tLUfSxYxnRB7tV85WvfOXv/t2/+9u//dtFF1KA+eWwf/Wv/tXi+C/8wi8UUU4x3vve9/6z\nf/bPilov/KM/+qN3v/vd/j/qYx/72Mc+9rGiqyhGpVL5r//1vx6+5xW4JAS7yPzqr/7qaqY6\ncp/5zGd+7ud+7vd+7/cKefdXXnlFqltxs9ls8ZL0Srn4m4wPyO8qvuDNOge4yfjyizLYZVl2\nwsipLhMs8y9qfhbkgT1xRbl161bRJVCot/gMQDEK3JS96Jd/+Zd/+Zd/ucAC3GR8yUUZ7Gq1\n2gkjp1pRWP5f1D/6oz9a/mXPz1vf+taiS6BQ3w5/9s/+2aLe/OGHHy7qrbk8VvZjUMim7EUH\n2gIUYsVvMo5ClMEuP8knNxwOD4ycyjL/ou7u7t64ceMDH/jA/b3F2frZn/3ZL3zhC7/2a79W\ndCEU44Mf/OBHPvKRot79B3/wBz/4wQ/+u3/374oqgMJ98IMf/MEf/MGiqyhGIZuyD/g7f+fv\nFFsAl18l9jtmLmBXbL/f73a7n/nMZ+r1+vm9CwAQhclk8t73vrfX613CrroxnRULAMAJBDsA\ngJKI8h67RUmSFF0CAMClEH2wm0wmRZcAAHApuBQLAFAS0Qe7fr/fbrfb7Xb+5Xg8LrYeAICi\nRHwpdjweN5vN+ZeDwSCE0Gw2kyRxfRYAWEGxrtjNU91oNFocb7VaaZr2+/2C6gIAKEyswW5r\nayuEMBqNGo3G4vj169dDCDs7O8WUBQBQnFiDXZqmIYQDqS6EUK1W588CAKyUWIMdAAAHxBrs\n8r7EWZYdGM93xepaDACsoFiD3fr6eghhY2NjcXC+oyJ/FgBgpcTa7qTT6ezt7Q2Hw0qlko/M\nH7RarU6nU1xpAADFiHXFLoQwGAxGo9HiVdckSUajUd7QDgBg1cS6YpdrNBqHN8YCAKymiFfs\nAABYFOWKXZZlu7u7e3t7N2/ezEeuXbt29erVtbW1vI8dAMAKiizYZVm2sbFxuP9wPtLtdpMk\nuXHjhngHAKygmIJdlmW1Wi2EkCTJ5uZm/nhuOp1ubW2laVqr1abT6Zlnuy9+8YsPP/zw2b4m\np/X666//6q/+6qOPPvrQQ+4ioAC3b9/OsqxarfoEcvFu37795S9/+UMf+tCb3/zmomtZdV/8\n4heLLuF4s3i0Wq0QQqvVesA5p/XMM88U/VsCAC6XZ5555gzDxlmpzGazon8yy8o71d2z4CWn\nLe9b3/rWjRs3Xn311bN6Qe7b5z//+d/4jd9473vf++ijjxZdC6voy1/+8mc+8xmfQAqRf/w+\n8IEP/PAP/3DRtRAeeeSRjY2Nt7/97UUXcpBgR0x+8zd/c319fWdn5+mnny66FlaRTyAF8vFj\nGTHdJpL3Is5Pgz2Os2IBgJUVU7Db3NwMITSbzeOy3fys2HwmAMBKiWlXbKPR6PV63W43T29h\nYWVusQFKr9dzHAUAsIJiCnYhhE6ns7a2tru7u7Ozk6bpYp5LkmR9fV2PYgBgZUUW7EII1Wq1\n0+l0Op2iCwEAuFxiuscOAIATCHYAACUh2AEAlIRgBwBQEoIdMXnkkUfm/4SL5xNIgXz8WEZM\nR4rBrVu3Pv3pTz/++ONXrlwpuhZWkU8gBfLxYxmCHQBASbgUCwBQEoIdAEBJCHYAACUh2AEA\nlIRgBwBQEoIdAEBJCHYAACUh2AEAlIRgBwBQEoIdAEBJCHYAACUh2AEAlIRgBwBQEoIdAEBJ\nCHYAACUh2AEAlIRgR3z6/X6lUim6ClZIlmXtdrtyR71eH4/HRRfFCsn/ozf/+GVZVnRFXF6V\n2WxWdA1wCuPxuNlshhB8dLkYWZbVarXD461WazAYXHw9rJp6vZ6m6YHB0WjUaDQKqYdLzood\nMen3+3mqgwuzvb0dQmi1WtPpdDabzWaz0WgUQhgOh9btOG/j8ThN0yRJ5h+/Xq8XQvBfQo5j\nxY44ZFm2sbGR/wcuhJCmqY8uF6NSqSRJMplMFgf7/X6327Vox3nLl+um02m1Wp0Pttvt4XB4\nYBByVuyIw/b2dpqmvV7vwN+vcN5ms9nhT93a2lohxbBqJpPJbDY7EOBu3rxZVD1cfm8qugBY\nyvXr1y2NcHns7u6GEK5evVp0Iaycdrudpmmr1bJcx5FciiU++bUJH12KMt9O4UPIhVncxNPr\n9TqdTrH1cGm5FAtwCvO/X/MtFHBhkiTJbzLudrvtdrvocrikXIoFWNa8245mE1ywarU6v9cz\n3zwRQnCDCoe5FEt8XIqlEPO/TaU6Cpc3afefQQ5zKRbg3ur1+nA4zNuJSXUULr8mC4cJdgD3\nkC8St1qtyWRiKyIXKT9G7MBglmXzpp5wgGAHcJJ+v5/3UHQ/Exev1WqFEBbPhx2Px/n2nfX1\n9SIr47Jyjx3xcY8dF+nwesnc4RMp4MwdeVasU084jhU7gGM5DZbCTSaT/HzYXJIko9FIquM4\nVuwAAErCih0AQEkIdgAAJSHYAQCUhGAHAFASgh0AQEkIdgAAJSHYAQCUhGAHAFASgh0AQEkI\ndgAAJSHYAQCUhGAHAFASgh0AQEkIdgAAJSHYAQCUhGAHAFASgh0AQEkIdgAAJSHYAQCUhGAH\nAFASgh0AQEkIdgAAJSHYAQCUhGAHAFASgh0AQEkIdgAAJSHYAQCUhGAHAFASgh0AQEkIdgAA\nJSHYAQCUhGAHAFASgh1QjHG7UqnU+9l5v0/Wr1cqlfb47sC437/71UWVcfB9z/dPAatKsANW\nSdav15rdvUjet6hqgWgJdgAAJSHYAQCUhGAHFCzrt+uVSqVSqdTb/fHBm92y8fzpSqVebx+a\nkM+oz+ccMyvkd9vVumkIYdg8fGPdyWXc+7sYtxdq2PcSx77viZUf+aeOuCPwiHsIj60EKD3B\nDijS3na91h2m+RfpsNvcWEwt+T1md54OIU2HzVp9IcTMZ6Tp3bF81qn2Q5xcxj1l/XqtOVyo\nIR1271XCGVV+BpUAJSLYAQVKh8PQG01ns9lsNu0lIYR0Z/dOCBm3a900JK3RNJ8wm01HrSSk\nw+Y82o23u2kIyZ2XeGNO2P86d1Q7kzfepDWazWaTTnWpMu7tThXTuzX0kruvceT73rPy46u9\n/0qA0hPsgCIlvRudRh5Zqp3NVggh3ZvmT40/Pgwh6d0YNKp3Mk21MZiMWiEMP/5GsmsMZrPZ\nbHLnJfI513vJGZZxCtM76ana6EzuEcbOqvIHrwQoE8EOKFCyvraQOGpXF3JN9qWbIYS0W6vs\n1xyGEG5+af8KVJZl4/F43G+32/V6fmfaGZWxjMb7WyGEtNusvXGjXH+cLbtC9oCVn2ElQBm8\nqegCAI403Tsh5KR70xCqIYSQ9dsbC3fhFaMxmI7CRjMvI02HaXPYDSFpjSaDxrF/6Fwqv69K\ngPKwYgdcTvm6WWs0O9IbMWXczvc8JK1Wq9cbjUaj6fSNO9MuVrUxmMxms+l01Gu1kvz9F28G\nPOTcKj91JUCZWLEDLqfqu6+FkA4/Ph40jl1reuM2vOm+W8iy/k5hC3jVaqMzaHTCIGT9dq07\nvPmlLDSOur3tQSq/u1wZQsh2j/4zS1cClIoVO+CSyu8XGzYXu9Jl4369sr9tW0h3du82fxu3\nl7pR7UtneeNZ3kluf/O8bLo3vOf7Llv53T+VL2MOt95oX5KN+xuLf+YUlQAldfRVDoBzNmqF\nfY05ZrPF7h535hx5afLujLxByNHemLX/RRcudubvvUwZy30vhy22HTnyfU+u/Ig/dfhibdJq\nLVZ770qAUrNiB1xejcFkOrpzo1gIISRJazSd3d0H0Bgcer43mubp5+DO2RBCCNXOjd4bafF+\nGpocW+dsfx0hafVGi9dZD73vUpUfrrbamUzvDIWk1ZtOrl89XSVAqVVms1nRNQAAcAas2AEA\nlIRgB3CMcbuyBI1EgMtDsAMAKAn32AEAlIQVOwCAkhDsAABKQrADACgJwQ4AoCQEOwCAkhDs\nAABKQrADACgJwQ4AoCQEOwCAkhDsAABKQrADACgJwQ4AoCQEOwCAkhDsAABKQrADACgJwQ4A\noCQEOwCAkhDsAABKQrADACgJwQ4AoCQEOwCAkhDsAABKQrADACgJwQ4AoCT+f6JPM2Cvc2Ha\nAAAAAElFTkSuQmCC",
      "text/plain": [
       "Plot with title \"Taxa 2 - Abundance\""
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "boxplot(count ~ health_status, data=taxa2, ylim = c(0, 1e+5), col=c(\"green\", 'red', 'pink'),\n",
    "       main=\"Taxa 2 - Abundance\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "b062803f-60f9-44a5-8ed6-1537bec91276",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 143</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>S1</th><th scope=col>S3</th><th scope=col>S4</th><th scope=col>S5</th><th scope=col>S7</th><th scope=col>S8</th><th scope=col>S9</th><th scope=col>S10</th><th scope=col>S11</th><th scope=col>S12</th><th scope=col>⋯</th><th scope=col>S141</th><th scope=col>S142</th><th scope=col>S143</th><th scope=col>S144</th><th scope=col>S145</th><th scope=col>S146</th><th scope=col>S147</th><th scope=col>S148</th><th scope=col>S149</th><th scope=col>S150</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>T1</th><td>-0.1078375</td><td>-0.624916</td><td>-0.7285940</td><td>-0.24781539</td><td>-0.33746108</td><td>       NA</td><td>      NA</td><td>-0.4915143</td><td>       NA</td><td>-0.6883827</td><td>⋯</td><td>-0.4340369</td><td>-0.6018567</td><td>-0.4951947</td><td>-0.8182918</td><td>-0.8200758</td><td>-0.8071122</td><td>-0.9205346</td><td>-0.7530298</td><td>-0.3608264</td><td>-0.7311376</td></tr>\n",
       "\t<tr><th scope=row>T2</th><td> 0.3049522</td><td> 0.212757</td><td> 0.2428773</td><td> 0.02863789</td><td> 0.04547567</td><td>0.2892208</td><td>0.385502</td><td> 0.1891549</td><td>0.2956597</td><td> 0.1870306</td><td>⋯</td><td> 1.0327260</td><td> 1.3386929</td><td> 1.0992331</td><td> 1.4476971</td><td> 1.4476542</td><td> 1.4258788</td><td> 1.5716465</td><td> 1.4003771</td><td> 0.9382495</td><td> 1.3390795</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 143\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & S1 & S3 & S4 & S5 & S7 & S8 & S9 & S10 & S11 & S12 & ⋯ & S141 & S142 & S143 & S144 & S145 & S146 & S147 & S148 & S149 & S150\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\tT1 & -0.1078375 & -0.624916 & -0.7285940 & -0.24781539 & -0.33746108 &        NA &       NA & -0.4915143 &        NA & -0.6883827 & ⋯ & -0.4340369 & -0.6018567 & -0.4951947 & -0.8182918 & -0.8200758 & -0.8071122 & -0.9205346 & -0.7530298 & -0.3608264 & -0.7311376\\\\\n",
       "\tT2 &  0.3049522 &  0.212757 &  0.2428773 &  0.02863789 &  0.04547567 & 0.2892208 & 0.385502 &  0.1891549 & 0.2956597 &  0.1870306 & ⋯ &  1.0327260 &  1.3386929 &  1.0992331 &  1.4476971 &  1.4476542 &  1.4258788 &  1.5716465 &  1.4003771 &  0.9382495 &  1.3390795\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 143\n",
       "\n",
       "| <!--/--> | S1 &lt;dbl&gt; | S3 &lt;dbl&gt; | S4 &lt;dbl&gt; | S5 &lt;dbl&gt; | S7 &lt;dbl&gt; | S8 &lt;dbl&gt; | S9 &lt;dbl&gt; | S10 &lt;dbl&gt; | S11 &lt;dbl&gt; | S12 &lt;dbl&gt; | ⋯ ⋯ | S141 &lt;dbl&gt; | S142 &lt;dbl&gt; | S143 &lt;dbl&gt; | S144 &lt;dbl&gt; | S145 &lt;dbl&gt; | S146 &lt;dbl&gt; | S147 &lt;dbl&gt; | S148 &lt;dbl&gt; | S149 &lt;dbl&gt; | S150 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| T1 | -0.1078375 | -0.624916 | -0.7285940 | -0.24781539 | -0.33746108 |        NA |       NA | -0.4915143 |        NA | -0.6883827 | ⋯ | -0.4340369 | -0.6018567 | -0.4951947 | -0.8182918 | -0.8200758 | -0.8071122 | -0.9205346 | -0.7530298 | -0.3608264 | -0.7311376 |\n",
       "| T2 |  0.3049522 |  0.212757 |  0.2428773 |  0.02863789 |  0.04547567 | 0.2892208 | 0.385502 |  0.1891549 | 0.2956597 |  0.1870306 | ⋯ |  1.0327260 |  1.3386929 |  1.0992331 |  1.4476971 |  1.4476542 |  1.4258788 |  1.5716465 |  1.4003771 |  0.9382495 |  1.3390795 |\n",
       "\n"
      ],
      "text/plain": [
       "   S1         S3        S4         S5          S7          S8        S9      \n",
       "T1 -0.1078375 -0.624916 -0.7285940 -0.24781539 -0.33746108        NA       NA\n",
       "T2  0.3049522  0.212757  0.2428773  0.02863789  0.04547567 0.2892208 0.385502\n",
       "   S10        S11       S12        ⋯ S141       S142       S143      \n",
       "T1 -0.4915143        NA -0.6883827 ⋯ -0.4340369 -0.6018567 -0.4951947\n",
       "T2  0.1891549 0.2956597  0.1870306 ⋯  1.0327260  1.3386929  1.0992331\n",
       "   S144       S145       S146       S147       S148       S149       S150      \n",
       "T1 -0.8182918 -0.8200758 -0.8071122 -0.9205346 -0.7530298 -0.3608264 -0.7311376\n",
       "T2  1.4476971  1.4476542  1.4258788  1.5716465  1.4003771  0.9382495  1.3390795"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "log_value = output$bias_correct_log_table\n",
    "log_value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "a035ca96-7fc2-4207-a69d-4893b6e73f4b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 1 × 143</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>S1</th><th scope=col>S3</th><th scope=col>S4</th><th scope=col>S5</th><th scope=col>S7</th><th scope=col>S8</th><th scope=col>S9</th><th scope=col>S10</th><th scope=col>S11</th><th scope=col>S12</th><th scope=col>⋯</th><th scope=col>S141</th><th scope=col>S142</th><th scope=col>S143</th><th scope=col>S144</th><th scope=col>S145</th><th scope=col>S146</th><th scope=col>S147</th><th scope=col>S148</th><th scope=col>S149</th><th scope=col>S150</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>T2</th><td>0.3049522</td><td>0.212757</td><td>0.2428773</td><td>0.02863789</td><td>0.04547567</td><td>0.2892208</td><td>0.385502</td><td>0.1891549</td><td>0.2956597</td><td>0.1870306</td><td>⋯</td><td>1.032726</td><td>1.338693</td><td>1.099233</td><td>1.447697</td><td>1.447654</td><td>1.425879</td><td>1.571646</td><td>1.400377</td><td>0.9382495</td><td>1.33908</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 1 × 143\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & S1 & S3 & S4 & S5 & S7 & S8 & S9 & S10 & S11 & S12 & ⋯ & S141 & S142 & S143 & S144 & S145 & S146 & S147 & S148 & S149 & S150\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\tT2 & 0.3049522 & 0.212757 & 0.2428773 & 0.02863789 & 0.04547567 & 0.2892208 & 0.385502 & 0.1891549 & 0.2956597 & 0.1870306 & ⋯ & 1.032726 & 1.338693 & 1.099233 & 1.447697 & 1.447654 & 1.425879 & 1.571646 & 1.400377 & 0.9382495 & 1.33908\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 1 × 143\n",
       "\n",
       "| <!--/--> | S1 &lt;dbl&gt; | S3 &lt;dbl&gt; | S4 &lt;dbl&gt; | S5 &lt;dbl&gt; | S7 &lt;dbl&gt; | S8 &lt;dbl&gt; | S9 &lt;dbl&gt; | S10 &lt;dbl&gt; | S11 &lt;dbl&gt; | S12 &lt;dbl&gt; | ⋯ ⋯ | S141 &lt;dbl&gt; | S142 &lt;dbl&gt; | S143 &lt;dbl&gt; | S144 &lt;dbl&gt; | S145 &lt;dbl&gt; | S146 &lt;dbl&gt; | S147 &lt;dbl&gt; | S148 &lt;dbl&gt; | S149 &lt;dbl&gt; | S150 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| T2 | 0.3049522 | 0.212757 | 0.2428773 | 0.02863789 | 0.04547567 | 0.2892208 | 0.385502 | 0.1891549 | 0.2956597 | 0.1870306 | ⋯ | 1.032726 | 1.338693 | 1.099233 | 1.447697 | 1.447654 | 1.425879 | 1.571646 | 1.400377 | 0.9382495 | 1.33908 |\n",
       "\n"
      ],
      "text/plain": [
       "   S1        S3       S4        S5         S7         S8        S9      \n",
       "T2 0.3049522 0.212757 0.2428773 0.02863789 0.04547567 0.2892208 0.385502\n",
       "   S10       S11       S12       ⋯ S141     S142     S143     S144     S145    \n",
       "T2 0.1891549 0.2956597 0.1870306 ⋯ 1.032726 1.338693 1.099233 1.447697 1.447654\n",
       "   S146     S147     S148     S149      S150   \n",
       "T2 1.425879 1.571646 1.400377 0.9382495 1.33908"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "4e4cbb76-34a6-46e0-930a-f88a3d2953be",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>1</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>2</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li><li>3</li></ol>\n",
       "\n",
       "<details>\n",
       "\t<summary style=display:list-item;cursor:pointer>\n",
       "\t\t<strong>Levels</strong>:\n",
       "\t</summary>\n",
       "\t<style>\n",
       "\t.list-inline {list-style: none; margin:0; padding: 0}\n",
       "\t.list-inline>li {display: inline-block}\n",
       "\t.list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "\t</style>\n",
       "\t<ol class=list-inline><li>'1'</li><li>'2'</li><li>'3'</li></ol>\n",
       "</details>"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\emph{Levels}: \\begin{enumerate*}\n",
       "\\item '1'\n",
       "\\item '2'\n",
       "\\item '3'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 1\n",
       "2. 1\n",
       "3. 1\n",
       "4. 1\n",
       "5. 1\n",
       "6. 1\n",
       "7. 1\n",
       "8. 1\n",
       "9. 1\n",
       "10. 1\n",
       "11. 1\n",
       "12. 1\n",
       "13. 1\n",
       "14. 1\n",
       "15. 1\n",
       "16. 1\n",
       "17. 1\n",
       "18. 1\n",
       "19. 1\n",
       "20. 1\n",
       "21. 1\n",
       "22. 1\n",
       "23. 1\n",
       "24. 1\n",
       "25. 1\n",
       "26. 1\n",
       "27. 1\n",
       "28. 1\n",
       "29. 1\n",
       "30. 1\n",
       "31. 1\n",
       "32. 1\n",
       "33. 1\n",
       "34. 1\n",
       "35. 1\n",
       "36. 1\n",
       "37. 1\n",
       "38. 1\n",
       "39. 1\n",
       "40. 1\n",
       "41. 1\n",
       "42. 1\n",
       "43. 1\n",
       "44. 1\n",
       "45. 1\n",
       "46. 1\n",
       "47. 1\n",
       "48. 1\n",
       "49. 1\n",
       "50. 1\n",
       "51. 2\n",
       "52. 2\n",
       "53. 2\n",
       "54. 2\n",
       "55. 2\n",
       "56. 2\n",
       "57. 2\n",
       "58. 2\n",
       "59. 2\n",
       "60. 2\n",
       "61. 2\n",
       "62. 2\n",
       "63. 2\n",
       "64. 2\n",
       "65. 2\n",
       "66. 2\n",
       "67. 2\n",
       "68. 2\n",
       "69. 2\n",
       "70. 2\n",
       "71. 2\n",
       "72. 2\n",
       "73. 2\n",
       "74. 2\n",
       "75. 2\n",
       "76. 2\n",
       "77. 2\n",
       "78. 2\n",
       "79. 2\n",
       "80. 2\n",
       "81. 2\n",
       "82. 2\n",
       "83. 2\n",
       "84. 2\n",
       "85. 2\n",
       "86. 2\n",
       "87. 2\n",
       "88. 2\n",
       "89. 2\n",
       "90. 2\n",
       "91. 2\n",
       "92. 2\n",
       "93. 2\n",
       "94. 2\n",
       "95. 2\n",
       "96. 2\n",
       "97. 2\n",
       "98. 2\n",
       "99. 2\n",
       "100. 2\n",
       "101. 3\n",
       "102. 3\n",
       "103. 3\n",
       "104. 3\n",
       "105. 3\n",
       "106. 3\n",
       "107. 3\n",
       "108. 3\n",
       "109. 3\n",
       "110. 3\n",
       "111. 3\n",
       "112. 3\n",
       "113. 3\n",
       "114. 3\n",
       "115. 3\n",
       "116. 3\n",
       "117. 3\n",
       "118. 3\n",
       "119. 3\n",
       "120. 3\n",
       "121. 3\n",
       "122. 3\n",
       "123. 3\n",
       "124. 3\n",
       "125. 3\n",
       "126. 3\n",
       "127. 3\n",
       "128. 3\n",
       "129. 3\n",
       "130. 3\n",
       "131. 3\n",
       "132. 3\n",
       "133. 3\n",
       "134. 3\n",
       "135. 3\n",
       "136. 3\n",
       "137. 3\n",
       "138. 3\n",
       "139. 3\n",
       "140. 3\n",
       "141. 3\n",
       "142. 3\n",
       "143. 3\n",
       "144. 3\n",
       "145. 3\n",
       "146. 3\n",
       "147. 3\n",
       "148. 3\n",
       "149. 3\n",
       "150. 3\n",
       "\n",
       "\n",
       "\n",
       "**Levels**: 1. '1'\n",
       "2. '2'\n",
       "3. '3'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  [1] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1\n",
       " [38] 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2\n",
       " [75] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 3 3 3 3 3 3 3 3 3 3 3\n",
       "[112] 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3\n",
       "[149] 3 3\n",
       "Levels: 1 2 3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "smd[,'health_status']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "83199c59-f14b-40bc-ab37-bed3ac8bb5c4",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cbind(...):\n",
      "\"number of rows of result is not a multiple of vector length (arg 2)\"\n"
     ]
    }
   ],
   "source": [
    "taxa2 = cbind(t(log_value['T2',]), smd$health_status)\n",
    "colnames(taxa2) = c('log_value', 'health_status')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "ca660392-4245-485a-be0c-30373fedc407",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 2 × 2 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>log_value</th><th scope=col>health_status</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>S1</th><td>0.3049522</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>S3</th><td>0.2127570</td><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 2 × 2 of type dbl\n",
       "\\begin{tabular}{r|ll}\n",
       "  & log\\_value & health\\_status\\\\\n",
       "\\hline\n",
       "\tS1 & 0.3049522 & 1\\\\\n",
       "\tS3 & 0.2127570 & 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 2 × 2 of type dbl\n",
       "\n",
       "| <!--/--> | log_value | health_status |\n",
       "|---|---|---|\n",
       "| S1 | 0.3049522 | 1 |\n",
       "| S3 | 0.2127570 | 1 |\n",
       "\n"
      ],
      "text/plain": [
       "   log_value health_status\n",
       "S1 0.3049522 1            \n",
       "S3 0.2127570 1            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(taxa2,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "26cbeede-6265-45e6-88ee-94aae1b29060",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAIAAAByhViMAAAABmJLR0QA/wD/AP+gvaeTAAAg\nAElEQVR4nOzdf3TV9Z0n/s8lMoaKa7S6BYu1dhLtMmBn1o6VG8+MGVBJEMRtga2dGbr1eGPn\n1Cauh9l21VOsjnumzAzJdseatHbK7IxVdB0W5F6BKsz3TC52Tj3TEUqryak6VNBaq61ugWJy\nv39cCReSm9wIyedz33k8Tk8OfD4f7n3mR+Hp+/P5vD6pQqEQAQBQ/abEHQAAgJNDsQMACIRi\nBwAQCMUOACAQih0AQCAUOwCAQCh2AACBUOwAAAKh2AEABEKxAwAIhGIHABAIxQ4AIBCKHQBA\nIBQ7AIBAKHYAAIFQ7AAAAqHYAQAEQrEDAAiEYgcAEAjFDgAgEIodAEAgFDsAgEAodgAAgVDs\nAAACodgBAARCsQMACIRiBwAQCMUOACAQih0AQCAUOwCAQCh2AACBUOwAAAKh2AEABEKxAwAI\nhGIHABAIxQ4AIBCKHQBAIBQ7AIBAKHYAAIFQ7CBAfZ2Nqco0dvbFHba8vr5ca2Pj0U+lsbG1\nMzeueY/7wrXmRjuqzBExy7VWxfcXGAeKHZBIfZ2NDQ0t3fl8fnBTPt/d3tLQOH5lqm/z+nzp\n77s3JLK3AZSl2AEJlGttaM8Pvyvf3TJO61DH97oo6r7bihdQVRQ7CFB9W0+hVDZzZE+6o/eY\nPT1t9XEGLaOv8+7uI78eTNzbkT6yLd++ZhyW0o72ukzmyBcsv36zZgdUEcUOiKIo6st1tjaW\nXmDW2NjYeuwFbUev3Dr22rJy2yt5zeGzHF05S3esO1I969vWHa12u5876X3r6LumZ69aOoZm\n15c7+lkO8wmWu+Bt+Cv1Bo9u7Ow79gtY7mtX+vapUb6+lX1Hxp7hncshS1+33OWQxx44/pdN\nwqRTAIJXfsWuqGQx7HiZ7LCvc3RH6Z8tPbjS1xwuTjaTTqfT6SFxR/s8TsTRvOmO3pHfquRT\ny2QyQz/NdJkv2rGvVPoq2aFHp4d75SFhssMddHTB8ZjjK/6OjDHD8CHeed1jDi2XID3aTwRQ\nMcUOJoGRC9FxlaZQKPRmy7S10mqX7uitpNaN/poVG74KnRxDAo/wNRuhIQ2T710WuwpeefSD\nS95yDN+RMWUY5YtR8jmP+LInv6nDJOVULEx2uTXv3KZw9LRnffOi5cOf9WxeNfjveL59TWfn\nysFbHDLZruZ395oV6it9s6XNIx479hc/eh52+aL6KIqi5krPxqY7skeK0tHmcnLuujjy0qXl\n6eiduqVXIg6ujR3T1kq86+/IyBmOvm6ZDIOXQw533WTJgeNz2SRMRhPfJYGJ9i5OYZb/I8Mu\n0VS0gnYiZ1JLl3vG8zxsyWdSNm7pV6DsKtzgjne9Yld6eMnmYc6Blz1jPdqXavhslWcY8ZNL\nZzo6sr1Db3055mdlPFdhYVKyYgeU6Ovry3V2tjY2tnSXO+SYmxiKjlmtezevOUqoXGuqpWS5\nZ93JvpW35G6NkrXAStbsjqzvvaP+wjlHfpnf03tioY556YbZQ9p07578cEcOf/CxKv6OjJah\n77ndR34558JjvhBtPYWerra25vr649Meu9paf3TVcBxuiIFJSLED3rlbMpVKpRoaGlra27tL\npgIP47hql+5YNVytG9trlo/W2dhQ2up6R5/QMsyDN0Ycajx8r6uo2R1bZyroVBOhpF4e4yR9\nR0qUlMvZDSMdWNIAu1uO+c4cnVd4wlUYiBQ7IMq1Nja0tHcfucYsnenI9vaOeKX7sZN8h2s9\nY3/N4d+os/HoP/zpTLaCVjd2pZ/Nsa3jaKGs/nl2J+k7AiScYgeTXF/n3YP/2nf0Fgo9PV1H\nT6CV+RMrj30qRL595bG3Coz9NYd/n9JW19Hb09U8HtOUhz5vYjhlLu4//vTh0SWsUY3h0FEc\nv9RVsjw2uOmkfEeGOLpAOcpyW8kiYvlL6UY6nw9USLGDya30NOTtR5fDhukGR/YcX+ui6Phq\nN+bXHE7JQ8XSmeyYHpFx/IM3RuwMlfW6qMyDY49bySv5FIe5cfeY8jO2r8YwSs76Hlsvh/mM\nTsp3ZBglfe24L0+utbGxsbVzcPrw0bQewAvjSrEDhlFuOemYmSPZ3t6ScRUrRxvwMYYlqr7O\no9f1Z7I947iUc8xTLoa5g7Tkrs3hm117w5GHMfSVPuD26DVnJQWs++53ik5frnWYejw2Jbcd\nHBui8lc+8UXDo5chRiWP8O3Ltd7dnc/nu9vbWxqKW0vSdrc0dg6m7WxMpUoLIHCiTu5NtkAS\njTRn5NinSfQWhoxCK/kjw8z5GHb0x1hec+S85ZysiSflZ4YM9zm/cxZxbAOKK/hsRh9n/C6n\nGQ++yJi+I2PKMNoX4+hLjJLWiGI4KazYwSTX3JUtWXRpSKVSDS3HLPkcPYFYMox2cOZIfdvt\ng398cNVuDK85nNyGdzkXZcxKz1oef4PrO0qWmoau2RUffHa8dEdv6Rpj6Vej9Khhn9c1Js1d\nw9eqkmeKDZPhXXxHRlDf1lP+kWLpTPboOfRyad858KQPsYHJSbGDSa+5qzfbUfJvczqdyZY+\nLfWdpyjkWgfPjh4zSq60MwzeYlDhaw4rnl5X9mEWIza75et6erOZo+0unekY5tbd474a6XQm\n29vbs2r2iaWPoiiqb+sZ+tKFrqVDjzyR78gomrt6eo/5KkRR+p3P8diT6PVt7xyYPubAjux4\n3RgDk1CqUCjEnQEAgJPAih0AQCAUOwCAQCh2AACBUOwAAAKh2AEABEKxAwAIhGIHABAIxQ4A\nIBCKHQBAIBQ7AIBAKHYAAIFQ7AAAAqHYAQAEQrEDAAiEYgcAEAjFDgAgEIodAEAgFDsAgEAo\ndgAAgVDsAAACodgBAARCsQMACIRiBwAQCMUOACAQih0AQCAUOwCAQCh2AACBUOwAAAKh2AEA\nBEKxAwAIhGIHABAIxQ4AIBCKHQBAIBQ7AIBAKHYAAIFQ7AAAAqHYAQAEQrEDAAjEKXEHqAK/\n+MUv1q1bd+DAgbiDAACJMG3atJUrV55xxhlxBzmeYje6v//7v29ra4s7BQCQIKeccsqf/Mmf\nxJ3ieIrd6A4fPhxF0f333/+Rj3wk7iwAQMz+9V//9YYbbijWg6RR7Cp10UUXXXLJJXGnAABi\ndvDgwbgjlOXmCQCAQCh2AACBUOwAAAKh2AEABEKxAwAIhGIHABAIxQ4AIBCKHQBAIBQ7AIBA\nKHYAAIFQ7AAAAqHYAQAEQrEDAAiEYgcAEIhT4g4AUAV+9atffec733n22WejKLrooosWLFjw\nnve8J+5QAMdT7ABG0dXVddttt7322muDW9773vfec889mUwmxlQAQzkVCzCSP/uzP7vppptO\nP/30zs7O7373u9/97nc7OzunT5/e2tp6zz33xJ0O4BhW7ADK+sEPfrB69epLLrnkiSeeOOOM\nM4obL7300pUrV/7BH/zBl770pWuvvfa3fuu34g0JMMiKHUBZ3/zmN99+++177713sNUVnXHG\nGffee+/bb7/9N3/zN3FlAxhKsQMo6+mnnz7nnHMuvfTSobs+9rGPnX322d/73vcmPhVAOYod\nQFm//OUvj1urK1VXV/fmm29OZB6AkSl2AGXNnDlz3759hw4dGrrr4MGDL7300syZMyc+FUA5\nih1AWVdeeeWvfvWrBx54YOiuBx544MCBAwsWLJj4VADlKHYAZd1www0zZ878/Oc/v2XLltLt\nW7ZsaWtrmzlz5g033BBXNoChjDsBKOv000//h3/4h4ULFy5cuHDevHnz5s1LpVL5fH7nzp11\ndXUbNmw4/fTT484IcJRiBzCSj33sY9///vfvvPPORx55ZOfOnVEU/bt/9+/+y3/5L6tXr/7A\nBz4QdzqAYyh2AKM4//zzv/nNb3Z3d+/bty+KonPPPfeUU/zlCSSRv5sAKnLKKadYogMSTrED\ngNH19/dns9mDBw/GFWBgYGDXrl1z586dMiW2Gx9ra2tbWlpqamriCsCoFDsAGN327duXLFkS\nd4r4bdu2zZSfJFPsAGB0TU1NGzdujHHFbufOnWvXrr3lllvmzZsXV4ba2tqmpqa43p1KKHYA\nMLqamprFixfHm2Ht2rXz5s1btmxZvDFIMgOKAQACodgBQBWYNm3a4Ecox6lYAKgCzc3NW7Zs\nmT9/ftxBSDTFDgCqQE1NzVVXXRV3CpLOqVgAqAL9/f1bt27t7++POwiJptgBQBXI5XJXX311\nLpeLOwiJptgBQBU4cODA4EcoR7EDAAiEYgcAEAh3xQKMolAoPP744w899NCzzz4bRdFFF120\nYsWKhQsXplKpuKMxiZhjRyUUO4CR/OIXv1ixYsWWLVuiKJo1a1YURU899dS6desWLlz44IMP\nnnHGGXEHZLIwx45KOBULUFahUFi2bNmWLVs+85nP/Nu//dvevXv37t37b//2b5/5zGcef/zx\nZcuWFQqFuDMyWRTn2NXU1MQdhERT7ADK2rRp07Zt2zKZzP3333/eeecVN5533nn333//jTfe\nuG3btk2bNsWbkMnDHDsqodgBlLV+/fopU6bceeedQ3d9+ctfnjJlysMPPzzxqZiczLGjEood\nQFl9fX3nnXfejBkzhu6aMWPGrFmzent7Jz4Vk5M5dlRCsQMoa+RL6NwVCySNYgdQVkNDw09+\n8pNXXnll6K5XXnll79699fX1E58KoBzFDqCsT3ziE/39/cNeY7d69eqBgYFPfOITE5+Kyckc\nOyqh2AGUde21186fP/9rX/vaTTfdtH///uLG/fv333TTTffdd9/8+fOvvfbaeBMyeRTn2DU3\nN8cdhEQzoBigrFQq9cgjjyxbtqyrq+vrX//6zJkzoyjav3//wMDAggULHn74YZfZMWGKc+zi\nTkHSWbEDGEldXd2aNWsuueSSQqHw0ksvvfTSS4VC4ZJLLlmzZk1dXV3c6ZhEzLGjElbsAEay\nZcuWpUuXvv3229dee+28efOiKNq5c+djjz02b968DRs2XH311XEHnEReeumlH/7wh3GniM1T\nTz11xx133HXXXZdddlncWeIxZcqUxsbGU089Ne4giabYAZT1s5/97JOf/OT06dNzudxHP/rR\nwe3f+973mpubP/nJTz733HNnn312jAknlY9//OPf/e53404RszvuuCPuCHH6i7/4i1tvvTXu\nFImm2AGU9Y1vfOP1119/+OGHS1tdFEUf/ehH77333uXLl3/jG9/4whe+EFe8yeatt9469+xz\nvrTyxriDxGOgUPjRi89/+PwLpkzKKzv3v/az1d/qfvPNN+MOknSKHUBZ27dvnz59+rC3vi5d\nunT69Ok7duxQ7CbSmdNPzyy+Lu4UxGDPC8+v/lZ33CmqgJsnAMp69dVXZ86cOXXq1KG7pk6d\nOmPGjJ/+9KcTnwqgHMUOoKy6urrXXntt2AeLFQqF11577cwzz5z4VADlKHYAZc2bN+/nP//5\njh07hu7asWPH66+/PmnvTwSSSbEDKOuGG26ora296aab9u3bV7q9+PCJ2traG264Ia5sAEO5\neQKgrA996EMdHR033XTTxRdf3NraOm/evFQqlc/nu7q6Xnvtte7u7g996ENxZwQ4SrEDGElr\na+v73ve+W2+99Z577hnc+KEPfej+++/3oFggaRQ7gFEsXbo0iqK77rrrxRdfjKLo/PPPv+OO\nO7Q6IIEUO4CRvPXWWx/72Mf27NkzuOW111677rrrZs+e/d3vfnf69OkxZgM4jpsnAEbykY98\nZM+ePWedddZXv/rV119//fXXX/+f//N/nnXWWXv27PnIRz4SdzqAYyh2AGV1dnb++Mc/njlz\n5ssvv/y5z32urq6urq7u5ptvfvnll2fMmPHjH/+4s7Mz7owARyl2AGV97Wtfi6LogQceOO7h\nE1OnTn3ggQeiKLrvvvviSQYwHMUOoKx9+/bV1NRcccUVQ3c1NTXV1NQcN98OIF6KHUBZAwMD\nU6aU/XsylUr19/dPZB6AkSl2AGXV1dUdPnx47969Q3e9+OKLb7/9tmfFAomi2AGUtWjRoiiK\n/viP/3jork9/+tNRFLW0tExwJIARKHYAZXV0dEybNm3Hjh3Nzc1vvPFGceMbb7zR3Ny8Y8eO\nadOmdXR0xJsQoJQBxQBlTZs27cknn7ziiisef/zxs846q66uLoqiN954o1AonHrqqU8++eS0\nadPizji5/PSN11NX/G7cKSC5rNgBjOSyyy7bt2/fddddN3369DfeeOONN96YPn36ddddt2/f\nvssuuyzudADHsGIHMIqzzjrr0UcfjTsFwOis2AEABEKxAwAIhFOxAKMbGBj4/ve//6Mf/SiK\nog9/+MO//du/PcLgYsbPv68786cbtsadghjseeH53/r08rhTVAHFDmAUjz322K233vrcc88N\nbrnwwgv/6q/+qjjlDiA5/BcnwEi+8Y1vLFmy5NVXX121atWjjz766KOPrlq16tVXX128ePH9\n998fdzqAY1ixAyjrhRdeuPnmm3/zN39z+/bts2bNKm687rrrPv/5zzc1NX3uc5+bP3/+Bz/4\nwVgzAhxlxQ6grG984xsHDx687777Bltd0axZs+67776DBw9atAMSRbEDKCufz5955plNTU1D\ndzU1NdXV1eXz+YlPBVCOYgdQ1htvvHH22WcPewPslClTzjnnnNdff33iUwGUo9gBlHXOOee8\n/PLLb7/99tBdb7/99v79+88555yJTwVQjmIHUNbv//7vv/nmmxs3bhy6a+PGjW+99dbv//7v\nT3wqgHIUO4Cybrzxxrq6us9+9rPf//73S7d///vf/+xnP1tXV3fjjTfGlQ1gKONOAMo655xz\n/u7v/u4//af/dOmlly5dujSdTkdRlM/nN2zYMGXKlEcffdSpWCBRFDuAkSxatOipp5760z/9\n0//zf/7Pww8/HEXRlClT/uAP/uArX/nK7/zO78SdDuAYih3AKH7nd35n27ZtP/vZz5599tko\nii666KKzzz477lAAw1DsACpy9tln63NAwil2AFSN1996s3vTP8Sdghjsf+1ncUeoDoodwCgK\nhcLjjz/+0EMPDZ6KXbFixcKFC1OpVNzRJpfTTjvtBz97tfUv74k7CLGZPn163BGSTrEDGMkv\nfvGLFStWbNmyJYqi4hNjn3rqqXXr1i1cuPDBBx8844wz4g44iTzyyCPFbj05PfXUU3fcccdd\nd9112WWXxZ0lHlOmTCnemc4IFDuAsgqFwrJly7Zt2/aZz3xm9erV5513XhRFe/fuXb169Te/\n+c1ly5Zt2bLFut2EOe+884rfgsmp+Py6iy66aMGCBXFnIbkUO4CyNm3atG3btkwm09XVNbjx\nvPPOu//++2tqar7+9a9v2rRpyZIlMSYEKOXJEwBlrV+/fsqUKXfeeefQXV/+8penTJlSnGwH\nkBCKHUBZfX1955133owZM4bumjFjxqxZs3p7eyc+FZPTtGnTBj9COU7FArxLrq5jIjU3N2/Z\nsmX+/PlxByHRrNgBlFVfX/+Tn/zklVdeGbrrlVde2bt3b319/cSnYnKqqam56qqrampq4g5C\noil2AGV94hOf6O/vX7169dBdq1evHhgY+MQnPjHhoZik+vv7t27d2t/fH3cQEk2xAyjr2muv\nnT9//n333dfa2rp///7ixn379rW2tt53330LFiy49tpr403I5JHL5a6++upcLhd3EBLNNXYA\nZaVSqUceeWTZsmXd3d1f//rXP/jBD0ZR9MILLxQKhSuvvHL9+vUus2PCHDhwYPAjlKPYAYyk\nrq5u69atjz322Pr16/fs2ZNKpRobG5cvX37NNddodUDSKHYAo0ilUosXL168eHHcQQBG4Ro7\nAKgC5thRCSt2AFAFzLGjEoodAFSB4hy7uFOQdE7FAkAVMMeOSih2AFAFzLGjEoodAFQBc+yo\nhGIHABAIxQ4AIBCKHQBUAXPsqIRxJwBQBcyxoxKKHQBUAXPsqIRTsQBQBcyxoxKKHQBUAXPs\nqEQ1nYptbGys/OCenp7xSwIAE8wcOypRTcVuzpw53d3dJ/c1+/v7s9nswYMHRzjmX/7lX6Io\nOnz48Ml9awCAk6uail1XV9fs2bPb29szmUxXV9dJec3t27cvWbKkkiMfeOCBK6644qS8KQDA\neKimYhdFUVtb2549e7q7u5cuXdrc3HziL9jU1LRx48aRV+zuvffeHTt2zJo168TfDgDeHXPs\nqESVFbsoirq6unbv3t3S0lIoFE781WpqahYvXjzyMdlsNoqiKVPcaAJAbMyxoxJVWVZuv/32\nTCbjziAAJo/iHLuampq4g5Bo1bdiF0VRc3PzSTkPCwDVor+//4knnpg/f75uxwiqcsUOACYb\nc+yohGIHAFXAHDsqodgBAASiKq+xA4AJVslA+3G1c+fOwY9xqa2tbWlpcZFfkil2ADC6ygfa\nj6u1a9euXbs2xgDbtm1bsGBBjAEYmWIHAKOrZKD9uBoYGNi1a9fcuXNjnKtaW1vb1NQU17tT\nCcUOAEZXyUD78bZixYp4A5B8bp4AAAiEYgcAEAjFDgAgEIodAEAgFDsAgEAodgAAgVDsAAAC\nodgBAARCsQMACIRiBwAQCMUOACAQih0AQCAUOwCAQCh2AACBUOwAAAKh2AEABEKxAwAIhGIH\nABAIxQ4AIBCKHQBAIBQ7AIBAKHYAAIFQ7AAAAqHYAQAEQrEDAAiEYgcAEAjFDgAgEIodAEAg\nFDsAgEAodgAAgVDsAAACodgBAARCsQMACIRiBwAQCMUOACAQih0AQCAUOwCAQCh2AACBUOwA\nAAKh2AEABEKxAwAIhGIHABAIxQ4AIBCnxB2AatLf35/NZg8ePBhXgIGBgV27ds2dO3fKlNj+\nm6S2tralpaWmpiauAABQjmLHGGzfvn3JkiVxp4jftm3bFixYEHcKADieYscYNDU1bdy4McYV\nu507d65du/aWW26ZN29eXBlqa2ubmpriencAGIFixxjU1NQsXrw43gxr166dN2/esmXL4o0B\nAAnk5gkAgEAodlSTadOmDX4EAI7jVCzVpLm5ecuWLfPnz487CAAkkWJHNampqbnqqqviTgEA\nCeVULNWkv79/69at/f39cQcBgCRS7KgmuVzu6quvzuVycQcBgCRS7KgmBw4cGPwIABxHsQMA\nCIRiBwAQCMWOamKOHQCMwLgTqok5dgAwAsWOamKOHQCMwKlYqok5dgAwAsWOamKOHQCMQLGj\nmphjBwAjUOwAAAKh2AEABEKxo5qYYwcAIzDuhGpijh0AjECxo5qYYwcAI3Aqlmpijh0AjECx\no5qYYwcAI1DsqCbm2AHACBQ7AIBAKHYAAIFQ7Kgm5tgBwAiMO6GamGMHACNQ7Kgm5tgBwAic\niqWamGMHACNQ7Kgm5tgBwAgUO6qJOXYAMALFDgAgEIodAEAgFDuqiTl2ADAC406oJubYAcAI\nFLsq8+abb27dunVgYCDuIHF69NFH444Qp4svvviiiy6KOwUASaTYVZl77733C1/4QtwpiNMl\nl1zyve99L+4UACSRYldl3pn08VdRNCvuKMTi1ujgwYNxhwAgoRS76nR1FM2OOwOxuDPuAAAk\nmLtiAQACodgBAARCsQMACIRiBwAQCMUOACAQih0AQCAUOwCAQCh2AACBUOwAAAKh2AEABMIj\nxarTb8UdgLicE0X/Pu4MACSVFTsAgEAodgAAgVDsAAACodgBAATCzRPV6QdRNDvuDMRiTtwB\nAEgwK3YAAIFQ7AAAAqHYAQAEQrEDAAiEYgcAEAjFDgAgEIodAEAgFDsAgEAodgAAgVDsAAAC\nodgBAARCsQMACIRiBwAQCMUOACAQp8QdgHfll1H0etwZYtEfRf9fFP1eFNXEnSQu/ZP4cwdg\nNIpdlZkyZUoURdG8uHMQn9ScVNwRAEgoxa7KLF++fP/+/QMDA3EHiUehUHjppZfe//73p1KT\nt9xcccUVcUcAIKEUuyrz4Q9/+Gtf+1rcKQCAJHLzBNWkv79/69at/f39cQcBgCRS7KgmuVzu\n6quvzuVycQcBgCRS7KgmBw4cGPwIABxHsQMACIRiBwAQCMWOajJt2rTBjwDAcYw7oZo0Nzdv\n2bJl/vz5cQcBgCQa84pdZ2dna2tra2tr8bfuT2Qi1dTUXHXVVTU1HqoFAMMYw4pdLpdraWkZ\n/G1XV1cURS0tLel0uqen5+RHgyH6+/ufeOKJ+fPn63YAMFSlK3aDrS6bzZZuz2Qy+Xy+s7Pz\n5EeDIcyxA4ARVFrs7r777iiKstlsc3Nz6fZVq1ZFUbR+/fqTngyGMscOAEZQabHL5/NRFB3X\n6qIoqq+vH9wLAECMjDsBAAhEpcUunU5HUdTX13fc9uLVTsW9MN7MsQOAEVRa7JYvXx5F0cqV\nK0s3Dt5RUdwL4604x27oJQEAQFT5uJO2trY9e/Z0d3enUqnilsFfZDKZtra2cUkHxyrOsYs7\nBQAk1Biusevq6spms6VnXdPpdDabLQ60gwnQ39+/devW/v7+uIMAQBKN7ZFizc3NzoIRo1wu\nt3jx4k2bNl1zzTVxZwGAxHFXLNXEHDsAGMHYVuwSoq+vb/PmzXv27Nm9e3dxy5w5c2bPnr1o\n0aLiXD0AgEmo0mI3eKtEOYVC4YTDjK6vr2/lypVD5yEXt7S3t6fT6XXr1ql346S/vz+bzR48\neDCuADt37hz8GJfa2tqWlhYPqwUggSotdplMZujG7u7ucrvGQ19fX0NDQxRF6XT69ttvL/56\nUG9v7913353P5xsaGnp7e3W78bB9+/YlS5bEnSJau3bt2rVrYwywbdu2BQsWxBgAAIZVabEb\n9tbXrq6u1tbW3bt39/T0nNRUw1uzZk0URZlMZtgw9fX1zc3Nra2t3d3da9asqfBe3UqWoF54\n4YUoigYGBt5N6LA0NTVt3LgxxhW7gYGBXbt2zZ07d8qU2C4Pra2tbWpqiuvdAWAEqRM8hVpc\nRStXtk6u4ungUQNXeFjRd77znSuvvLKSIz/96U//zd/8TSVHAgAB6+npuf/i7iMAAB9KSURB\nVPzyyzs6OhI4x/dEb54onvHs7u6u0ml2lSxB3XvvvTt27LjgggsmLBUAwLtwosVu6NNjx086\nnc7n87lcboRZemN9dm1NTc3ixYtHPiabzUZRFOO5PwCASlRa7IqFaajis2In5v6J22+/vaWl\npaWlJZvNDtvtBp9de/vtt09AHgCARKm02BULUzmrVq06GWFG0dzc3NHR0d7ePhhmcGWudABK\nR0eHx2MAAJPQCY07iaJo6dKlE9mi2traFi1atHnz5vXr1+fz+dI+l06nly9fbkYxADBpndC4\nk1jU19e3tbUl8D4UAIB4uSEAACAQih0AQCBGOhU76vNhS03Ms2IBAChnpGI3YQ+BBQDgxI1U\n7JJzwwQAAKNyjR0AQCDG9kixoQ8Q6+3tjaJow4YNlvcAAOI1hmLX2NhYOhD4OIodAEC8Kj0V\n29raWmx1g0/xGvxFFEXZbPakJwMAYEwqLXa7d++Ooiibzfb09BQr3bp16wZHnDz33HPjlA8A\ngApVWuyKy3XFx8LOmTMnOnJ1XXGtbv369eMVEACAyrybu2Jnz54dHVmlK1a9Ea69AwBgYlRa\n7EqvqLvwwgujKNqzZ8+4JAIA4F2ptNgVT7/mcrkoihoaGqIjV90NHYACAEAsKi12q1atiqKo\npaUliqL6+vpMJpPP51OpVLHkefgYAEDsKi129fX1vb29gydku7q6Bn+dyWQMsQMAiN0YBhTX\n19f39PQM/rb01wAAxM6zYgEAAlFpsUulUo2NjcWbJwAASKAxjDvJ5/MtLS2pVKq1tdXNsAAA\nSVNpsevp6SkUCh0dHVEUdXd3NzQ0pFKpzs7O8cwGAMAYjO0au7a2tkKh0NvbW5xv0t7e7hQt\nAEBCvJubJ+rr67u6uooNb/AU7UlPBgDAmIxh3Empvr6+zZs3t7e3F39b+sAxAABiMeZi19nZ\nuX79+nw+X/xtR0fHokWL6uvrT3YwAADGptJil8vl7r777sE+l06n161bp88BACRHpcWueBVd\nOp2+/fbbm5ubxzMSAADvRqXFLpPJrFq1yhIdAEBiVVrsurq6KjkslUpFUVQoFN59IgAA3hXP\nigUACIRiBwAQCMUOACAQih0AQCAUOwCAQCh2AACBUOwAAAKh2AEABEKxAwAIhGIHABAIxQ4A\nIBCVPiu2Qp4SCwAQFyt2AACBqLTYpSrQ2NjY19c3rnEBACin0mKXyWQymUzx1+l0OpPJpNPp\n0r3pdDqfzzc0NOh2AACxqLTYrVq1qru7O51OFwqFnp6erq6unp6eQqFQrHezZ8/u6enp6OiI\nomjNmjXjmBcAgDIqLXYrV66MomjdunXHbS9uWb9+fRRFixYtiqJo9+7dJzMgAACVqbTY5fP5\nKIrq6+uP217cUrq3+GsAACbY2O6KHXr9nCvqAAASYgw3T0RHTsiWKl5RV9yby+WiKCq9qQIA\ngAlT6YDirq6u7u7ufD6fSqUGb4/t7u4e3BtFUUtLSxRFy5cvH4ecAACMYgynYnt7e4uVrvuI\nKIoymUzp0yYymUxbW9tJTwkAwKjG8Eix+vr6rq6u4uLcsDxPDAAgRh4pBgAQiDGs2EVR1NfX\nt2bNmsFJdXPmzFm1atXQGSgAAEy8MRS7zs7O9vb20i35fL67u7ujo8N1dQAAsav0VGwulyu2\nuo6Ojt7e3kKh0NvbW3yGWHt7e3HQCQAAMaq02N19991RFBUX54rnXuvr69va2ordbsOGDeMX\nEQCASoztkWJDT7kWtwwOtAMAIC7uigUACESlxa74oLCh19J5jBgAQEJUWuyKDwpraWnp6+sb\n3NjX1+cxYgAACVHpuJO2trb169fn8/mGhobB9bnihXfpdNq4EwCA2I3hGruenp7iPbD5I6Io\n6ujo6OnpGa90AABUbGxPnmhra7M4BwCQTO6KBQAIhGIHABCIkU7FplKpyl+oUCiccBgAAN69\nkYpdJpOZsBwAAJygkYpdV1fXhOUAAOAEucYOACAQih0AQCAUOwCAQCh2AACBUOwAAAKh2AEA\nBEKxAwAIhGIHABAIxQ4AIBCKHQBAIBQ7AIBAKHYAAIFQ7AAAAqHYAQAEQrEDAAiEYgcAEAjF\nDgAgEIodAEAgFDsAgEAodgAAgVDsAAACodgBAARCsQMACIRiBwAQCMUOACAQih0AQCAUOwCA\nQCh2AACBUOwAAAKh2AEABEKxAwAIhGIHABAIxQ4AIBCKHQBAIBQ7AIBAKHYAAIFQ7AAAAqHY\nAQAEQrEDAAiEYgcAEAjFDgAgEIodAEAgFDsAgEAodgAAgVDsAAACodgBAARCsQMACIRiBwAQ\nCMUOACAQih0AQCAUOwCAQCh2AACBUOwAAAKh2AEABEKxAwAIhGIHABAIxQ4AIBCKHQBAIBQ7\nAIBAKHYAAIFQ7AAAAqHYAQAEQrEDAAiEYgcAEAjFDgAgEIodAEAgFDsAgEAodgAAgVDsAAAC\nodgBAARCsQMACIRiBwAQCMUOACAQih0AQCAUOwCAQJwSd4AxaGxsrPzgnp6e8UsCAJBA1VTs\n5syZ093dHXcKAICEqqZTsV1dXR0dHVEUZTKZwmjiDgsAMNGqqdhFUdTW1pbJZLq7u3O5XNxZ\nAACSpZpOxRZ1dXXt3r27paXlpCzL9ff3Z7PZgwcPjnDMCy+8EEXRwMDAib8dAMD4qb5iF0XR\n7bffvmHDhlwu19zcfIIvtX379iVLllRy5PPPP3+C7wUAMK6qstg1NzefeKUrampq2rhx48gr\ndvfee++OHTsuuOCCk/KOAADjpCqL3UlUU1OzePHikY/JZrNRFE2ZUmXXIwIAk42yAgAQiBCK\nXSqVSqVScacAAIhZCMUOAIBIsQMACIZiBwAQCMUOACAQIYw7KY4jAQCY5EIodidrWDEAQFVz\nKhYAIBCKHQBAIBQ7AIBAKHYAAIFQ7AAAAqHYAQAEQrEDAAiEYgcAEAjFDgAgEIodAEAgFDsA\ngEAodgAAgVDsAAACodgBAARCsQMACIRiBwAQCMUOACAQih0AQCAUOwCAQCh2AACBUOwAAAJx\nStwBoCL9/f2bNm3asmXLyy+/fNppp33sYx/71Kc+ddZZZ8WdCwASRLGjCjz77LPLly9/5pln\noij6jd/4jV//+td///d/f9ttt33ta1/71Kc+FXc6AEgKp2JJuldeeWX+/Pk//OEPb7vtthdf\nfPHQoUNvvvnmAw88UFdX98d//MePPvpo3AEBICkUO5LuzjvvfOmll9atW3f33Xd/4AMfiKJo\n+vTpn/zkJ3fu3HnOOefcfPPNhw4dijsjACSCYkei9ff3P/jggx/96Ec/+clPHrfr/e9//623\n3rpv374dO3bEEQ0AEkexI9Fefvnl119//fLLLx92b2NjYxRFP/zhDyc2FAAklGJHoh0+fDiK\not/4jd8Ydm9x+69//esJzQQASaXYkWgzZsw49dRTd+3aNeze4vYPfvCDE5oJAJJKsSPRamtr\nr7766q1bt/7Lv/zLcbsOHTrU0dExbdq0K6+8MpZsAJA0ih1Jd9ddd51yyimLFi3asmXL4MYX\nX3xxyZIlzzzzzH//7//9zDPPjDEeACSHAcUk3cUXX/zQQw/90R/90cKFCz/wgQ9ceOGFP//5\nz//1X/+1v7//T/7kT2677ba4AwJAUlixowpce+21u3fvvvXWW88444ynn376tdde+/jHP/7k\nk0/+9V//dSqVijsdACSFFTuqwwc+8IG/+Iu/iDsFACSaFTsAgEAodgAAgVDsAAAC4Ro7oGr0\n9/dns9mDBw/GFWBgYGDXrl1z586dMiWe/yqura1taWmpqamJ5d2B5FPsgKqxffv2JUuWxJ0i\nZtu2bVuwYEHcKYCEUuyAqtHU1LRx48YYV+x27ty5du3aW265Zd68ebEEqK2tbWpqiuWtgaqg\n2AFVo6amZvHixfFmWLt27bx585YtWxZvDIBhuXkCACAQih1ApaZNmzb4ESCBnIoFqFRzc/OW\nLVvmz58fdxCA4Sl2AJWqqam56qqr4k4BUJZTsQCV6u/v37p1a39/f9xBAIan2AFUKpfLXX31\n1blcLu4gAMNT7AAqdeDAgcGPAAmk2AEABEKxAwAIhGIHUClz7ICEM+4EoFLm2AEJp9gBVMoc\nOyDhnIoFqJQ5dkDCKXYAlTLHDkg4xQ6gUubYAQmn2AEABEKxAwAIhGIHUClz7ICEM+4EoFLm\n2AEJp9gBVMocOyDhnIoFqJQ5dkDCKXYAlTLHDkg4xQ6gUubYAQmn2AEABEKxAwAIhGIHUClz\n7ICEM+4EoFLm2AEJp9gBVMocOyDhnIoFqJQ5dkDCKXYAlTLHDkg4xQ6gUubYAQmn2AEABEKx\nAwAIhGIHUClz7ICEM+4EqNShQ4duvvnmN954I+4gsSkUCr/3e7+3bt26v/3bv407Szzq6uq+\n+tWvnnrqqXEHAYan2AGV+vGPf/z1r3897hTE7JZbbvkP/+E/xJ0CGJ5TsUClCoVCFEVfiqKC\n/03K/32p5McASCbFjuqwd+/e//bf/ttv//Zvn3POOb/5m795/fXX/+M//mPcoQAgWRQ7qsCm\nTZvmzJnzla985bXXXps7d+5pp5320EMPXXHFFTfffLPFAwAYpNiRdLt27Vq+fPl73vOezZs3\n792798knn3zmmWf6+voWLFjwv/7X/7rnnnviDggASaHYkXR33HHH4cOHN2/e3NLSMrjxggsu\n2LRp09y5c++5557JfJMmAJRS7Ei0gwcPPv7441deeeV//I//8bhdtbW1t9xyy69+9att27bF\nkg0AkkaxI9FefvnlQ4cOXXzxxcPunTt3bhRFzz///MSGAoCEUuxItKlTp0ZRdPjw4WH3FrcX\njwEAFDsSbcaMGWeeeeY//dM/Dbu3uN2sVAAo8uQJEq2mpmbFihX33Xfft7/97fnz5z/55JMv\nvfTSGWeccemll5599tl/+Zd/OXPmzKamprhjAkAiKHYk3erVqzdu3PiHf/iHqVSqv79/cPup\np556+PDhhx9+2GMrAaDIqViSrq6u7txzzx0YGCi2umnTphW3Hzp0qLa2tqGhIdZ0AJAgVuxI\nuo6Oju9973vt7e2XX3751q1bf/KTn5xxxhmXXXbZ+eefv2LFis985jP//M//nEql4o4JAPFT\n7Ei6rq6u888//ytf+crUqVM//vGPl+767Gc/29HR8fTTT3/0ox+NKx4AJIdTsSTaq6+++vzz\nzy9cuHDYmSbXXHNNFEX//M//POG5ACCJrNiRaL/4xS+iKDrrrLOG3fve97538BgmzJ1RdGfc\nGQAYlhU7Eu1973tfKpV68cUXh91bfObEjBkzJjYUACSUYkeinX766Zdeeuljjz326quvDt37\nrW99K5VKmWMHAEWKHUn3xS9+8Ze//OV1113305/+dHDjwMDAl770pY0bN15//fUf/OAH40sH\nAAniGjuS7tprr/3iF7/4P/7H/6ivr1+6dOlFF1306quvZrPZ3t7eSy655K//+q/jDggASaHY\nUQXuueeeSy+99M/+7M/+7u/+rlAoRFE0Y8aML33pS1/4whdqa2vjTjfpfCmKVsedgVisdt8M\nJJ5iR3VYunTp0qVLf/aznxWfFXv++ecbSgwAx1HsqCZnn3322WefHXcKAEgoN08AAARCsQMA\nCIRTsVSH/v7+//t//++WLVv2798/ffr0yy677Prrr3daFgBKKXZUgR/96EfLli3bvXt3FEXT\npk07cODAt7/97dtuu+3ee+/9oz/6o7jTAUBSOBVL0r388svz589/9tln77jjjr179/7qV796\n6623Hnzwwfe+970rV6585JFH4g4IAEmh2JF0d9555759+/72b//2y1/+8qxZs6IoOu2001as\nWJHP59/3vvd9/vOfP3ToUNwZASARFDsS7e23337ooYd+93d/9z//5/983K5zzz33v/7X/7p/\n//4dO3bEEQ0AEkexI9FeeeWV119//fLLLx92b3H7D3/4w4kNBQAJ5eYJEu3w4cNRFE2dOnXY\nvcXtv/71ryc006T3v6OoJ+4McSlE0c+j6KwompyPPflx3AGAUSl2JNrMmTNPPfXUXbt2Dbu3\nuP2CCy6Y2FCT17nnnnthQ8Orr776dNxJ4nL48OG3/t//m37aaeX+YyN4c97//nPPPTfuFEBZ\nih2Jduqppy5cuPCxxx57+umnL7nkktJdhw4dWrt27Xve854rr7wyrniTTV1d3bPPPRd3ijg9\n9thjixcv/vaDD15zzTVxZwEYhmvsSLq77rpr6tSp11xzTTabHdz4/PPPX3PNNbt27brtttvq\n6upijMek0tzcvGXLlubm5riDAAzPih1JN3fu3PXr1//hH/7hokWLZs2a1dDQ8Nprr+3evXtg\nYOBzn/vcF7/4xbgDMonU1NRcddVVcacAKMuKHVVg8eLFP/jBD/70T//0ve997zPPPPPWW2+t\nWLHiH//xH7/61a+mUpPzKnbi0d/fv3Xr1v7+/riDAAzPih3VYdasWX/+53/+53/+53EHYVLL\n5XKLFy/etGmTa+yAZLJiB1CpAwcODH4ESCDFDgAgEIodAEAgFDuASk2bNm3wI0ACuXkCoFLF\nOXbz58+POwjA8BQ7gEqZYwcknFOxAJUyxw5IOMUOoFK5XO7qq6/O5XJxBwEYnmIHUClz7ICE\nU+wAAAKh2AEABEKxA6iUOXZAwhl3AlApc+yAhFPsACpljh2QcE7FAlTKHDsg4RQ7gEqZYwck\nnGIHUClz7ICEU+wAAAKh2AEABEKxA6iUOXZAwhl3AlApc+yAhKvKYle8Ja25uXnYva2trVEU\ndXV1TWgmYBIwxw5IuCo7FZvL5VKpVEtLS0tLSyqVamxs7OvrO+6Y7u7u7u7uWOIBYTPHDki4\naip2uVyupaUliqJ0Op3JZNLpdD6fb2hoGNrtAMaDOXZAwlVTsduwYUMURR0dHT09PV1dXT09\nPdlsNooi3Q6YGObYAQlXTcWueIJ10aJFg1uam5sHu11ssQAAkqGail1RfX196W+bm5s7Ojqi\nKGpsbIwpEQBAIlRTsUun01EUDT3r2tbWlslk8vl8Z2dnHLmAycIcOyDhqmncyZw5c/L5/ObN\nm9va2o7b1dXVtXv37vb29rG+Zn9/fzabPXjw4AjHvPDCC1EUDQwMjPXFgcCYYwckXKpQKMSd\noVJ9fX2D19ING7uxsTGfz49wwFDf+c53rrzyykqOvPHGG01RAQB6enouv/zyjo6OoStNsaum\nFbv6+vre3t6VK1cOtrfj9PT0tLa2jql+NTU1bdy4ceQVu82bN69bt+76668fW1wgOP39/U88\n8cT8+fNramrizgIwjGoqdlEU1dfX9/T0jHBAV1fXmJ45UVNTs3jx4pGP2bdv37p166ZOnVr5\nywJByuVyixcv3rRp0zXXXBN3FoBhVNPNEwDxMscOSLgQil0qlUqlUnGnAACIWZWdigUms0pu\nYx9XO3fuHPwYi9ra2paWFlf4AeUodkDV2L59+5IlS+JOEa1du3bt2rVxvfu2bdsWLFgQ17sD\nCafYAVWjktvYx9XAwMCuXbvmzp07ZUo817HU1tY2NTXF8tZAVVDsgKpRyW3s423FihXxBgAY\nQQjFLpvNxh0BACB+IRS75ubmuCMAAMQvhHEnAABEih0AQDAUOwCAQCh2AACBUOwAAAKh2AEA\nBEKxAwAIhGIHABAIxQ4AIBCKHQBAIBQ7AIBAKHYAAIFQ7AAAAqHYAQAEQrEDAAiEYgcAEAjF\nDgAgEKfEHaBqPPvss7W1tXGnmOwOHz78rW996/zzz58yxX+TEIOBgYG+vr76+no/gUy8gYGB\nF1988dOf/vTUqVPjzjLZPfvss3FHKEuxG13x/0I33HBD3EEAmOy6urrijsA7ktmwFbvRfepT\nn3r77bcPHDgQdxCiZ5555oEHHrj88svPP//8uLMwGb344ov/9E//5CeQWBR//K6//vqLL744\n7ixE06ZN+9SnPhV3imGkCoVC3BmgUg8//PDy5cvXr1+/bNmyuLMwGfkJJEZ+/KiEy0QAAAKh\n2AEABEKxAwAIhGIHABAIxQ4AIBCKHQBAIBQ7AIBAKHYAAIFQ7AAAAqHYUU2mTZs2+BEmnp9A\nYuTHj0p4pBjVpL+//4knnpg/f35NTU3cWZiM/AQSIz9+VEKxAwAIhFOxAACBUOwAAAKh2AEA\nBEKxAwAIhGIHABAIxQ4AIBCKHQBAIBQ7AIBAKHYAAIFQ7AAAAqHYAQAEQrEDAAiEYgcAEAjF\nDgAgEIodAEAgFDsAgEAodlSfzs7OVCoVdwomkb6+vtbW1tQRjY2NuVwu7lBMIsW/9AZ//Pr6\n+uJORHKlCoVC3BlgDHK5XEtLSxRFfnSZGH19fQ0NDUO3ZzKZrq6uic/DZNPY2JjP54/bmM1m\nm5ubY8lDwlmxo5p0dnYWWx1MmDVr1kRRlMlkent7C4VCoVDIZrNRFHV3d1u3Y7zlcrl8Pp9O\npwd//Do6OqIo8jch5Vixozr09fWtXLmy+BdcFEX5fN6PLhMjlUql0+menp7SjZ2dne3t7Rbt\nGG/F5bre3t76+vrBja2trd3d3cdthCIrdlSHNWvW5PP5jo6O4/59hfFWKBSG/tQtWrQoljBM\nNj09PYVC4bgCt3v37rjykHynxB0AKrJq1SpLIyTH5s2boyiaPXt23EGYdFpbW/P5fCaTsVzH\nsJyKpfoUz0340SUug7dT+CFkwpTexNPR0dHW1hZvHhLLqViAMRj897V4CwVMmHQ6XbzIuL29\nvbW1Ne44JJRTsQCVGpy2Y9gEE6y+vn7wWs/izRNRFLlAhaGciqX6OBVLLAb/NdXqiF1xSLu/\nBhnKqViA0TU2NnZ3dxfHiWl1xK54ThaGUuwARlFcJM5kMj09PW5FZCIVHyN23Ma+vr7BoZ5w\nHMUOYCSdnZ3FGYquZ2LiZTKZKIpKnw+by+WKt+8sX748zmQklWvsqD6usWMiDV0vGTT0iRRw\n0g37rFhPPaEcK3YAZXkaLLHr6ekpPh+2KJ1OZ7NZrY5yrNgBAATCih0AQCAUOwCAQCh2AACB\nUOwAAAKh2AEABEKxAwAIhGIHABAIxQ4AIBCKHQBAIBQ7AIBAKHYAAIFQ7AAAAqHYAQAEQrED\nAAiEYgcAEAjFDgAgEIodAEAgFDsAgEAodgAAgVDsAAACodgBAARCsQMACIRiBwAQCMUOACAQ\nih0AQCAUOwCAQCh2AACBUOwAAAKh2AEABEKxAwAIhGIHABAIxQ4AIBCKHRCPXGsqlWrs7Bvv\n9+nrbEylUq25oxtynZ1HfzdRMY5/3/H9U8BkpdgBk0lfZ2NDS/ueKnnfuNICVUuxAwAIhGIH\nABAIxQ6IWV9na2MqlUqlUo2tnbnjL3bryw3uTqUaG1uHHFA8onHwmDJHRcWr7Rra81EUdbcM\nvbBu5Bijfxa51pIMx7xE2fcdMfmwf2qYKwKHuYawbBIgeIodEKc9axob2rvzxd/ku9tbVpa2\nluI1Zkd2R1E+393S0FhSYgaPyOePbiseNab7IUaOMaq+zsaGlu6SDPnu9tEinKTkJyEJEBDF\nDohRvrs76sj2FgqFQqG3Ix1FUX795iMlJNfa0J6P0plsb/GAQqE3m0lH+e6WwWqXW9Oej6L0\nkZd455jo2Nc5or6t5503yWQLhUJPW31FMUZ3JEXv0Qwd6aOvMez7jpq8fNp3nwQInmIHxCnd\nsa6tuVhZ6ttuz0RRlN/TW9yV29AdRemOdV3N9Uc6TX1zV082E0XdG95pds1dhUKh0HPkJYrH\nrOpIn8QYY9B7pD3VN7f1jFLGTlbyE08ChESxA2KUXr6opHE0zC7pNX3P7Y6iKN/ekDpWS3cU\nRbufO3YFqq+vL5fL5TpbW1sbG4tXpp2kGJVoXpqJoijf3tLwzoVynbm+SlfITjD5SUwChOCU\nuAMADKt3zwglJ7+nN4rqoyiK+jpbV5ZchReP5q7ebLSypRgjn+/Ot3S3R1E6k+3pai77h8Yl\n+btKAoTDih2QTMV1s0y2MKx3akqutXjPQzqTyXR0ZLPZbG/vO1emTaz65q6eQqHQ25vtyGTS\nxfcvvRhwiHFLPuYkQEis2AHJVH/hnCjKd2/IdTWXXWt65zK83mMuIevrXB/bAl59fXNbV3Nb\n1BX1dbY2tHfvfq4vah7u8rYTSX50uTKKor7Nw/+ZipMAQbFiByRU8Xqx7pbSqXR9uc7G1LFj\n26L8+s1Hh7/lWiu6UO25k3nhWXGS3LHD8/p693SP+r6VJj/6p4rLmN13vzO+pC/XubL0z4wh\nCRCo4c9yAIyzbCY6ZjBHoVA63ePIMcOemjx6RHFAyPDeOerYFy052Vl870piVPa5DFU6dmTY\n9x05+TB/aujJ2nQmU5p29CRA0KzYAcnV3NXTmz1yoVgURVE6ncn2Fo7eB9DcNWR/R7a32H6O\nv3M2iqIoqm9b1/FOW3w3A03K5iwcmyNKZzqypedZh7xvRcmHpq1v6+k9silKZzp6e1bNHlsS\nIGipQqEQdwYAAE4CK3YAAIFQ7ADKyLWmKmCQCJAcih0AQCBcYwcAEAgrdgAAgVDsAAACodgB\nAARCsQMACIRiBwAQCMUOACAQih0AQCAUOwCAQCh2AACBUOwAAAKh2AEABEKxAwAIhGIHABAI\nxQ4AIBCKHQBAIBQ7AIBAKHYAAIFQ7AD4/9utAxkAAACAQf7W9/iKImBC7AAAJsQOAGBC7AAA\nJsQOAGBC7AAAJsQOAGAi1CVGYKf+1aEAAAAASUVORK5CYII=",
      "text/plain": [
       "Plot with title \"Taxa 2 - Abundance\""
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "boxplot(log_value ~ health_status, data=taxa2, col=c(\"green\", 'red', 'pink'),\n",
    "       main=\"Taxa 2 - Abundance\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "7cda9be1-5763-46de-be68-013317da7a6b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "\tPairwise comparisons using t tests with pooled SD \n",
       "\n",
       "data:  taxa2[, \"log_value\"] and taxa2[, \"health_status\"] \n",
       "\n",
       "  1       2      \n",
       "2 < 2e-16 -      \n",
       "3 5.4e-12 < 2e-16\n",
       "\n",
       "P value adjustment method: BH "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "t = pairwise.t.test(taxa2[,'log_value'], taxa2[,'health_status'], p.adjust.method = \"BH\")\n",
    "t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "ee78d831-88c5-457d-8b68-3af21213cce6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 2 × 2 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>1</th><th scope=col>2</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>2</th><td>9.807144e-18</td><td>          NA</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>5.408914e-12</td><td>1.388275e-35</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 2 × 2 of type dbl\n",
       "\\begin{tabular}{r|ll}\n",
       "  & 1 & 2\\\\\n",
       "\\hline\n",
       "\t2 & 9.807144e-18 &           NA\\\\\n",
       "\t3 & 5.408914e-12 & 1.388275e-35\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 2 × 2 of type dbl\n",
       "\n",
       "| <!--/--> | 1 | 2 |\n",
       "|---|---|---|\n",
       "| 2 | 9.807144e-18 |           NA |\n",
       "| 3 | 5.408914e-12 | 1.388275e-35 |\n",
       "\n"
      ],
      "text/plain": [
       "  1            2           \n",
       "2 9.807144e-18           NA\n",
       "3 5.408914e-12 1.388275e-35"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "t$p.value"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "570ba1f5-3f3a-4599-aa07-4a9c1713f372",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "9d1abac9-e1b7-4224-9181-5d5c50e54e62",
   "metadata": {
    "tags": [],
    "toc-hr-collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table style=\"width: 100%;\"><tr><td>ancombc {ANCOMBC}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2 id='ancombc'>Analysis of Compositions of Microbiomes with Bias Correction\n",
       "(ANCOM-BC)</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p>Determine taxa whose absolute abundances, per unit volume, of\n",
       "the ecosystem (e.g., gut) are significantly different with changes in the\n",
       "covariate of interest (e.g., group). The current version of\n",
       "<code>ancombc</code> function implements Analysis of Compositions of Microbiomes\n",
       "with Bias Correction (ANCOM-BC) in cross-sectional data while allowing\n",
       "for covariate adjustment.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Usage</h3>\n",
       "\n",
       "<pre><code class='language-R'>ancombc(\n",
       "  data = NULL,\n",
       "  assay.type = NULL,\n",
       "  assay_name = \"counts\",\n",
       "  rank = NULL,\n",
       "  tax_level = NULL,\n",
       "  phyloseq = NULL,\n",
       "  formula,\n",
       "  p_adj_method = \"holm\",\n",
       "  prv_cut = 0.1,\n",
       "  lib_cut = 0,\n",
       "  group = NULL,\n",
       "  struc_zero = FALSE,\n",
       "  neg_lb = FALSE,\n",
       "  tol = 1e-05,\n",
       "  max_iter = 100,\n",
       "  conserve = FALSE,\n",
       "  alpha = 0.05,\n",
       "  global = FALSE,\n",
       "  n_cl = 1,\n",
       "  verbose = FALSE\n",
       ")\n",
       "</code></pre>\n",
       "\n",
       "\n",
       "<h3>Arguments</h3>\n",
       "\n",
       "<table>\n",
       "<tr><td><code id=\"ancombc_:_data\">data</code></td>\n",
       "<td>\n",
       "<p>the input data. The <code>data</code> parameter should be either a\n",
       "<code>phyloseq</code> or a <code>TreeSummarizedExperiment</code> object, which\n",
       "consists of a feature table (microbial count table), a sample metadata table,\n",
       "a taxonomy table (optional), and a phylogenetic tree (optional).\n",
       "Ensure that the row names of the metadata table match the sample names in the\n",
       "feature table, and the row names of the taxonomy table match the taxon\n",
       "(feature) names in the feature table. For detailed information, refer to\n",
       "<code>?phyloseq::phyloseq</code> or\n",
       "<code>?TreeSummarizedExperiment::TreeSummarizedExperiment</code>.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"ancombc_:_assay.type\">assay.type</code></td>\n",
       "<td>\n",
       "<p>alias for <code>assay_name</code>.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"ancombc_:_assay_name\">assay_name</code></td>\n",
       "<td>\n",
       "<p>character. Name of the count table in the data object\n",
       "(only applicable if data object is a <code>(Tree)SummarizedExperiment</code>).\n",
       "Default is &quot;counts&quot;.\n",
       "See <code>?SummarizedExperiment::assay</code> for more details.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"ancombc_:_rank\">rank</code></td>\n",
       "<td>\n",
       "<p>alias for <code>tax_level</code>.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"ancombc_:_tax_level\">tax_level</code></td>\n",
       "<td>\n",
       "<p>character. The taxonomic level of interest. The input data\n",
       "can be agglomerated at different taxonomic levels based on your research\n",
       "interest. Default is NULL, i.e., do not perform agglomeration, and the\n",
       "ANCOM-BC anlysis will be performed at the lowest taxonomic level of the\n",
       "input <code>data</code>.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"ancombc_:_phyloseq\">phyloseq</code></td>\n",
       "<td>\n",
       "<p>a <code>phyloseq</code> object. Will be deprecated.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"ancombc_:_formula\">formula</code></td>\n",
       "<td>\n",
       "<p>the character string expresses how microbial absolute\n",
       "abundances for each taxon depend on the variables in metadata. When\n",
       "specifying the <code>formula</code>, make sure to include the <code>group</code> variable\n",
       "in the formula if it is not NULL.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"ancombc_:_p_adj_method\">p_adj_method</code></td>\n",
       "<td>\n",
       "<p>character. method to adjust p-values. Default is &quot;holm&quot;.\n",
       "Options include &quot;holm&quot;, &quot;hochberg&quot;, &quot;hommel&quot;, &quot;bonferroni&quot;, &quot;BH&quot;, &quot;BY&quot;,\n",
       "&quot;fdr&quot;, &quot;none&quot;. See <code>?stats::p.adjust</code> for more details.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"ancombc_:_prv_cut\">prv_cut</code></td>\n",
       "<td>\n",
       "<p>a numerical fraction between 0 and 1. Taxa with prevalences\n",
       "(the proportion of samples in which the taxon is present)\n",
       "less than <code>prv_cut</code> will be excluded in the analysis. For example,\n",
       "if there are 100 samples, and a taxon has nonzero counts present in less than\n",
       "100*prv_cut samples, it will not be considered in the analysis.\n",
       "Default is 0.10.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"ancombc_:_lib_cut\">lib_cut</code></td>\n",
       "<td>\n",
       "<p>a numerical threshold for filtering samples based on library\n",
       "sizes. Samples with library sizes less than <code>lib_cut</code> will be\n",
       "excluded in the analysis. Default is 0, i.e. do not discard any sample.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"ancombc_:_group\">group</code></td>\n",
       "<td>\n",
       "<p>character. the name of the group variable in metadata.\n",
       "The <code>group</code> parameter should be a character string representing the name\n",
       "of the group variable in the metadata. The <code>group</code> variable should be\n",
       "discrete, meaning it consists of categorical values. Specifying the\n",
       "<code>group</code> variable is required if you are interested in detecting\n",
       "structural zeros and performing global tests. However, if these analyses are\n",
       "not of interest to you, you can leave the <code>group</code> parameter as NULL.\n",
       "If the <code>group</code> variable of interest contains only two categories, you\n",
       "can also leave the <code>group</code> parameter as NULL. Default is NULL.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"ancombc_:_struc_zero\">struc_zero</code></td>\n",
       "<td>\n",
       "<p>logical. whether to detect structural zeros based on\n",
       "<code>group</code>. Default is FALSE.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"ancombc_:_neg_lb\">neg_lb</code></td>\n",
       "<td>\n",
       "<p>logical. whether to classify a taxon as a structural zero using\n",
       "its asymptotic lower bound. Default is FALSE.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"ancombc_:_tol\">tol</code></td>\n",
       "<td>\n",
       "<p>numeric. the iteration convergence tolerance for the E-M\n",
       "algorithm. Default is 1e-05.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"ancombc_:_max_iter\">max_iter</code></td>\n",
       "<td>\n",
       "<p>numeric. the maximum number of iterations for the E-M\n",
       "algorithm. Default is 100.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"ancombc_:_conserve\">conserve</code></td>\n",
       "<td>\n",
       "<p>logical. whether to use a conservative variance estimator for\n",
       "the test statistic. It is recommended if the sample size is small and/or\n",
       "the number of differentially abundant taxa is believed to be large.\n",
       "Default is FALSE.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"ancombc_:_alpha\">alpha</code></td>\n",
       "<td>\n",
       "<p>numeric. level of significance. Default is 0.05.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"ancombc_:_global\">global</code></td>\n",
       "<td>\n",
       "<p>logical. whether to perform the global test. Default is FALSE.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"ancombc_:_n_cl\">n_cl</code></td>\n",
       "<td>\n",
       "<p>numeric. The number of nodes to be forked. For details, see\n",
       "<code>?parallel::makeCluster</code>. Default is 1 (no parallel computing).</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"ancombc_:_verbose\">verbose</code></td>\n",
       "<td>\n",
       "<p>logical. Whether to generate verbose output during the\n",
       "ANCOM-BC fitting process. Default is FALSE.</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>Details</h3>\n",
       "\n",
       "<p>A taxon is considered to have structural zeros in some (&gt;=1)\n",
       "groups if it is completely (or nearly completely) missing in these groups.\n",
       "For instance, suppose there are three groups: g1, g2, and g3.\n",
       "If the counts of taxon A in g1 are 0 but nonzero in g2 and g3,\n",
       "then taxon A will be considered to contain structural zeros in g1.\n",
       "In this example, taxon A is declared to be differentially abundant between\n",
       "g1 and g2, g1 and g3, and consequently, it is globally differentially\n",
       "abundant with respect to this group variable.\n",
       "Such taxa are not further analyzed using ANCOM-BC, but the results are\n",
       "summarized in the overall summary. For more details about the structural\n",
       "zeros, please go to the\n",
       "<a href=\"https://doi.org/10.3389/fmicb.2017.02114\">ANCOM-II</a> paper.\n",
       "Setting <code>neg_lb = TRUE</code> indicates that you are using both criteria\n",
       "stated in section 3.2 of\n",
       "<a href=\"https://doi.org/10.3389/fmicb.2017.02114\">ANCOM-II</a>\n",
       "to detect structural zeros; otherwise, the algorithm will only use the\n",
       "equation 1 in section 3.2 for declaring structural zeros. Generally, it is\n",
       "recommended to set <code>neg_lb = TRUE</code> when the sample size per group is\n",
       "relatively large (e.g. &gt; 30).\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Value</h3>\n",
       "\n",
       "<p>a <code>list</code> with components:\n",
       "</p>\n",
       "\n",
       "<ul>\n",
       "<li> <p><code>feature_table</code>, a <code>data.frame</code> of pre-processed\n",
       "(based on <code>prv_cut</code> and <code>lib_cut</code>) microbial count table.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>zero_ind</code>, a logical <code>data.frame</code> with TRUE\n",
       "indicating the taxon is detected to contain structural zeros in\n",
       "some specific groups.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>samp_frac</code>, a numeric vector of estimated sampling\n",
       "fractions in log scale (natural log).\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>delta_em</code>, estimated sample-specific biases\n",
       "through E-M algorithm.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>delta_wls</code>, estimated sample-specific biases through\n",
       "weighted least squares (WLS) algorithm.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>res</code>,  a <code>list</code> containing ANCOM-BC primary result,\n",
       "which consists of:\n",
       "</p>\n",
       "\n",
       "<ul>\n",
       "<li> <p><code>lfc</code>, a <code>data.frame</code> of log fold changes\n",
       "obtained from the ANCOM-BC log-linear (natural log) model.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>se</code>, a <code>data.frame</code> of standard errors (SEs) of\n",
       "<code>lfc</code>.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>W</code>, a <code>data.frame</code> of test statistics.\n",
       "<code>W = lfc/se</code>.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>p_val</code>, a <code>data.frame</code> of p-values. P-values are\n",
       "obtained from two-sided Z-test using the test statistic <code>W</code>.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>q_val</code>, a <code>data.frame</code> of adjusted p-values.\n",
       "Adjusted p-values are obtained by applying <code>p_adj_method</code>\n",
       "to <code>p_val</code>.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>diff_abn</code>, a logical <code>data.frame</code>. TRUE if the\n",
       "taxon has <code>q_val</code> less than <code>alpha</code>.\n",
       "</p>\n",
       "</li></ul>\n",
       "\n",
       "</li>\n",
       "<li> <p><code>res_global</code>,  a <code>data.frame</code> containing ANCOM-BC\n",
       "global test result for the variable specified in <code>group</code>,\n",
       "each column is:\n",
       "</p>\n",
       "\n",
       "<ul>\n",
       "<li> <p><code>W</code>, test statistics.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>p_val</code>, p-values, which are obtained from two-sided\n",
       "Chi-square test using <code>W</code>.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>q_val</code>, adjusted p-values. Adjusted p-values are\n",
       "obtained by applying <code>p_adj_method</code> to <code>p_val</code>.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>diff_abn</code>, A logical vector. TRUE if the taxon has\n",
       "<code>q_val</code> less than <code>alpha</code>.\n",
       "</p>\n",
       "</li></ul>\n",
       "\n",
       "</li></ul>\n",
       "\n",
       "\n",
       "\n",
       "<h3>Author(s)</h3>\n",
       "\n",
       "<p>Huang Lin\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>References</h3>\n",
       "\n",
       "<p>Kaul A, Mandal S, Davidov O, Peddada SD (2017).\n",
       "&ldquo;Analysis of microbiome data in the presence of excess zeros.&rdquo;\n",
       "<em>Frontiers in microbiology</em>, <b>8</b>, 2114.\n",
       "</p>\n",
       "<p>Lin H, Peddada SD (2020).\n",
       "&ldquo;Analysis of compositions of microbiomes with bias correction.&rdquo;\n",
       "<em>Nature communications</em>, <b>11</b>(1), 1&ndash;11.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>See Also</h3>\n",
       "\n",
       "<p><code>ancom</code> <code>ancombc2</code>\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Examples</h3>\n",
       "\n",
       "<pre><code class='language-R'>#===========Build a TreeSummarizedExperiment Object from Scratch=============\n",
       "library(mia)\n",
       "\n",
       "# microbial count table\n",
       "otu_mat = matrix(sample(1:100, 100, replace = TRUE), nrow = 10, ncol = 10)\n",
       "rownames(otu_mat) = paste0(\"taxon\", 1:nrow(otu_mat))\n",
       "colnames(otu_mat) = paste0(\"sample\", 1:ncol(otu_mat))\n",
       "assays = SimpleList(counts = otu_mat)\n",
       "\n",
       "# sample metadata\n",
       "smd = data.frame(group = sample(LETTERS[1:4], size = 10, replace = TRUE),\n",
       "                 row.names = paste0(\"sample\", 1:ncol(otu_mat)),\n",
       "                 stringsAsFactors = FALSE)\n",
       "smd = DataFrame(smd)\n",
       "\n",
       "# taxonomy table\n",
       "tax_tab = matrix(sample(letters, 70, replace = TRUE),\n",
       "                 nrow = nrow(otu_mat), ncol = 7)\n",
       "rownames(tax_tab) = rownames(otu_mat)\n",
       "colnames(tax_tab) = c(\"Kingdom\", \"Phylum\", \"Class\", \"Order\",\n",
       "                      \"Family\", \"Genus\", \"Species\")\n",
       "tax_tab = DataFrame(tax_tab)\n",
       "\n",
       "# create TSE\n",
       "tse = TreeSummarizedExperiment(assays = assays,\n",
       "                               colData = smd,\n",
       "                               rowData = tax_tab)\n",
       "\n",
       "# convert TSE to phyloseq\n",
       "pseq = makePhyloseqFromTreeSummarizedExperiment(tse)\n",
       "\n",
       "#========================Run ANCOMBC Using a Real Data=======================\n",
       "library(ANCOMBC)\n",
       "data(atlas1006, package = \"microbiome\")\n",
       "tse = mia::makeTreeSummarizedExperimentFromPhyloseq(atlas1006)\n",
       "\n",
       "# subset to baseline\n",
       "tse = tse[, tse$time == 0]\n",
       "\n",
       "# run ancombc function\n",
       "set.seed(123)\n",
       "out = ancombc(data = tse, assay_name = \"counts\",\n",
       "              tax_level = \"Family\", phyloseq = NULL,\n",
       "              formula = \"age + nationality + bmi_group\",\n",
       "              p_adj_method = \"holm\", prv_cut = 0.10, lib_cut = 1000,\n",
       "              group = \"bmi_group\", struc_zero = TRUE, neg_lb = FALSE,\n",
       "              tol = 1e-5, max_iter = 100, conserve = TRUE,\n",
       "              alpha = 0.05, global = TRUE, n_cl = 1, verbose = TRUE)\n",
       "\n",
       "res_prim = out$res\n",
       "res_global = out$res_global\n",
       "\n",
       "# to run ancombc using the phyloseq object\n",
       "tse_alt = agglomerateByRank(tse, \"Family\")\n",
       "pseq = makePhyloseqFromTreeSummarizedExperiment(tse_alt)\n",
       "set.seed(123)\n",
       "out = ancombc(data = NULL, assay_name = NULL,\n",
       "              tax_level = \"Family\", phyloseq = pseq,\n",
       "              formula = \"age + nationality + bmi_group\",\n",
       "              p_adj_method = \"holm\", prv_cut = 0.10, lib_cut = 1000,\n",
       "              group = \"bmi_group\", struc_zero = TRUE, neg_lb = FALSE,\n",
       "              tol = 1e-5, max_iter = 100, conserve = TRUE,\n",
       "              alpha = 0.05, global = TRUE, n_cl = 1, verbose = TRUE)\n",
       "\n",
       "</code></pre>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>ANCOMBC</em> version 2.4.0 ]</div>\n",
       "</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{ancombc}{Analysis of Compositions of Microbiomes with Bias Correction (ANCOM-BC)}{ancombc}\n",
       "%\n",
       "\\begin{Description}\n",
       "Determine taxa whose absolute abundances, per unit volume, of\n",
       "the ecosystem (e.g., gut) are significantly different with changes in the\n",
       "covariate of interest (e.g., group). The current version of\n",
       "\\code{ancombc} function implements Analysis of Compositions of Microbiomes\n",
       "with Bias Correction (ANCOM-BC) in cross-sectional data while allowing\n",
       "for covariate adjustment.\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Usage}\n",
       "\\begin{verbatim}\n",
       "ancombc(\n",
       "  data = NULL,\n",
       "  assay.type = NULL,\n",
       "  assay_name = \"counts\",\n",
       "  rank = NULL,\n",
       "  tax_level = NULL,\n",
       "  phyloseq = NULL,\n",
       "  formula,\n",
       "  p_adj_method = \"holm\",\n",
       "  prv_cut = 0.1,\n",
       "  lib_cut = 0,\n",
       "  group = NULL,\n",
       "  struc_zero = FALSE,\n",
       "  neg_lb = FALSE,\n",
       "  tol = 1e-05,\n",
       "  max_iter = 100,\n",
       "  conserve = FALSE,\n",
       "  alpha = 0.05,\n",
       "  global = FALSE,\n",
       "  n_cl = 1,\n",
       "  verbose = FALSE\n",
       ")\n",
       "\\end{verbatim}\n",
       "\\end{Usage}\n",
       "%\n",
       "\\begin{Arguments}\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{data}] the input data. The \\code{data} parameter should be either a\n",
       "\\code{phyloseq} or a \\code{TreeSummarizedExperiment} object, which\n",
       "consists of a feature table (microbial count table), a sample metadata table,\n",
       "a taxonomy table (optional), and a phylogenetic tree (optional).\n",
       "Ensure that the row names of the metadata table match the sample names in the\n",
       "feature table, and the row names of the taxonomy table match the taxon\n",
       "(feature) names in the feature table. For detailed information, refer to\n",
       "\\code{?phyloseq::phyloseq} or\n",
       "\\code{?TreeSummarizedExperiment::TreeSummarizedExperiment}.\n",
       "\n",
       "\\item[\\code{assay.type}] alias for \\code{assay\\_name}.\n",
       "\n",
       "\\item[\\code{assay\\_name}] character. Name of the count table in the data object\n",
       "(only applicable if data object is a \\code{(Tree)SummarizedExperiment}).\n",
       "Default is \"counts\".\n",
       "See \\code{?SummarizedExperiment::assay} for more details.\n",
       "\n",
       "\\item[\\code{rank}] alias for \\code{tax\\_level}.\n",
       "\n",
       "\\item[\\code{tax\\_level}] character. The taxonomic level of interest. The input data\n",
       "can be agglomerated at different taxonomic levels based on your research\n",
       "interest. Default is NULL, i.e., do not perform agglomeration, and the\n",
       "ANCOM-BC anlysis will be performed at the lowest taxonomic level of the\n",
       "input \\code{data}.\n",
       "\n",
       "\\item[\\code{phyloseq}] a \\code{phyloseq} object. Will be deprecated.\n",
       "\n",
       "\\item[\\code{formula}] the character string expresses how microbial absolute\n",
       "abundances for each taxon depend on the variables in metadata. When\n",
       "specifying the \\code{formula}, make sure to include the \\code{group} variable\n",
       "in the formula if it is not NULL.\n",
       "\n",
       "\\item[\\code{p\\_adj\\_method}] character. method to adjust p-values. Default is \"holm\".\n",
       "Options include \"holm\", \"hochberg\", \"hommel\", \"bonferroni\", \"BH\", \"BY\",\n",
       "\"fdr\", \"none\". See \\code{?stats::p.adjust} for more details.\n",
       "\n",
       "\\item[\\code{prv\\_cut}] a numerical fraction between 0 and 1. Taxa with prevalences\n",
       "(the proportion of samples in which the taxon is present)\n",
       "less than \\code{prv\\_cut} will be excluded in the analysis. For example,\n",
       "if there are 100 samples, and a taxon has nonzero counts present in less than\n",
       "100*prv\\_cut samples, it will not be considered in the analysis.\n",
       "Default is 0.10.\n",
       "\n",
       "\\item[\\code{lib\\_cut}] a numerical threshold for filtering samples based on library\n",
       "sizes. Samples with library sizes less than \\code{lib\\_cut} will be\n",
       "excluded in the analysis. Default is 0, i.e. do not discard any sample.\n",
       "\n",
       "\\item[\\code{group}] character. the name of the group variable in metadata.\n",
       "The \\code{group} parameter should be a character string representing the name\n",
       "of the group variable in the metadata. The \\code{group} variable should be\n",
       "discrete, meaning it consists of categorical values. Specifying the\n",
       "\\code{group} variable is required if you are interested in detecting\n",
       "structural zeros and performing global tests. However, if these analyses are\n",
       "not of interest to you, you can leave the \\code{group} parameter as NULL.\n",
       "If the \\code{group} variable of interest contains only two categories, you\n",
       "can also leave the \\code{group} parameter as NULL. Default is NULL.\n",
       "\n",
       "\\item[\\code{struc\\_zero}] logical. whether to detect structural zeros based on\n",
       "\\code{group}. Default is FALSE.\n",
       "\n",
       "\\item[\\code{neg\\_lb}] logical. whether to classify a taxon as a structural zero using\n",
       "its asymptotic lower bound. Default is FALSE.\n",
       "\n",
       "\\item[\\code{tol}] numeric. the iteration convergence tolerance for the E-M\n",
       "algorithm. Default is 1e-05.\n",
       "\n",
       "\\item[\\code{max\\_iter}] numeric. the maximum number of iterations for the E-M\n",
       "algorithm. Default is 100.\n",
       "\n",
       "\\item[\\code{conserve}] logical. whether to use a conservative variance estimator for\n",
       "the test statistic. It is recommended if the sample size is small and/or\n",
       "the number of differentially abundant taxa is believed to be large.\n",
       "Default is FALSE.\n",
       "\n",
       "\\item[\\code{alpha}] numeric. level of significance. Default is 0.05.\n",
       "\n",
       "\\item[\\code{global}] logical. whether to perform the global test. Default is FALSE.\n",
       "\n",
       "\\item[\\code{n\\_cl}] numeric. The number of nodes to be forked. For details, see\n",
       "\\code{?parallel::makeCluster}. Default is 1 (no parallel computing).\n",
       "\n",
       "\\item[\\code{verbose}] logical. Whether to generate verbose output during the\n",
       "ANCOM-BC fitting process. Default is FALSE.\n",
       "\\end{ldescription}\n",
       "\\end{Arguments}\n",
       "%\n",
       "\\begin{Details}\n",
       "A taxon is considered to have structural zeros in some (>=1)\n",
       "groups if it is completely (or nearly completely) missing in these groups.\n",
       "For instance, suppose there are three groups: g1, g2, and g3.\n",
       "If the counts of taxon A in g1 are 0 but nonzero in g2 and g3,\n",
       "then taxon A will be considered to contain structural zeros in g1.\n",
       "In this example, taxon A is declared to be differentially abundant between\n",
       "g1 and g2, g1 and g3, and consequently, it is globally differentially\n",
       "abundant with respect to this group variable.\n",
       "Such taxa are not further analyzed using ANCOM-BC, but the results are\n",
       "summarized in the overall summary. For more details about the structural\n",
       "zeros, please go to the\n",
       "\\Rhref{https://doi.org/10.3389/fmicb.2017.02114}{ANCOM-II} paper.\n",
       "Setting \\code{neg\\_lb = TRUE} indicates that you are using both criteria\n",
       "stated in section 3.2 of\n",
       "\\Rhref{https://doi.org/10.3389/fmicb.2017.02114}{ANCOM-II}\n",
       "to detect structural zeros; otherwise, the algorithm will only use the\n",
       "equation 1 in section 3.2 for declaring structural zeros. Generally, it is\n",
       "recommended to set \\code{neg\\_lb = TRUE} when the sample size per group is\n",
       "relatively large (e.g. > 30).\n",
       "\\end{Details}\n",
       "%\n",
       "\\begin{Value}\n",
       "a \\code{list} with components:\n",
       "\\begin{itemize}\n",
       "\n",
       "\\item{}  \\code{feature\\_table}, a \\code{data.frame} of pre-processed\n",
       "(based on \\code{prv\\_cut} and \\code{lib\\_cut}) microbial count table.\n",
       "\\item{}  \\code{zero\\_ind}, a logical \\code{data.frame} with TRUE\n",
       "indicating the taxon is detected to contain structural zeros in\n",
       "some specific groups.\n",
       "\\item{}  \\code{samp\\_frac}, a numeric vector of estimated sampling\n",
       "fractions in log scale (natural log).\n",
       "\\item{}  \\code{delta\\_em}, estimated sample-specific biases\n",
       "through E-M algorithm.\n",
       "\\item{}  \\code{delta\\_wls}, estimated sample-specific biases through\n",
       "weighted least squares (WLS) algorithm.\n",
       "\\item{}  \\code{res},  a \\code{list} containing ANCOM-BC primary result,\n",
       "which consists of:\n",
       "\\begin{itemize}\n",
       "\n",
       "\\item{}  \\code{lfc}, a \\code{data.frame} of log fold changes\n",
       "obtained from the ANCOM-BC log-linear (natural log) model.\n",
       "\\item{}  \\code{se}, a \\code{data.frame} of standard errors (SEs) of\n",
       "\\code{lfc}.\n",
       "\\item{}  \\code{W}, a \\code{data.frame} of test statistics.\n",
       "\\code{W = lfc/se}.\n",
       "\\item{}  \\code{p\\_val}, a \\code{data.frame} of p-values. P-values are\n",
       "obtained from two-sided Z-test using the test statistic \\code{W}.\n",
       "\\item{}  \\code{q\\_val}, a \\code{data.frame} of adjusted p-values.\n",
       "Adjusted p-values are obtained by applying \\code{p\\_adj\\_method}\n",
       "to \\code{p\\_val}.\n",
       "\\item{}  \\code{diff\\_abn}, a logical \\code{data.frame}. TRUE if the\n",
       "taxon has \\code{q\\_val} less than \\code{alpha}.\n",
       "\n",
       "\\end{itemize}\n",
       "\n",
       "\\item{}  \\code{res\\_global},  a \\code{data.frame} containing ANCOM-BC\n",
       "global test result for the variable specified in \\code{group},\n",
       "each column is:\n",
       "\\begin{itemize}\n",
       "\n",
       "\\item{}  \\code{W}, test statistics.\n",
       "\\item{}  \\code{p\\_val}, p-values, which are obtained from two-sided\n",
       "Chi-square test using \\code{W}.\n",
       "\\item{}  \\code{q\\_val}, adjusted p-values. Adjusted p-values are\n",
       "obtained by applying \\code{p\\_adj\\_method} to \\code{p\\_val}.\n",
       "\\item{}  \\code{diff\\_abn}, A logical vector. TRUE if the taxon has\n",
       "\\code{q\\_val} less than \\code{alpha}.\n",
       "\n",
       "\\end{itemize}\n",
       "\n",
       "\n",
       "\\end{itemize}\n",
       "\n",
       "\\end{Value}\n",
       "%\n",
       "\\begin{Author}\n",
       "Huang Lin\n",
       "\\end{Author}\n",
       "%\n",
       "\\begin{References}\n",
       "Kaul A, Mandal S, Davidov O, Peddada SD (2017).\n",
       "``Analysis of microbiome data in the presence of excess zeros.''\n",
       "\\emph{Frontiers in microbiology}, \\bold{8}, 2114.\n",
       "\n",
       "Lin H, Peddada SD (2020).\n",
       "``Analysis of compositions of microbiomes with bias correction.''\n",
       "\\emph{Nature communications}, \\bold{11}(1), 1--11.\n",
       "\\end{References}\n",
       "%\n",
       "\\begin{SeeAlso}\n",
       "\\code{\\LinkA{ancom}{ancom}} \\code{\\LinkA{ancombc2}{ancombc2}}\n",
       "\\end{SeeAlso}\n",
       "%\n",
       "\\begin{Examples}\n",
       "\\begin{ExampleCode}\n",
       "#===========Build a TreeSummarizedExperiment Object from Scratch=============\n",
       "library(mia)\n",
       "\n",
       "# microbial count table\n",
       "otu_mat = matrix(sample(1:100, 100, replace = TRUE), nrow = 10, ncol = 10)\n",
       "rownames(otu_mat) = paste0(\"taxon\", 1:nrow(otu_mat))\n",
       "colnames(otu_mat) = paste0(\"sample\", 1:ncol(otu_mat))\n",
       "assays = SimpleList(counts = otu_mat)\n",
       "\n",
       "# sample metadata\n",
       "smd = data.frame(group = sample(LETTERS[1:4], size = 10, replace = TRUE),\n",
       "                 row.names = paste0(\"sample\", 1:ncol(otu_mat)),\n",
       "                 stringsAsFactors = FALSE)\n",
       "smd = DataFrame(smd)\n",
       "\n",
       "# taxonomy table\n",
       "tax_tab = matrix(sample(letters, 70, replace = TRUE),\n",
       "                 nrow = nrow(otu_mat), ncol = 7)\n",
       "rownames(tax_tab) = rownames(otu_mat)\n",
       "colnames(tax_tab) = c(\"Kingdom\", \"Phylum\", \"Class\", \"Order\",\n",
       "                      \"Family\", \"Genus\", \"Species\")\n",
       "tax_tab = DataFrame(tax_tab)\n",
       "\n",
       "# create TSE\n",
       "tse = TreeSummarizedExperiment(assays = assays,\n",
       "                               colData = smd,\n",
       "                               rowData = tax_tab)\n",
       "\n",
       "# convert TSE to phyloseq\n",
       "pseq = makePhyloseqFromTreeSummarizedExperiment(tse)\n",
       "\n",
       "#========================Run ANCOMBC Using a Real Data=======================\n",
       "library(ANCOMBC)\n",
       "data(atlas1006, package = \"microbiome\")\n",
       "tse = mia::makeTreeSummarizedExperimentFromPhyloseq(atlas1006)\n",
       "\n",
       "# subset to baseline\n",
       "tse = tse[, tse$time == 0]\n",
       "\n",
       "# run ancombc function\n",
       "set.seed(123)\n",
       "out = ancombc(data = tse, assay_name = \"counts\",\n",
       "              tax_level = \"Family\", phyloseq = NULL,\n",
       "              formula = \"age + nationality + bmi_group\",\n",
       "              p_adj_method = \"holm\", prv_cut = 0.10, lib_cut = 1000,\n",
       "              group = \"bmi_group\", struc_zero = TRUE, neg_lb = FALSE,\n",
       "              tol = 1e-5, max_iter = 100, conserve = TRUE,\n",
       "              alpha = 0.05, global = TRUE, n_cl = 1, verbose = TRUE)\n",
       "\n",
       "res_prim = out$res\n",
       "res_global = out$res_global\n",
       "\n",
       "# to run ancombc using the phyloseq object\n",
       "tse_alt = agglomerateByRank(tse, \"Family\")\n",
       "pseq = makePhyloseqFromTreeSummarizedExperiment(tse_alt)\n",
       "set.seed(123)\n",
       "out = ancombc(data = NULL, assay_name = NULL,\n",
       "              tax_level = \"Family\", phyloseq = pseq,\n",
       "              formula = \"age + nationality + bmi_group\",\n",
       "              p_adj_method = \"holm\", prv_cut = 0.10, lib_cut = 1000,\n",
       "              group = \"bmi_group\", struc_zero = TRUE, neg_lb = FALSE,\n",
       "              tol = 1e-5, max_iter = 100, conserve = TRUE,\n",
       "              alpha = 0.05, global = TRUE, n_cl = 1, verbose = TRUE)\n",
       "\n",
       "\\end{ExampleCode}\n",
       "\\end{Examples}"
      ],
      "text/plain": [
       "ancombc                package:ANCOMBC                 R Documentation\n",
       "\n",
       "_\bA_\bn_\ba_\bl_\by_\bs_\bi_\bs _\bo_\bf _\bC_\bo_\bm_\bp_\bo_\bs_\bi_\bt_\bi_\bo_\bn_\bs _\bo_\bf _\bM_\bi_\bc_\br_\bo_\bb_\bi_\bo_\bm_\be_\bs _\bw_\bi_\bt_\bh _\bB_\bi_\ba_\bs _\bC_\bo_\br_\br_\be_\bc_\bt_\bi_\bo_\bn (_\bA_\bN_\bC_\bO_\bM-_\bB_\bC)\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     Determine taxa whose absolute abundances, per unit volume, of the\n",
       "     ecosystem (e.g., gut) are significantly different with changes in\n",
       "     the covariate of interest (e.g., group). The current version of\n",
       "     'ancombc' function implements Analysis of Compositions of\n",
       "     Microbiomes with Bias Correction (ANCOM-BC) in cross-sectional\n",
       "     data while allowing for covariate adjustment.\n",
       "\n",
       "_\bU_\bs_\ba_\bg_\be:\n",
       "\n",
       "     ancombc(\n",
       "       data = NULL,\n",
       "       assay.type = NULL,\n",
       "       assay_name = \"counts\",\n",
       "       rank = NULL,\n",
       "       tax_level = NULL,\n",
       "       phyloseq = NULL,\n",
       "       formula,\n",
       "       p_adj_method = \"holm\",\n",
       "       prv_cut = 0.1,\n",
       "       lib_cut = 0,\n",
       "       group = NULL,\n",
       "       struc_zero = FALSE,\n",
       "       neg_lb = FALSE,\n",
       "       tol = 1e-05,\n",
       "       max_iter = 100,\n",
       "       conserve = FALSE,\n",
       "       alpha = 0.05,\n",
       "       global = FALSE,\n",
       "       n_cl = 1,\n",
       "       verbose = FALSE\n",
       "     )\n",
       "     \n",
       "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
       "\n",
       "    data: the input data. The 'data' parameter should be either a\n",
       "          'phyloseq' or a 'TreeSummarizedExperiment' object, which\n",
       "          consists of a feature table (microbial count table), a sample\n",
       "          metadata table, a taxonomy table (optional), and a\n",
       "          phylogenetic tree (optional). Ensure that the row names of\n",
       "          the metadata table match the sample names in the feature\n",
       "          table, and the row names of the taxonomy table match the\n",
       "          taxon (feature) names in the feature table. For detailed\n",
       "          information, refer to '?phyloseq::phyloseq' or\n",
       "          '?TreeSummarizedExperiment::TreeSummarizedExperiment'.\n",
       "\n",
       "assay.type: alias for 'assay_name'.\n",
       "\n",
       "assay_name: character. Name of the count table in the data object (only\n",
       "          applicable if data object is a '(Tree)SummarizedExperiment').\n",
       "          Default is \"counts\". See '?SummarizedExperiment::assay' for\n",
       "          more details.\n",
       "\n",
       "    rank: alias for 'tax_level'.\n",
       "\n",
       "tax_level: character. The taxonomic level of interest. The input data\n",
       "          can be agglomerated at different taxonomic levels based on\n",
       "          your research interest. Default is NULL, i.e., do not perform\n",
       "          agglomeration, and the ANCOM-BC anlysis will be performed at\n",
       "          the lowest taxonomic level of the input 'data'.\n",
       "\n",
       "phyloseq: a 'phyloseq' object. Will be deprecated.\n",
       "\n",
       " formula: the character string expresses how microbial absolute\n",
       "          abundances for each taxon depend on the variables in\n",
       "          metadata. When specifying the 'formula', make sure to include\n",
       "          the 'group' variable in the formula if it is not NULL.\n",
       "\n",
       "p_adj_method: character. method to adjust p-values. Default is \"holm\".\n",
       "          Options include \"holm\", \"hochberg\", \"hommel\", \"bonferroni\",\n",
       "          \"BH\", \"BY\", \"fdr\", \"none\". See '?stats::p.adjust' for more\n",
       "          details.\n",
       "\n",
       " prv_cut: a numerical fraction between 0 and 1. Taxa with prevalences\n",
       "          (the proportion of samples in which the taxon is present)\n",
       "          less than 'prv_cut' will be excluded in the analysis. For\n",
       "          example, if there are 100 samples, and a taxon has nonzero\n",
       "          counts present in less than 100*prv_cut samples, it will not\n",
       "          be considered in the analysis. Default is 0.10.\n",
       "\n",
       " lib_cut: a numerical threshold for filtering samples based on library\n",
       "          sizes. Samples with library sizes less than 'lib_cut' will be\n",
       "          excluded in the analysis. Default is 0, i.e. do not discard\n",
       "          any sample.\n",
       "\n",
       "   group: character. the name of the group variable in metadata. The\n",
       "          'group' parameter should be a character string representing\n",
       "          the name of the group variable in the metadata. The 'group'\n",
       "          variable should be discrete, meaning it consists of\n",
       "          categorical values. Specifying the 'group' variable is\n",
       "          required if you are interested in detecting structural zeros\n",
       "          and performing global tests. However, if these analyses are\n",
       "          not of interest to you, you can leave the 'group' parameter\n",
       "          as NULL. If the 'group' variable of interest contains only\n",
       "          two categories, you can also leave the 'group' parameter as\n",
       "          NULL. Default is NULL.\n",
       "\n",
       "struc_zero: logical. whether to detect structural zeros based on\n",
       "          'group'. Default is FALSE.\n",
       "\n",
       "  neg_lb: logical. whether to classify a taxon as a structural zero\n",
       "          using its asymptotic lower bound. Default is FALSE.\n",
       "\n",
       "     tol: numeric. the iteration convergence tolerance for the E-M\n",
       "          algorithm. Default is 1e-05.\n",
       "\n",
       "max_iter: numeric. the maximum number of iterations for the E-M\n",
       "          algorithm. Default is 100.\n",
       "\n",
       "conserve: logical. whether to use a conservative variance estimator for\n",
       "          the test statistic. It is recommended if the sample size is\n",
       "          small and/or the number of differentially abundant taxa is\n",
       "          believed to be large. Default is FALSE.\n",
       "\n",
       "   alpha: numeric. level of significance. Default is 0.05.\n",
       "\n",
       "  global: logical. whether to perform the global test. Default is\n",
       "          FALSE.\n",
       "\n",
       "    n_cl: numeric. The number of nodes to be forked. For details, see\n",
       "          '?parallel::makeCluster'. Default is 1 (no parallel\n",
       "          computing).\n",
       "\n",
       " verbose: logical. Whether to generate verbose output during the\n",
       "          ANCOM-BC fitting process. Default is FALSE.\n",
       "\n",
       "_\bD_\be_\bt_\ba_\bi_\bl_\bs:\n",
       "\n",
       "     A taxon is considered to have structural zeros in some (>=1)\n",
       "     groups if it is completely (or nearly completely) missing in these\n",
       "     groups. For instance, suppose there are three groups: g1, g2, and\n",
       "     g3. If the counts of taxon A in g1 are 0 but nonzero in g2 and g3,\n",
       "     then taxon A will be considered to contain structural zeros in g1.\n",
       "     In this example, taxon A is declared to be differentially abundant\n",
       "     between g1 and g2, g1 and g3, and consequently, it is globally\n",
       "     differentially abundant with respect to this group variable. Such\n",
       "     taxa are not further analyzed using ANCOM-BC, but the results are\n",
       "     summarized in the overall summary. For more details about the\n",
       "     structural zeros, please go to the ANCOM-II paper. Setting 'neg_lb\n",
       "     = TRUE' indicates that you are using both criteria stated in\n",
       "     section 3.2 of ANCOM-II to detect structural zeros; otherwise, the\n",
       "     algorithm will only use the equation 1 in section 3.2 for\n",
       "     declaring structural zeros. Generally, it is recommended to set\n",
       "     'neg_lb = TRUE' when the sample size per group is relatively large\n",
       "     (e.g. > 30).\n",
       "\n",
       "_\bV_\ba_\bl_\bu_\be:\n",
       "\n",
       "     a 'list' with components:\n",
       "\n",
       "        * 'feature_table', a 'data.frame' of pre-processed (based on\n",
       "          'prv_cut' and 'lib_cut') microbial count table.\n",
       "\n",
       "        * 'zero_ind', a logical 'data.frame' with TRUE indicating the\n",
       "          taxon is detected to contain structural zeros in some\n",
       "          specific groups.\n",
       "\n",
       "        * 'samp_frac', a numeric vector of estimated sampling fractions\n",
       "          in log scale (natural log).\n",
       "\n",
       "        * 'delta_em', estimated sample-specific biases through E-M\n",
       "          algorithm.\n",
       "\n",
       "        * 'delta_wls', estimated sample-specific biases through\n",
       "          weighted least squares (WLS) algorithm.\n",
       "\n",
       "        * 'res', a 'list' containing ANCOM-BC primary result, which\n",
       "          consists of:\n",
       "\n",
       "            * 'lfc', a 'data.frame' of log fold changes obtained from\n",
       "              the ANCOM-BC log-linear (natural log) model.\n",
       "\n",
       "            * 'se', a 'data.frame' of standard errors (SEs) of 'lfc'.\n",
       "\n",
       "            * 'W', a 'data.frame' of test statistics.  'W = lfc/se'.\n",
       "\n",
       "            * 'p_val', a 'data.frame' of p-values. P-values are\n",
       "              obtained from two-sided Z-test using the test statistic\n",
       "              'W'.\n",
       "\n",
       "            * 'q_val', a 'data.frame' of adjusted p-values.  Adjusted\n",
       "              p-values are obtained by applying 'p_adj_method' to\n",
       "              'p_val'.\n",
       "\n",
       "            * 'diff_abn', a logical 'data.frame'. TRUE if the taxon has\n",
       "              'q_val' less than 'alpha'.\n",
       "\n",
       "        * 'res_global', a 'data.frame' containing ANCOM-BC global test\n",
       "          result for the variable specified in 'group', each column is:\n",
       "\n",
       "            * 'W', test statistics.\n",
       "\n",
       "            * 'p_val', p-values, which are obtained from two-sided\n",
       "              Chi-square test using 'W'.\n",
       "\n",
       "            * 'q_val', adjusted p-values. Adjusted p-values are\n",
       "              obtained by applying 'p_adj_method' to 'p_val'.\n",
       "\n",
       "            * 'diff_abn', A logical vector. TRUE if the taxon has\n",
       "              'q_val' less than 'alpha'.\n",
       "\n",
       "_\bA_\bu_\bt_\bh_\bo_\br(_\bs):\n",
       "\n",
       "     Huang Lin\n",
       "\n",
       "_\bR_\be_\bf_\be_\br_\be_\bn_\bc_\be_\bs:\n",
       "\n",
       "     Kaul A, Mandal S, Davidov O, Peddada SD (2017). \"Analysis of\n",
       "     microbiome data in the presence of excess zeros.\" _Frontiers in\n",
       "     microbiology_, *8*, 2114.\n",
       "\n",
       "     Lin H, Peddada SD (2020). \"Analysis of compositions of microbiomes\n",
       "     with bias correction.\" _Nature communications_, *11*(1), 1-11.\n",
       "\n",
       "_\bS_\be_\be _\bA_\bl_\bs_\bo:\n",
       "\n",
       "     'ancom' 'ancombc2'\n",
       "\n",
       "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
       "\n",
       "     #===========Build a TreeSummarizedExperiment Object from Scratch=============\n",
       "     library(mia)\n",
       "     \n",
       "     # microbial count table\n",
       "     otu_mat = matrix(sample(1:100, 100, replace = TRUE), nrow = 10, ncol = 10)\n",
       "     rownames(otu_mat) = paste0(\"taxon\", 1:nrow(otu_mat))\n",
       "     colnames(otu_mat) = paste0(\"sample\", 1:ncol(otu_mat))\n",
       "     assays = SimpleList(counts = otu_mat)\n",
       "     \n",
       "     # sample metadata\n",
       "     smd = data.frame(group = sample(LETTERS[1:4], size = 10, replace = TRUE),\n",
       "                      row.names = paste0(\"sample\", 1:ncol(otu_mat)),\n",
       "                      stringsAsFactors = FALSE)\n",
       "     smd = DataFrame(smd)\n",
       "     \n",
       "     # taxonomy table\n",
       "     tax_tab = matrix(sample(letters, 70, replace = TRUE),\n",
       "                      nrow = nrow(otu_mat), ncol = 7)\n",
       "     rownames(tax_tab) = rownames(otu_mat)\n",
       "     colnames(tax_tab) = c(\"Kingdom\", \"Phylum\", \"Class\", \"Order\",\n",
       "                           \"Family\", \"Genus\", \"Species\")\n",
       "     tax_tab = DataFrame(tax_tab)\n",
       "     \n",
       "     # create TSE\n",
       "     tse = TreeSummarizedExperiment(assays = assays,\n",
       "                                    colData = smd,\n",
       "                                    rowData = tax_tab)\n",
       "     \n",
       "     # convert TSE to phyloseq\n",
       "     pseq = makePhyloseqFromTreeSummarizedExperiment(tse)\n",
       "     \n",
       "     #========================Run ANCOMBC Using a Real Data=======================\n",
       "     library(ANCOMBC)\n",
       "     data(atlas1006, package = \"microbiome\")\n",
       "     tse = mia::makeTreeSummarizedExperimentFromPhyloseq(atlas1006)\n",
       "     \n",
       "     # subset to baseline\n",
       "     tse = tse[, tse$time == 0]\n",
       "     \n",
       "     # run ancombc function\n",
       "     set.seed(123)\n",
       "     out = ancombc(data = tse, assay_name = \"counts\",\n",
       "                   tax_level = \"Family\", phyloseq = NULL,\n",
       "                   formula = \"age + nationality + bmi_group\",\n",
       "                   p_adj_method = \"holm\", prv_cut = 0.10, lib_cut = 1000,\n",
       "                   group = \"bmi_group\", struc_zero = TRUE, neg_lb = FALSE,\n",
       "                   tol = 1e-5, max_iter = 100, conserve = TRUE,\n",
       "                   alpha = 0.05, global = TRUE, n_cl = 1, verbose = TRUE)\n",
       "     \n",
       "     res_prim = out$res\n",
       "     res_global = out$res_global\n",
       "     \n",
       "     # to run ancombc using the phyloseq object\n",
       "     tse_alt = agglomerateByRank(tse, \"Family\")\n",
       "     pseq = makePhyloseqFromTreeSummarizedExperiment(tse_alt)\n",
       "     set.seed(123)\n",
       "     out = ancombc(data = NULL, assay_name = NULL,\n",
       "                   tax_level = \"Family\", phyloseq = pseq,\n",
       "                   formula = \"age + nationality + bmi_group\",\n",
       "                   p_adj_method = \"holm\", prv_cut = 0.10, lib_cut = 1000,\n",
       "                   group = \"bmi_group\", struc_zero = TRUE, neg_lb = FALSE,\n",
       "                   tol = 1e-5, max_iter = 100, conserve = TRUE,\n",
       "                   alpha = 0.05, global = TRUE, n_cl = 1, verbose = TRUE)\n",
       "     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "help(ancombc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d3d866e-2970-4442-936f-c00ddbaaf94f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
